#+TITLE: Clarifying confusion in scRNA-seq analysis
#+SETUPFILE: setup.org
#+OPTIONS: toc:nil

The development of single cell RNA sequencing (scRNA-seq) technologies has
greatly facilitated investigation of important biological questions that were
previously difficult or impossible to study. However, how to model and analyze
scRNA-seq data has been the subject of considerable confusion and debate. We
develop a simple way of thinking and talking about scRNA-seq data that can
clarify much of this confusion. We outline a general framework which explicitly
models, and distinguishes between, the measurement process and variation in
actual expression levels between cells. Although our simple framework cannot on
its own address all issues that arise in scRNA-seq data, we show how it can
help clarify thinking, guide interpretation of existing analysis methods,
address questions of biological interest, and suggest directions for
development of new methods and analysis strategies.

Sarkar and Stephens,
[[https://dx.doi.org/10.1101/2020.04.07.030007]["Separating measurement and
expression models clarifies confusion in single cell RNA-seq analysis."]]
/BioRxiv/ 2020.

* Empirical results

- [[file:gof.org][Goodness of fit of deconvolved distributions]] ([[http://shiny.stephenslab.uchicago.edu:5009/gof][interactive browser]])
- [[file:lra.org][Low rank approximation]]
- [[file:fitted-values.org][Comparison of fitted values]]
- [[file:mlcb.org][MLCB figures]]
- [[file:figures.org][Manuscript figures]]

* Additional analyses

- [[file:deconvolution.org][Comparison of distribution deconvolution methods]]
- [[file:deconv-examples.org][Distribution deconvolution examples]]
- [[file:marker-genes.org][Marker genes in /in silico/ mixtures]]
  ([[http://shiny.stephenslab.uchicago.edu:5009/markers][interactive browser]])
- [[file:immune-census.org][Expression variation in Census of Immune Cells]]
- [[file:chromium3.org][Expression variation in 10X v3 PBMC data]]
- [[file:brain-dronc-seq.org][Expression variation in human brain cells]]
- [[file:retina.org][Expression variation in human retina cells]]
- [[file:liver.org][Expression variation in human liver]]
- [[file:imputation.org][Imputation of count matrices]]
- [[file:llr.org][Marginal log likelihood comparison of expression models]]

* Technical details

- [[file:pitfalls.org][Technical zero-generating mechanism in scRNA-seq data]]
- [[file:modes.org][Estimating the number of modes]]
- [[file:browser.org][Interactive simulations]]
- [[file:capture.org][Effect of capture rate on sampling variation]]
- [[file:ash-estmode.org][Speed up ash mode estimation]]
- [[file:mix-nb.org][Mixture of negative binomials]]
- [[file:gamma.org][Convergence of Gamma deconvolution]]
- [[file:poisson.org][Deconvolution of near-Poisson data]]
- [[file:validation.org][Validation set log likelihood comparison]]
- [[file:npmle.org][NPMLE on C1 spike-in data]]
- [[file:deconv-de.org][Differential expression from distribution deconvolution]]
- [[file:ep.org][Posterior approximation for scRNA-seq data]]
- [[file:link.org][Link functions in Poisson LRA]]
- [[file:random-effect.org][Distributional assumptions in Poisson LRA]]
- [[file:pois-truncnorm.org][Poisson-truncated normal data]]
- [[file:gamma-mix.org][Poisson-unimodal Gamma mixture model]]
- [[file:pois-gam-lik.org][Poisson-Gamma log likelihood surface]]
- [[file:vae.org][Variational autoencoders for scRNA-seq data]]
- [[file:wnbmf.org][Weighted Negative Binomial Matrix Factorization]]
- [[file:ash-pois-log-link.org][Log link in Poisson ash]]
- [[file:transforms.org][Transformations of deconvolved gene expression
  distributions]]
- [[file:gaussian.org][Gaussian methods]]
- [[file:quantiles.org][Randomized quantiles]]
- [[file:independence.org][Relaxing the independence assumption on expression models]]
