#+TITLE: Clarifying confusion in scRNA-seq analysis
#+SETUPFILE: setup.org
#+OPTIONS: toc:nil

The high proportion of zeros in typical scRNA-seq datasets has led to
widespread but inconsistent use of terminology such as "dropout" and "missing
data". Here, we argue that much of this terminology is unhelpful and confusing,
and outline simple ideas to help reduce confusion. These include: (1) observed
scRNA-seq counts reflect both true gene expression levels and measurement
error, and carefully distinguishing these contributions helps clarify thinking;
and (2) method development should start with a Poisson measurement model,
rather than more complex models, because it is simple and generally consistent
with existing data. We outline how several existing methods can be viewed
within this framework and highlight how these methods differ in their
assumptions about expression variation. We also illustrate how our perspective
helps address questions of biological interest, such as whether mRNA expression
levels are multimodal among cells.

Sarkar, A. and Stephens,
M. "[[https://www.nature.com/articles/s41588-021-00873-4][Separating
measurement and expression models clarifies confusion in single-cell RNA
sequencing analysis.]]" /Nature Genetics/ (2021).

* Empirical results

- [[file:llr.org][Marginal log likelihood comparison of expression models]]
- [[file:ebnbm.org][Negative binomial measurement model]]
- [[file:lra.org][Low rank approximation]]
- [[file:fitted-values.org][Comparison of fitted values]]
- [[file:mlcb.org][MLCB figures]]
- [[file:figures.org][Manuscript figures]]

* Additional analyses

- [[file:deconvolution.org][Comparison of distribution deconvolution methods]]
- [[file:deconv-examples.org][Distribution deconvolution examples]]
- [[file:gof.org][Goodness of fit of deconvolved distributions]] ([[https://shiny.stephenslab.uchicago.edu/scmodes/gof][interactive browser]])
- [[file:marker-genes.org][Marker genes in /in silico/ mixtures]]
  ([[http://shiny.stephenslab.uchicago.edu:5009/markers][interactive browser]])
- [[file:immune-census.org][Expression variation in Census of Immune Cells]]
- [[file:chromium3.org][Expression variation in 10X v3 PBMC data]]
- [[file:brain-dronc-seq.org][Expression variation in human brain cells]]
- [[file:retina.org][Expression variation in human retina cells]]
- [[file:liver.org][Expression variation in human liver]]
- [[file:imputation.org][Imputation of count matrices]]

* Technical details

- [[file:pitfalls.org][Technical zero-generating mechanism in scRNA-seq data]]
- [[file:modes.org][Estimating the number of modes]]
- [[file:browser.org][Interactive simulations]]
- [[file:capture.org][Effect of capture rate on sampling variation]]
- [[file:ash-estmode.org][Speed up ash mode estimation]]
- [[file:mix-nb.org][Mixture of negative binomials]]
- [[file:gamma.org][Convergence of Gamma deconvolution]]
- [[file:poisson.org][Deconvolution of near-Poisson data]]
- [[file:validation.org][Validation set log likelihood comparison]]
- [[file:npmle.org][NPMLE on C1 spike-in data]]
- [[file:deconv-de.org][Differential expression from distribution deconvolution]]
- [[file:ep.org][Posterior approximation for scRNA-seq data]]
- [[file:link.org][Link functions in Poisson LRA]]
- [[file:random-effect.org][Distributional assumptions in Poisson LRA]]
- [[file:pois-truncnorm.org][Poisson-truncated normal data]]
- [[file:gamma-mix.org][Poisson-unimodal Gamma mixture model]]
- [[file:pois-gam-lik.org][Poisson-Gamma log likelihood surface]]
- [[file:vae.org][Variational autoencoders for scRNA-seq data]]
- [[file:wnbmf.org][Weighted Negative Binomial Matrix Factorization]]
- [[file:ash-pois-log-link.org][Log link in Poisson ash]]
- [[file:transforms.org][Transformations of deconvolved gene expression
  distributions]]
- [[file:gaussian.org][Gaussian methods]]
- [[file:quantiles.org][Randomized quantiles]]
- [[file:independence.org][Relaxing the independence assumption on expression models]]
- [[file:zinb-em.org][EM algorithm for point-Gamma expression model]]
- [[file:npmle-grid.org][Iterative refinement of NPMLE grid in ashr]]
- [[file:halfuniform.org][Uniform vs half-uniform mixture prior]]
