<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-08-27 Tue 23:40 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Comparison of expression deconvolution approaches</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Comparison of expression deconvolution approaches</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org63a4104">Methods</a>
<ul>
<li><a href="#deconvolution">Distribution deconvolution</a></li>
<li><a href="#orgc6b0597">Benchmarking metric</a></li>
<li><a href="#data">Benchmarking data</a></li>
</ul>
</li>
<li><a href="#orga70e5f8">Results</a>
<ul>
<li><a href="#examples">Example</a></li>
<li><a href="#spike-ins">Spike-in data</a></li>
<li><a href="#homogeneous">Homogeneous cell populations</a></li>
<li><a href="#synthetic">Synthetic cell mixtures</a></li>
<li><a href="#heterogeneous">Heterogeneous cell populations</a></li>
<li><a href="#org99de990">Finalize plots</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orgbd3415a" class="outline-2">
<h2 id="introduction"><a id="orgbd3415a"></a>Introduction</h2>
<div class="outline-text-2" id="text-introduction">
<p>
Suppose we have observations \(x_i \sim f(\theta_i), i = 1, \ldots, n\), and
\(\theta_i \sim g(\cdot)\). <i>Distribution deconvolution</i> is the problem of
estimating \(g \in \mathcal{G}\) from \(x_1, \ldots, x_n\), assuming \(f\) is known
(<a href="https://academic.oup.com/biomet/article/103/1/1/2390141">Efron
2016</a>).
</p>

<p>
Recent work suggests that scRNA-seq data follows this generative model
(<a href="http://dx.doi.org/10.1073/pnas.1721085115">Wang et al. 2018</a>). Here, we
investigate the trade-off between model complexity/flexibility and
generalization for different choices of \(\mathcal{G}\) in real data.
</p>
</div>
</div>

<div id="outline-container-org8d85f12" class="outline-2">
<h2 id="setup"><a id="org8d85f12"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython" id="orgcb42039"><span class="org-keyword">import</span> multiprocessing <span class="org-keyword">as</span> mp
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scanpy
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scmodes
<span class="org-keyword">import</span> sklearn.model_selection <span class="org-keyword">as</span> skms

<span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
<span class="org-keyword">import</span> rpy2.robjects.numpy2ri

rpy2.robjects.pandas2ri.activate()
rpy2.robjects.numpy2ri.activate()

<span class="org-variable-name">ashr</span> = rpy2.robjects.packages.importr(<span class="org-string">'ashr'</span>)
<span class="org-variable-name">descend</span> = rpy2.robjects.packages.importr(<span class="org-string">'descend'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org63a4104" class="outline-2">
<h2 id="org63a4104">Methods</h2>
<div class="outline-text-2" id="text-org63a4104">
</div>
<div id="outline-container-org67b8f78" class="outline-3">
<h3 id="deconvolution"><a id="org67b8f78"></a>Distribution deconvolution</h3>
<div class="outline-text-3" id="text-deconvolution">
<p>
The general form of distribution deconvolution for scRNA-seq is:
</p>

<p>
\[ x_{ij} \sim \mathrm{Poisson}(\exp(\mathbf{z}_i' \mathbf{b}_j) \lambda_{ij}) \]
</p>

<p>
\[ \lambda_{ij} \sim g_j(\cdot) \]
</p>

<p>
where:
</p>

<ul class="org-ul">
<li>\(x_{ij}\) is the count of molecules of gene \(j\) in cell \(i\)</li>
<li>\(\mathbf{z}_i\) is a \(q\)-vector of covariates for cell \(i\)</li>
<li>\(\mathbf{b}_j\) is a \(q\)-vector of confounding effects on gene \(j\)</li>
<li>\(\lambda_{ij}\) is proportional to the relative abundance of gene \(j\)
in cell \(i\)</li>
</ul>

<p>
The primary inference goal is to recover \(g_j\). A secondary goal could be
to recover \(\lambda_{ij}\). We can trade off flexibility and complexity of
\(g_j\) for ease of implementation and speed.
</p>

<ol class="org-ol">
<li><b>Point mass:</b> \(g_j = \delta_\mu\). We mention it for completeness.</li>
<li><b>Gamma:</b> \(g_j = \mathrm{Gamma}(\cdot)\). This leads to the negative
binomial marginal likelihood, and can be motivated by the empirical
observation that the counts are overdispersed.</li>
<li><b>Point-Gamma:</b> \(g_j = \pi_j \delta_0(\cdot) + (1 - \pi_j)
      \mathrm{Gamma}(\cdot)\). This leads to the zero-inflated negative
binomial marginal likelihood, which is still analytic and therefore
computationally favorable. The inclusion of the point mass can be
motivated by theory suggesting a biological mechanism for bimodal gene
expression (<a href="http://dx.doi.org/10.1126/science.1216379">Munsky et
al. 2013</a>, <a href="http://dx.doi.org/10.1186/gb-2013-14-1-r7">Kim and Marioni
2013</a>).</li>
<li><b>Unimodal:</b> \(g_j\) is some unimodal distribution over non-negative
reals. In practice, we represent this family of distribution as \(g_j =
      \sum_k \pi_k \mathrm{Uniform}(\cdot; \lambda_0, a_{jk})\), where \(k = 1,
      \ldots, K\) are sufficiently many and \(\lambda_0\) is the mode
(<a href="http://dx.doi.org/10.1093/biostatistics/kxw041">Stephens 2016</a>).</li>
<li><b>Zero-inflated exponential family:</b> \(g_j = \exp(\mathbf{Q}\alpha -
      \phi(\alpha))\), where \(\mathbf{Q}\) is a
<a href="https://en.wikipedia.org/wiki/B-spline">B spline spline basis matrix</a>
for a <a href="https://en.wikipedia.org/wiki/Cubic_Hermite_spline">natural cubic
spline</a>
(<a href="https://www.rdocumentation.org/packages/splines/topics/ns"><code>ns</code>
function</a>;
<a href="https://academic.oup.com/biomet/article/103/1/1/2390141">Efron
2016</a>). The key idea of the method is use spline regression to find the
sufficient statistic and natural parameters which maximizes the penalized
likelihood of the observed data. The method has been extended to include
a point mass on zero (<a href="http://dx.doi.org/10.1073/pnas.1721085115">Wang
et al. 2018</a>).</li>
<li><b>Nonparametric:</b> \(g_j\) is some distribution over non-negative reals
(<a href="https://projecteuclid.org/euclid.aoms/1177728066">Kiefer and Wolfowitz
1956</a>). In practice, we discretize the representation
(<a href="https://amstat.tandfonline.com/doi/abs/10.1080/01621459.2013.869224">Koenker
and Mizera 2014</a>). In full detail, we use the representation \(g_j =
      \sum_k \pi_k \mathrm{Uniform}(\cdot; ak, a(k + 1))\), where \(a\) is a
fixed step size, which allows us to re-use the <code>ash</code> implementation.</li>
</ol>
</div>
</div>

<div id="outline-container-orgc6b0597" class="outline-3">
<h3 id="orgc6b0597">Benchmarking metric</h3>
<div class="outline-text-3" id="text-orgc6b0597">
<p>
In order to evaluate assumed families on their ability to explain the data,
we estimate the difference KL divergence between the fitted distribution and
the truth. The idea is:
</p>

<p>
\[ \hat{p} = \arg\max_{q} \frac{1}{n} \sum_i \ln q(x_i) \]
</p>

<p>
\[ \approx \arg\max_{q} \mathbb{E}_{x\sim p}[\ln q(x)] \]
</p>

<p>
\[ = \arg\max_{q} \mathbb{E}_{x\sim p}[\ln q(x)] - \mathbb{E}_{x\sim p}[\ln p(x)] \]
</p>

<p>
\[ = \arg\min_q \mathcal{KL}(p \Vert q) \]
</p>

<p>
We estimate \(\hat{p}\) on a training set, and then estimate
\(\mathbb{E}_{x\sim p}[\ln \hat{p}(x)]\) using an independent validation
set. By comparing against a baseline (Gamma), we don't need to estimate
\(\mathbb{E}_{x\sim p}[\ln p(x)]\).
</p>

<p>
The benchmarking code is implemented in the Python package
<a href="https://github.com/aksarkar/scmodes">scmodes</a>.
</p>
</div>
</div>

<div id="outline-container-org490227f" class="outline-3">
<h3 id="data"><a id="org490227f"></a>Benchmarking data</h3>
<div class="outline-text-3" id="text-data">
<p>
We deconvolve ERCC spike-in data generated on multiple platforms,
pre-processed as in
<a href="https://www.biorxiv.org/content/10.1101/582064v1">Svensson 2019</a>.
</p>

<ol class="org-ol">
<li>Drop-seq (Macosko et al., 2015)</li>
<li>InDrops (Klein et al., 2015)</li>
<li>10x Genomics, v1 chemistry (Zheng et al., 2017);</li>
<li>10x Genomics, ??? (Svensson et al., 2017).</li>
</ol>

<p>
We additionally deconvolve spike-in data included in iPSCs derived from 53
donor individuals. We
<a href="https://jdblischak.github.io/singlecell-qtl/descriptive-stats.html">previously
observed</a> individual-specific heterogeneity in this data.
</p>

<p>
As examples of homogeneous cell populations, we use:
</p>

<ol class="org-ol">
<li>Sorted immune cells sequenced on the 10X platform
(<a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>)</li>

<li>iPSCs derived from Yoruba LCLs, sequenced on the Fluidigm C1 platform
(<a href="http://dx.doi.org/10.1371/journal.pgen.1008045">Sarkar et al. 2018</a>)</li>
</ol>

<p>
We use the homogeneous cells populations to generate synthetic mixtures with
known cell type labels.
</p>

<p>
As examples of heterogeneous cell populations, we use:
</p>

<ol class="org-ol">
<li>Fresh PBMCs sequenced on the 10X platform
(<a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>)</li>

<li>Mouse neuron cells sequenced on the Fluidigm C1 platform
(<a href="https://science.sciencemag.org/content/347/6226/1138.full">Zeisel et
al. 2015</a>)</li>
</ol>
</div>
</div>
</div>

<div id="outline-container-orga70e5f8" class="outline-2">
<h2 id="orga70e5f8">Results</h2>
<div class="outline-text-2" id="text-orga70e5f8">
</div>
<div id="outline-container-org0629aa5" class="outline-3">
<h3 id="examples"><a id="org0629aa5"></a>Example</h3>
<div class="outline-text-3" id="text-examples">
<p>
As an example, use the highest expressed genes in 10K sorted CD8+ cytotoxic
T cells <a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19'</span>)
<span class="org-variable-name">xj</span> = pd.Series(x[:,x.mean(axis=0).argmax()])
<span class="org-variable-name">size_factor</span> = pd.Series(x.<span class="org-builtin">sum</span>(axis=1))
<span class="org-variable-name">lam</span> = xj / size_factor
</pre>
</div>

<p>
Fit unimodal distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">unimix_res</span> = ashr.ash_workhorse(
  pd.Series(np.zeros(x.shape[0])),
  1,
  lik=ashr.lik_pois(y=xj, scale=size_factor, link=<span class="org-string">'identity'</span>),
  outputlevel=<span class="org-string">'fitted_g'</span>,
  mixsd=pd.Series(np.geomspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 25)),
  mode=pd.Series([lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>()]))
<span class="org-variable-name">unimix_cdf</span> = ashr.cdf_ash(unimix_res, np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000))
</pre>
</div>

<p>
Fit NPMLE.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">K</span> = 200
<span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), K + 1)
<span class="org-variable-name">npmle_res</span> = ashr.ash_workhorse(
  pd.Series(np.zeros(x.shape[0])),
  1,
  outputlevel=<span class="org-string">'fitted_g'</span>,
  lik=ashr.lik_pois(y=xj, scale=size_factor, link=<span class="org-string">'identity'</span>),
  g=ashr.unimix(pd.Series(np.ones(K) / K), pd.Series(grid[:-1]), pd.Series(grid[1:])))
<span class="org-variable-name">npmle_cdf</span> = ashr.cdf_ash(npmle_res, np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000))
</pre>
</div>

<p>
Fit DESCEND.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">descend_res</span> = descend.deconvSingle(xj, scaling_consts=size_factor, verbose=<span class="org-constant">False</span>)2
</pre>
</div>

<p>
Fit NB/ZINB.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.io
<span class="org-keyword">import</span> scmodes
<span class="org-keyword">import</span> scqtl

<span class="org-variable-name">x</span> = scmodes.read10x(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>)
<span class="org-variable-name">size_factor</span> = x.<span class="org-builtin">sum</span>(axis=1).reshape(-1, 1)
<span class="org-variable-name">onehot</span> = np.ones((x.shape[0], 1))
<span class="org-variable-name">design</span> = np.zeros((x.shape[0], 1))
<span class="org-variable-name">init</span> = scqtl.tf.fit(
  umi=x.astype(np.float32),
  onehot=onehot.astype(np.float32),
  design=design.astype(np.float32),
  size_factor=size_factor.astype(np.float32),
  learning_rate=1e-3,
  max_epochs=30000,
  verbose=<span class="org-constant">True</span>)
pd.DataFrame(init[0]).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-nb-log-mu.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
pd.DataFrame(init[1]).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-nb-log-phi.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">log_mu</span>, <span class="org-variable-name">log_phi</span>, <span class="org-variable-name">logodds</span>, <span class="org-variable-name">nb_llik</span>, <span class="org-variable-name">zinb_llik</span> = scqtl.tf.fit(
  umi=x.astype(np.float32),
  onehot=onehot.astype(np.float32),
  design=design.astype(np.float32),
  size_factor=size_factor.astype(np.float32),
  learning_rate=1e-3,
  max_epochs=30000,
  warm_start=init[:3],
  verbose=<span class="org-constant">True</span>)
pd.DataFrame(log_mu).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-log-mu.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
pd.DataFrame(log_phi).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-log-phi.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
pd.DataFrame(logodds).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-logodds.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --time=60:00 --job-name=fit-nb
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python /project2/mstephens/aksarkar/projects/singlecell-modes/code/fit-nb.py
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">j</span> = <span class="org-builtin">str</span>(x.mean(axis=0).argmax())
<span class="org-variable-name">nb_log_mu</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-nb-log-mu.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">nb_log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-nb-log-phi.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-comment-delimiter"># </span><span class="org-comment">Gamma (Use MATLAB and MATHEMATICA (b=theta=scale, a=alpha=shape) definition)</span>
<span class="org-comment-delimiter"># </span><span class="org-comment">https://github.com/scipy/scipy/blob/v1.2.1/scipy/stats/_continuous_distns.py#L2479</span>
<span class="org-variable-name">gamma_cdf</span> = st.gamma(a=np.exp(-nb_log_phi[j]), scale=np.exp(nb_log_mu[j] + nb_log_phi[j])).cdf(np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">zinb_log_mu</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-log-mu.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">zinb_log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-log-phi.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">zinb_logodds</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-example/zheng-cd8-zinb-logodds.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">point_gamma_cdf</span> = st.gamma(a=np.exp(-zinb_log_phi[j]), scale=np.exp(zinb_log_mu[j] + zinb_log_phi[j])).cdf(np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000))
<span class="org-variable-name">point_gamma_cdf</span> *= sp.expit(-zinb_logodds[j].values)
<span class="org-variable-name">point_gamma_cdf</span> += sp.expit(zinb_logodds[j].values)
</pre>
</div>

<p>
Plot the observed counts and deconvolved distributions.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>).colors
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
<span class="org-variable-name">h</span> = ax[0].hist(xj, bins=np.arange(xj.<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_xlabel(<span class="org-string">'Molecule count'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)

ax[1].plot(np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000), gamma_cdf, color=cm[0], lw=1, label=<span class="org-string">'Gamma'</span>)
ax[1].plot(np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000), point_gamma_cdf, color=cm[1], lw=1, label=<span class="org-string">'Point-Gamma'</span>)
ax[1].plot(np.array(unimix_cdf.rx2(<span class="org-string">'x'</span>)),
           np.array(unimix_cdf.rx2(<span class="org-string">'y'</span>)).ravel(), c=cm[2], lw=1, label=<span class="org-string">'Unimodal'</span>)
<span class="org-variable-name">F</span> = np.cumsum(np.array(descend_res.slots[<span class="org-string">'density.points'</span>])[:,1])
ax[1].plot(np.array(descend_res.slots[<span class="org-string">'density.points'</span>])[:,0],
           F / F.<span class="org-builtin">max</span>(), c=cm[3], lw=1, label=<span class="org-string">'ZIEF'</span>)
ax[1].plot(np.array(npmle_cdf.rx2(<span class="org-string">'x'</span>)),
           np.array(npmle_cdf.rx2(<span class="org-string">'y'</span>)).ravel(), c=cm[4], lw=1, label=<span class="org-string">'NPMLE'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

ax[1].legend(frameon=<span class="org-constant">False</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/deconv-example.png" alt="deconv-example.png">
</p>
</div>

<p>
Print out the histogram for inspection.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.vstack((h[1][:-1], h[0])).T
</pre>
</div>

<pre class="example">
array([[  0.,   0.],
[  1.,   1.],
[  2.,   1.],
[  3.,   0.],
[  4.,   0.],
[  5.,   0.],
[  6.,   1.],
[  7.,   0.],
[  8.,   1.],
[  9.,   5.],
[ 10.,   8.],
[ 11.,   4.],
[ 12.,   9.],
[ 13.,   9.],
[ 14.,   8.],
[ 15.,  20.],
[ 16.,  17.],
[ 17.,  24.],
[ 18.,  31.],
[ 19.,  44.],
[ 20.,  39.],
[ 21.,  41.],
[ 22.,  46.],
[ 23.,  72.],
[ 24.,  68.],
[ 25.,  74.],
[ 26.,  74.],
[ 27.,  86.],
[ 28.,  94.],
[ 29., 114.],
[ 30.,  96.],
[ 31., 102.],
[ 32., 120.],
[ 33., 132.],
[ 34., 161.],
[ 35., 138.],
[ 36., 159.],
[ 37., 197.],
[ 38., 186.],
[ 39., 210.],
[ 40., 232.],
[ 41., 244.],
[ 42., 284.],
[ 43., 309.],
[ 44., 321.],
[ 45., 332.],
[ 46., 342.],
[ 47., 371.],
[ 48., 349.],
[ 49., 356.],
[ 50., 317.],
[ 51., 343.],
[ 52., 347.],
[ 53., 342.],
[ 54., 308.],
[ 55., 294.],
[ 56., 274.],
[ 57., 272.],
[ 58., 241.],
[ 59., 219.],
[ 60., 191.],
[ 61., 164.],
[ 62., 145.],
[ 63., 144.],
[ 64., 120.],
[ 65., 121.],
[ 66., 100.],
[ 67.,  95.],
[ 68.,  69.],
[ 69.,  62.],
[ 70.,  53.],
[ 71.,  47.],
[ 72.,  52.],
[ 73.,  32.],
[ 74.,  35.],
[ 75.,  36.],
[ 76.,  30.],
[ 77.,  28.],
[ 78.,  20.],
[ 79.,  27.],
[ 80.,  26.],
[ 81.,  21.],
[ 82.,  12.],
[ 83.,  12.],
[ 84.,   9.],
[ 85.,   7.],
[ 86.,  12.],
[ 87.,   8.],
[ 88.,   9.],
[ 89.,   6.],
[ 90.,   4.],
[ 91.,   5.],
[ 92.,   2.],
[ 93.,   2.],
[ 94.,   2.],
[ 95.,   3.],
[ 96.,   1.],
[ 97.,   2.],
[ 98.,   2.],
[ 99.,   0.],
[100.,   0.],
[101.,   0.],
[102.,   1.],
[103.,   1.],
[104.,   0.],
[105.,   0.],
[106.,   2.]])
</pre>

<p>
Look at what is going on with NPMLE.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res0</span> = ashr.ash_workhorse(
  pd.Series(np.zeros(x.shape[0])),
  1,
  outputlevel=pd.Series([<span class="org-string">'fitted_g'</span>, <span class="org-string">'loglik'</span>]),
  lik=ashr.lik_pois(y=xj, scale=size_factor, link=<span class="org-string">'identity'</span>),
  g=ashr.unimix(pd.Series(np.ones(100) / 100), pd.Series(grid[:-1]), pd.Series(grid[1:])))
<span class="org-variable-name">res1</span> = ashr.ash_workhorse(
  pd.Series(np.zeros(x.shape[0])),
  1,
  outputlevel=pd.Series([<span class="org-string">'fitted_g'</span>, <span class="org-string">'loglik'</span>]),
  lik=ashr.lik_pois(y=xj, scale=size_factor, link=<span class="org-string">'identity'</span>),
  g=ashr.unimix(pd.Series(np.ones(200) / 200), pd.Series(grid[:-1]), pd.Series(grid[1:])))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(4, 2)
<span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000)
plt.plot(grid, np.array(ashr.cdf_ash(res0, grid)[1]).ravel(), label=<span class="org-string">'NPMLE (100 components)'</span>, lw=1, c=<span class="org-string">'k'</span>)
plt.plot(grid, np.array(ashr.cdf_ash(res1, grid)[1]).ravel(), label=<span class="org-string">'NPMLE (200 components)'</span>, lw=2, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.legend(frameon=<span class="org-constant">False</span>)
plt.xlabel(<span class="org-string">'Latent gene expression'</span>)
plt.ylabel(<span class="org-string">'CDF'</span>)
</pre>
</div>

<pre class="example">
Text(0, 0.5, 'CDF')

</pre>

<div class="figure">
<p><img src="figure/deconvolution.org/npmle-example.png" alt="npmle-example.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">np.array(res0.rx2(<span class="org-string">'loglik'</span>)) - np.array(res1.rx2(<span class="org-string">'loglik'</span>))
</pre>
</div>

<pre class="example">
array([0.])

</pre>
</div>
</div>

<div id="outline-container-orgf1c6023" class="outline-3">
<h3 id="spike-ins"><a id="orgf1c6023"></a>Spike-in data</h3>
<div class="outline-text-3" id="text-spike-ins">
<p>
<a href="https://www.biorxiv.org/content/10.1101/582064v1">Svensson 2019</a> analyze
negative control data in which only spike-in data is sequenced. 
</p>

<p>
He claims that there should be no heterogeneity between samples in
spike-ins. However, there is random sampling of molecules in the spike-in
protocol, which will introduce noise. This was previously established by
<a href="https://www.pnas.org/content/115/28/E6437/">Wang et al. 2018</a>.
</p>

<p>
Further, <a href="http://dx.doi.org/10.1038/srep39921">Tung et al. 2016</a> establish
that in spike-in data added to real cells, there is considerable
heterogeneity in efficiency not just between cells, but also donor
individuals.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orge5c16cf"><span class="org-keyword">def</span> <span class="org-function-name">read_chromium</span>(sample):
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/svensson_chromium_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-keyword">return</span> x[x.obs[<span class="org-string">'sample'</span>] == sample][:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index].X.A

<span class="org-keyword">def</span> <span class="org-function-name">read_dropseq</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/macosko_dropseq_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-keyword">return</span> x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index].X.A

<span class="org-keyword">def</span> <span class="org-function-name">read_indrops</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/klein_indrops_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-keyword">return</span> x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index].X.A

<span class="org-keyword">def</span> <span class="org-function-name">read_gemcode</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/zheng_gemcode_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-keyword">return</span> x.X.A

<span class="org-keyword">def</span> <span class="org-function-name">read_c1</span>():
  <span class="org-variable-name">x</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/ercc-counts.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
  <span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, header=<span class="org-constant">None</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Throw out samples with too few spike-in molecule detected</span>
  <span class="org-keyword">return</span> x.loc[:,np.logical_and((x.<span class="org-builtin">sum</span>(axis=0) &gt; 1000).values, keep_samples.values.ravel())].values.T

<span class="org-variable-name">data</span> = {
  <span class="org-string">'dropseq'</span>: read_dropseq,
  <span class="org-string">'indrops'</span>: read_indrops,
  <span class="org-string">'chromium1'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20311'</span>),
  <span class="org-string">'chromium2'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20312'</span>),
  <span class="org-string">'gemcode'</span>: read_gemcode,
  <span class="org-string">'c1'</span>: read_c1,
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgab7cb06">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
&lt;&lt;ercc-data&gt;&gt;
<span class="org-variable-name">cpu_methods</span> = [<span class="org-string">'unimodal'</span>, <span class="org-string">'zief'</span>, <span class="org-string">'npmle'</span>]
<span class="org-variable-name">tasks</span> = [(d, c) <span class="org-keyword">for</span> d <span class="org-keyword">in</span> (<span class="org-string">'c1'</span>,) <span class="org-keyword">for</span> c <span class="org-keyword">in</span> cpu_methods]
<span class="org-variable-name">dataset</span>, <span class="org-variable-name">method</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-keyword">with</span> mp.Pool() <span class="org-keyword">as</span> pool:
  <span class="org-variable-name">x</span> = data[dataset]()
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x, pool=pool, test_size=0.1, methods=[method])
  res.to_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{dataset}-{method}.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch -a 0-2 --partition=broadwl -n1 -c28 --exclusive --time=6:00:00 --mem=16G --job-name=deconv-generalization --out=benchmark-cpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;run-ercc-benchmark-cpu&gt;&gt;</span>
EOF
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org19278c9">&lt;&lt;imports&gt;&gt;
&lt;&lt;ercc-data&gt;&gt;
<span class="org-keyword">for</span> dataset <span class="org-keyword">in</span> (<span class="org-string">'c1'</span>,):
  <span class="org-variable-name">x</span> = data[dataset]()
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x, pool=<span class="org-constant">None</span>, test_size=0.1, methods=[<span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>])
  res.to_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{dataset}-gpu.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=4G --time=60:00 --job-name=deconv-generalization --out=benchmark-gpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;run-ercc-benchmark-gpu&gt;&gt;</span>
EOF
</pre>
</div>

<p>
Process the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sample_sizes</span> = {k: <span class="org-builtin">int</span>(.1 * data[k]().shape[0]) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {}
<span class="org-keyword">for</span> data <span class="org-keyword">in</span> (<span class="org-string">'dropseq'</span>, <span class="org-string">'indrops'</span>, <span class="org-string">'chromium1'</span>, <span class="org-string">'chromium2'</span>, <span class="org-string">'gemcode'</span>):
  <span class="org-variable-name">benchmark</span>[data] = (
    pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-unimodal.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-zief.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-npmle.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)) / sample_sizes[data]
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.concat(benchmark).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/control-kl.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/control-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))}
</pre>
</div>

<p>
Plot the generalization performance of each method.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org7f68c9b"><span class="org-keyword">def</span> <span class="org-function-name">plot_benchmark</span>(benchmark, titles=<span class="org-constant">None</span>, xlabel_index=<span class="org-constant">None</span>, size=<span class="org-constant">None</span>):
  <span class="org-keyword">if</span> size <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">size</span> = (7, 3)
  <span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, <span class="org-builtin">len</span>(benchmark), sharey=<span class="org-constant">True</span>)
  fig.set_size_inches(*size)
  <span class="org-keyword">if</span> titles <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">titles</span> = <span class="org-builtin">sorted</span>(benchmark.keys())
  <span class="org-keyword">for</span> a, k, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), <span class="org-builtin">sorted</span>(benchmark.keys()), titles):
    <span class="org-variable-name">T</span> = (benchmark[f<span class="org-string">'{k}'</span>].values - benchmark[f<span class="org-string">'{k}'</span>][<span class="org-string">'gamma'</span>].values.reshape(-1, 1))[:,1:]
    a.boxplot([x[~np.isnan(x)] <span class="org-keyword">for</span> x <span class="org-keyword">in</span> T.T], widths=.5, medianprops={<span class="org-string">'color'</span>: <span class="org-string">'k'</span>}, flierprops={<span class="org-string">'marker'</span>: <span class="org-string">'.'</span>, <span class="org-string">'markersize'</span>: 4})
    a.axhline(y=0, ls=<span class="org-string">':'</span>, lw=1, c=<span class="org-string">'k'</span>)
    a.set_xticklabels([<span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>], rotation=90)
    a.set_title(t)
  <span class="org-keyword">if</span> xlabel_index <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">xlabel_index</span> = np.arange(ax.shape[0])
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">xlabel_index</span> = np.atleast_1d(xlabel_index)
  <span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax[xlabel_index]:
    a.set_xlabel(<span class="org-string">'Assumed latent dist'</span>)
  ax[0].set_ylabel(<span class="org-string">'Improvement $KL(p \Vert \hat{p})$\nover Gamma'</span>)
  fig.tight_layout()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_benchmark(benchmark, 2)
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/control-benchmark.png" alt="control-benchmark.png">
</p>
</div>

<p>
<b>Conclusions:</b>
</p>

<ol class="org-ol">
<li>ZIG does not perform better than Gamma, supporting the conclusions of
<a href="https://www.biorxiv.org/content/10.1101/582064v1">Svensson 2019</a></li>
<li>There is heterogeneity which isn't adequately described by Gamma, but
<a href="deconv-examples.html#org3205fdb">appears equally well
described by unimodal/NPMLE</a></li>
</ol>
</div>
</div>

<div id="outline-container-org3e95935" class="outline-3">
<h3 id="homogeneous"><a id="org3e95935"></a>Homogeneous cell populations</h3>
<div class="outline-text-3" id="text-homogeneous">
<p>
Run the benchmark.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgd89a606"><span class="org-variable-name">data</span> = {
  <span class="org-string">'cytotoxic_t'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'b_cells'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/b_cells/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'ipsc'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>, n=100, return_df=<span class="org-constant">True</span>),
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgeaeae79"><span class="org-keyword">import</span> os
&lt;&lt;homogeneous-data&gt;&gt;
<span class="org-variable-name">cpu_methods</span> = [<span class="org-string">'unimodal'</span>, <span class="org-string">'zief'</span>, <span class="org-string">'npmle'</span>]
<span class="org-variable-name">tasks</span> = [(d, c) <span class="org-keyword">for</span> d <span class="org-keyword">in</span> data <span class="org-keyword">for</span> c <span class="org-keyword">in</span> cpu_methods]
<span class="org-variable-name">task</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-keyword">with</span> mp.Pool(maxtasksperchild=20) <span class="org-keyword">as</span> pool:
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: this needs to be done after initializing the pool to avoid</span>
  <span class="org-comment-delimiter"># </span><span class="org-comment">memory duplication</span>
  <span class="org-variable-name">x</span> = data[task[0]]()
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, pool=pool, test_size=0.1, random_state=0, methods=[task[1]])
  <span class="org-variable-name">res.index</span> = x.columns
  res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/{task[0]}-{task[1]}.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --mem=32G -a 0-8 -n1 -c28 --exclusive --time=12:00:00 --job-name=benchmark --output=benchmark-cpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
&lt;&lt;run-homogeneous-benchmark-cpu&gt;&gt;
EOF
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org11112f9"><span class="org-keyword">import</span> os
&lt;&lt;homogeneous-data&gt;&gt;
<span class="org-variable-name">tasks</span> = <span class="org-builtin">list</span>(data.keys())
<span class="org-variable-name">task</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-variable-name">x</span> = data[task]()
<span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, pool=<span class="org-constant">None</span>, test_size=0.1, random_state=0, methods=[<span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>])
<span class="org-variable-name">res.index</span> = x.columns
res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/{task}-gpu.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 -a 1-2 --mem=16G --time=12:00:00 --job-name=benchmark --output=benchmark-gpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
module load cuda/9.0
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
&lt;&lt;run-homogeneous-benchmark-gpu&gt;&gt;
EOF
</pre>
</div>

<p>
Process the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sample_sizes</span> = {k: <span class="org-builtin">int</span>(.1 * data[k]().shape[0]) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {}
<span class="org-keyword">for</span> data <span class="org-keyword">in</span> (<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, <span class="org-string">'ipsc'</span>):
  <span class="org-variable-name">benchmark</span>[data] = (
    pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-unimodal.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-zief.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-npmle.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)) / sample_sizes[data]
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.concat(benchmark).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/homogeneous-kl.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/homogeneous-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))}

</pre>
</div>

<p>
Plot the generalization performance of each method.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plot_benchmark(benchmark, xlabel_index=1, titles=[<span class="org-string">'B cells'</span>, <span class="org-string">'Cytotoxic T cells'</span>, <span class="org-string">'iPSCs'</span>], size=(5, 3))
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/homogeneous-benchmark.png" alt="homogeneous-benchmark.png">
</p>
</div>

<p>
<b>Conclusions:</b>
</p>

<ul class="org-ul">
<li>The choice of unimodal discretization matters both for generalization
performance and computational speed. The default in <code>ashr</code> is flawed;
<a href="https://github.com/aksarkar/scmodes/blob/4d84236888daba32a7fee60642f3addd5a34ff9c/scmodes/benchmark/deconvolution.py#L83">our
choice</a> makes the method considerably slower</li>
<li>In the iPSC data, modeling all samples without considering individual of
origin is clearly flawed</li>
</ul>
</div>
</div>

<div id="outline-container-org76afb45" class="outline-3">
<h3 id="synthetic"><a id="org76afb45"></a>Synthetic cell mixtures</h3>
<div class="outline-text-3" id="text-synthetic">
</div>
<div id="outline-container-orgd2510e6" class="outline-4">
<h4 id="orgd2510e6">T cell-B cell mixture</h4>
<div class="outline-text-4" id="text-orgd2510e6">
<p>
Create a synthetic heterogeneous population of cells by combining sorted
CD8+ T cells and CD19+ B cells from
<a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org897e7b6"><span class="org-variable-name">t_cells</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">b_cells</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/b_cells/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(t_cells, b_cells)
<span class="org-comment-delimiter"># </span><span class="org-comment">Important: filter genes after mixing cells</span>
<span class="org-variable-name">keep_genes</span> = (x &gt; 0).mean(axis=0) &gt;= 0.25
<span class="org-variable-name">x</span> = x.loc[:,keep_genes]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org1ded154">      &lt;&lt;imports&gt;&gt;
      &lt;&lt;cd8-cd19-mix&gt;&gt;
      <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, stratify=y, pool=<span class="org-constant">None</span>, train_size=0.5, test_size=0.1, random_state=0, methods=[<span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>])
      res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/cd8-cd19-mix-gpu.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
(Figure \ref{fig:data-deconv}A)    <span class="org-comment-delimiter">#</span><span class="org-comment">+END_SRC</span>

    <span class="org-comment-delimiter">#</span><span class="org-comment">+BEGIN_SRC sh :noweb eval :dir /scratch/midway2/aksarkar/modes/</span>
      sbatch --<span class="org-variable-name">partition</span>=gpu2 --<span class="org-variable-name">gres</span>=gpu:1 --<span class="org-variable-name">mem</span>=16G --<span class="org-variable-name">time</span>=12:00:00 --job-<span class="org-variable-name">name</span>=benchmark --<span class="org-variable-name">output</span>=benchmark-gpu.out
      <span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
      module load cuda/9.0
      source activate scmodes
      python &lt;&lt;EOF
      &lt;&lt;run-cd8-cd19-mix-benchmark-gpu&gt;&gt;
      EOF
</pre>
</div>

<pre class="example">
Submitted batch job 60359981

</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="orgb9aa912">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
<span class="org-variable-name">tasks</span> = [<span class="org-string">'unimodal'</span>, <span class="org-string">'zief'</span>, <span class="org-string">'npmle'</span>]
<span class="org-variable-name">task</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-keyword">with</span> mp.Pool() <span class="org-keyword">as</span> pool:
  &lt;&lt;cd8-cd19-mix&gt;&gt;
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, stratify=y, pool=pool, train_size=0.5, test_size=0.1, random_state=0, methods=[task])
  res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/cd8-cd19-mix-{task}.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -a 0-2 -n1 -c28 --exclusive --mem=16G --time=12:00:00 --job-name=benchmark --output=benchmark-cpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
module load cuda/9.0
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;run-cd8-cd19-mix-benchmark-cpu&gt;&gt;</span>
EOF
</pre>
</div>
</div>
</div>

<div id="outline-container-org254b299" class="outline-4">
<h4 id="org254b299">Naive/activated T cell mixture</h4>
<div class="outline-text-4" id="text-org254b299">
<p>
Create a synthetic mixture of cells by combining sorted
CD8+ cytotoxic T cells and CD8+/CD45RA+ naive cytotoxic T cells
<a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgf397489"><span class="org-variable-name">cyto</span> = scmodes.dataset.read_10x(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">naive</span> = scmodes.dataset.read_10x(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/naive_cytotoxic/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(cyto, naive)
<span class="org-variable-name">keep_genes</span> = (x &gt; 0).mean(axis=0) &gt;= 0.25
<span class="org-variable-name">x</span> = x.loc[:,keep_genes]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=32G --time=1:00:00 --job-name=benchmark --output=benchmark-gpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
&lt;&lt;cyto-naive-mix&gt;&gt;
res = scmodes.benchmark.evaluate_deconv_generalization(x.values, <span class="org-variable-name">pool</span>=None, <span class="org-variable-name">stratify</span>=y, <span class="org-variable-name">train_size</span>=0.5, <span class="org-variable-name">test_size</span>=0.1, <span class="org-variable-name">random_state</span>=0, <span class="org-variable-name">methods</span>=[<span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>])
res.index = x.columns
<span class="org-function-name">res.to_csv</span>(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/cyto-naive-t-mix-gpu.txt.gz'</span>, <span class="org-variable-name">compression</span>=<span class="org-string">'gzip'</span>, <span class="org-variable-name">sep</span>=<span class="org-string">'\t'</span>)
EOF
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -a 0-2 -n1 -c28 --exclusive --mem=32G --time=24:00:00 --job-name=benchmark --output=benchmark-cpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
import os
tasks = [<span class="org-string">'unimodal'</span>, <span class="org-string">'zief'</span>, <span class="org-string">'npmle'</span>]
task = tasks[int(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
with mp.Pool(<span class="org-variable-name">maxtasksperchild</span>=10) as pool:
  &lt;&lt;cyto-naive-mix&gt;&gt;
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_deconv_generalization(x.values, pool=pool, stratify=y, train_size=0.5, test_size=0.1, random_state=0, methods=[task])</span>
<span class="org-sh-heredoc">  res.index = x.columns</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/deconv-generalization/cyto-naive-t-mix-{task}.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org1000070" class="outline-4">
<h4 id="org1000070">Plot the results</h4>
<div class="outline-text-4" id="text-org1000070">
<p>
Process the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sample_sizes</span> = {<span class="org-string">'cd8-cd19-mix'</span>: 2029,
                <span class="org-string">'cyto-naive-t-mix'</span>: 2216}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {}
<span class="org-keyword">for</span> data <span class="org-keyword">in</span> (<span class="org-string">'cd8-cd19-mix'</span>, <span class="org-string">'cyto-naive-t-mix'</span>):
  <span class="org-variable-name">benchmark</span>[data] = (
    pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-unimodal.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-zief.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-npmle.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)) / sample_sizes[data]
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.concat(benchmark).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/synthetic-mix-kl.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/synthetic-mix-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))}
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plot_benchmark(benchmark, titles=[<span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>], size=(4, 3))
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/synthetic-mix-benchmark.png" alt="synthetic-mix-benchmark.png">
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-org277d861" class="outline-3">
<h3 id="heterogeneous"><a id="org277d861"></a>Heterogeneous cell populations</h3>
<div class="outline-text-3" id="text-heterogeneous">
<div class="org-src-container">
<pre class="src src-ipython" id="org3d7c874"><span class="org-variable-name">data</span> = {
  <span class="org-string">'pbmcs_68k'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/fresh_68k_pbmc_donor_a/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'cortex'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.cortex(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/zeisel-2015/GSE60361_C1-3005-Expression.txt.gz'</span>, return_df=<span class="org-constant">True</span>)
}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org5f77663">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
&lt;&lt;heterogeneous-data&gt;&gt;
<span class="org-variable-name">tasks</span> = [<span class="org-string">'pbmcs_68k'</span>, <span class="org-string">'cortex'</span>]
<span class="org-variable-name">task</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-variable-name">x</span> = data[task]()
<span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, pool=<span class="org-constant">None</span>, test_size=0.1, random_state=0, methods=[<span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>])
<span class="org-variable-name">res.index</span> = x.columns
res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/{task}-gpu.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 -a 0-1 --gres=gpu:1 --mem=16G --time=12:00:00 --job-name=benchmark --output=benchmark-gpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;heterogeneous-benchmark-gpu&gt;&gt;</span>
EOF
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org6db4aa4">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
&lt;&lt;heterogeneous-data&gt;&gt;
<span class="org-variable-name">cpu_methods</span> = [<span class="org-string">'unimodal'</span>, <span class="org-string">'zief'</span>, <span class="org-string">'npmle'</span>]
<span class="org-variable-name">tasks</span> = [(d, c) <span class="org-keyword">for</span> d <span class="org-keyword">in</span> data <span class="org-keyword">for</span> c <span class="org-keyword">in</span> cpu_methods]
<span class="org-variable-name">dataset</span>, <span class="org-variable-name">method</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-keyword">with</span> mp.Pool() <span class="org-keyword">as</span> pool:
  <span class="org-variable-name">x</span> = data[dataset]()
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_deconv_generalization(x.values, pool=pool, train_size=0.5, test_size=0.1, random_state=0, methods=[method])
  <span class="org-variable-name">res.index</span> = x.columns
  res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/deconv-generalization/{dataset}-{method}.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -a 1 -n1 -c28 --exclusive --mem=32G --time=24:00:00 --job-name=benchmark --output=benchmark-cpu.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;heterogeneous-benchmark-cpu&gt;&gt;</span>
EOF
</pre>
</div>

<p>
Process the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sample_sizes</span> = {k: <span class="org-builtin">int</span>(.1 * data[k]().shape[0]) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {}
<span class="org-keyword">for</span> data <span class="org-keyword">in</span> (<span class="org-string">'cortex'</span>, <span class="org-string">'pbmcs_68k'</span>):
  <span class="org-variable-name">benchmark</span>[data] = (
    pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-unimodal.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-zief.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
    .merge(pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/{data}-npmle.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)) / sample_sizes[data]
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.concat(benchmark).to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/heterogeneous-kl.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/heterogeneous-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))}
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plot_benchmark(benchmark, titles=[<span class="org-string">'Mouse cortex'</span>, <span class="org-string">'Human PBMC'</span>], size=(4, 3))
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/heterogeneous-benchmark.png" alt="heterogeneous-benchmark.png">
</p>
</div>
</div>
</div>
<div id="outline-container-org99de990" class="outline-3">
<h3 id="org99de990">Finalize plots</h3>
<div class="outline-text-3" id="text-org99de990">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">benchmark</span> = {
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/homogeneous-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))}
benchmark.update({
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/synthetic-mix-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))})
benchmark.update({
  k: g.reset_index(level=0, drop=<span class="org-constant">True</span>) <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span>
  (pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/heterogeneous-kl.txt.gz'</span>,
               sep=<span class="org-string">'\t'</span>, index_col=[0, 1])
   .groupby(level=0))})
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">del</span> benchmark[<span class="org-string">'cortex'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 3, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">False</span>)
fig.set_size_inches(5, 4)
<span class="org-keyword">for</span> a, k, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), benchmark, [<span class="org-string">'B cell'</span>, <span class="org-string">'Cytotoxic T'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'B cell/T cell'</span>, <span class="org-string">'Naive/cytotoxic T'</span>, <span class="org-string">'PBMC'</span>]):
  <span class="org-variable-name">T</span> = (benchmark[f<span class="org-string">'{k}'</span>].values - benchmark[f<span class="org-string">'{k}'</span>][<span class="org-string">'gamma'</span>].values.reshape(-1, 1))[:,1:]
  a.boxplot([x[~np.isnan(x)] <span class="org-keyword">for</span> x <span class="org-keyword">in</span> T.T], widths=.5, medianprops={<span class="org-string">'color'</span>: <span class="org-string">'k'</span>}, flierprops={<span class="org-string">'marker'</span>: <span class="org-string">'.'</span>, <span class="org-string">'markersize'</span>: 4})
  a.axhline(y=0, ls=<span class="org-string">':'</span>, lw=1, c=<span class="org-string">'k'</span>)
  a.set_xticklabels([<span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>], rotation=90)
  a.set_title(t)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a[0].set_ylabel(<span class="org-string">'Improvement $KL(p \Vert \hat{p})$'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/mlcb-bench.png" alt="mlcb-bench.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">ipsc_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/ipsc-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">lrt</span> = st.chi2(1).sf(-2 * (ipsc_res[<span class="org-string">'gamma'</span>] - ipsc_res[<span class="org-string">'point_gamma'</span>]))
<span class="org-variable-name">ipsc_counts</span> = scmodes.dataset.ipsc(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>,
                                   query=<span class="org-builtin">list</span>(ipsc_res[lrt &lt; .05 / lrt.shape[0]].index), return_df=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">annotation</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, index_col=0, header=<span class="org-constant">None</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">annotation</span> = annotation.loc[keep_samples.values.ravel()]
<span class="org-variable-name">s</span> = annotation[<span class="org-string">'mol_hs'</span>]
<span class="org-variable-name">donor_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/20130606_sample_info.txt'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">ipsc_log_mu</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">ipsc_log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">ipsc_logodds</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-logodds.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">point_gamma_cdf</span>(grid, log_mu, log_phi, logodds):
  <span class="org-variable-name">res</span> = st.gamma(a=np.exp(-log_phi), scale=np.exp(log_mu + log_phi)).cdf(grid)
  <span class="org-variable-name">res</span> *= sp.expit(-logodds)
  <span class="org-variable-name">res</span> += sp.expit(logodds)
  <span class="org-keyword">return</span> res

<span class="org-variable-name">lam</span> = ipsc_counts.values / s.values.reshape(-1, 1)
<span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdfs</span> = np.array([[point_gamma_cdf(grid,
                                  ipsc_log_mu.loc[j, k],
                                  ipsc_log_phi.loc[j, k],
                                  ipsc_logodds.loc[j, k])
                  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> ipsc_log_mu]
                 <span class="org-keyword">for</span> j <span class="org-keyword">in</span> ipsc_counts.columns
                 <span class="org-keyword">if</span> j != <span class="org-builtin">list</span>(ipsc_log_mu.columns).index(<span class="org-string">'NA18498'</span>)])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
<span class="org-variable-name">colors</span> = {k: <span class="org-string">'r'</span> <span class="org-keyword">if</span> v == <span class="org-string">'male'</span> <span class="org-keyword">else</span> <span class="org-string">'b'</span> <span class="org-keyword">for</span> k, v <span class="org-keyword">in</span>
          donor_info.set_index(<span class="org-string">'Sample'</span>)
          .<span class="org-builtin">filter</span>(items=<span class="org-builtin">list</span>(ipsc_log_mu.columns), axis=0)
          [<span class="org-string">'Gender'</span>]
          .iteritems()}

</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">k</span> = <span class="org-string">'ENSG00000129824'</span>
<span class="org-variable-name">x</span> = ipsc_counts.loc[:,k].values
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4, 4)

ax[0].hist(x, bins=np.arange(x.<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_title(gene_info.loc[k, <span class="org-string">'name'</span>])
ax[0].set_xlabel(<span class="org-string">'Num mols'</span>)
ax[0].set_ylabel(<span class="org-string">'Num cells'</span>)

<span class="org-keyword">for</span> f, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(cdfs[1], ipsc_log_mu.columns):
  ax[1].plot(grid, f, lw=1, alpha=0.2, c=colors.get(k, <span class="org-string">'k'</span>))
ax[1].set_xlim(0, 2e-4)
ax[1].set_xticks(np.linspace(0, 2e-4, 3))
<span class="org-variable-name">dummy</span> = [plt.Line2D([0], [0], c=<span class="org-string">'r'</span>), plt.Line2D([0], [0], c=<span class="org-string">'b'</span>)]
ax[1].legend(dummy, [<span class="org-string">'Male'</span>, <span class="org-string">'Female'</span>], frameon=<span class="org-constant">False</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconvolution.org/mlcb-ex.png" alt="mlcb-ex.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-08-27 Tue 23:40</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
