<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-05-23 Thu 16:18 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Differential expression based on distribution deconvolution</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Differential expression based on distribution deconvolution</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgfb48832">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#orgf34f6e5">Methods</a>
<ul>
<li><a href="#orgc7a887d">Deconvolution-based test</a></li>
</ul>
</li>
<li><a href="#org2f0fa32">Results</a>
<ul>
<li><a href="#org2cbf224">Idealized simulation</a></li>
<li><a href="#org7b05d73">Realistic null simulation</a></li>
<li><a href="#orgb04ebae">Follow up</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orgfb48832" class="outline-2">
<h2 id="orgfb48832">Introduction</h2>
<div class="outline-text-2" id="text-orgfb48832">
<p>
For gene \(j\), consider the generative model for counts \(x_i, i=1, \ldots,
  n\):
</p>

<p>
\[ x_i \sim \mathrm{Pois}(s_i \lambda_i) \]
</p>

<p>
\[ \lambda_i \sim g_{z_i}(\cdot) \]
</p>

<p>
where \(z_i\) denotes group membership.
</p>

<p>
How do we use this model to test for differential expression?
</p>
</div>
</div>

<div id="outline-container-org5cc61b9" class="outline-2">
<h2 id="setup"><a id="org5cc61b9"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scmodes
<span class="org-keyword">import</span> scqtl.simple
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgf34f6e5" class="outline-2">
<h2 id="orgf34f6e5">Methods</h2>
<div class="outline-text-2" id="text-orgf34f6e5">
</div>
<div id="outline-container-orgc7a887d" class="outline-3">
<h3 id="orgc7a887d">Deconvolution-based test</h3>
<div class="outline-text-3" id="text-orgc7a887d">
<p>
In the simplest case, suppose \(g_k = \delta_{\mu_k}\). Then,
</p>

<p>
\[ \hat\mu_k = \frac{\sum_i [z_i = k] x_i}{\sum_i [z_i = k] s_i} \]
</p>

<p>
This suggests a simple likelihood ratio test, comparing the null model
</p>

<p>
\[ x_i \sim \mathrm{Pois}(s_i \mu_0) \]
</p>

<p>
against an alternative model:
</p>

<p>
\[ x_i \sim \mathrm{Pois}(s_i \mu_{z_i}) \]
</p>

<p>
More generally, we can compare the null model:
</p>

<p>
\[ x_i \sim \mathrm{Pois}(s_i \lambda_i) \]
</p>

<p>
\[ \lambda_i \sim g_0(\cdot) \]
</p>

<p>
against an alternative model:
</p>

<p>
\[ x_i \sim \mathrm{Pois}(s_i \lambda_i) \]
</p>

<p>
\[ \lambda_i \sim g_{z_i}(\cdot) \]
</p>

<p>
In the general case, it will not be true that \(-2 \ln(l_0 - l_1) \sim
   \chi^2_1\). However, for the case where \(g\) is assumed to be Gamma
distributed, it will be \(\chi^2_1\) or \(\chi^2_2\), depending on whether
we allow dispersions to vary across groups also.
</p>
</div>
</div>
</div>

<div id="outline-container-org2f0fa32" class="outline-2">
<h2 id="org2f0fa32">Results</h2>
<div class="outline-text-2" id="text-org2f0fa32">
</div>
<div id="outline-container-org2cbf224" class="outline-3">
<h3 id="org2cbf224">Idealized simulation</h3>
<div class="outline-text-3" id="text-org2cbf224">
<p>
Look at a case where
<a href="https://stephens999.github.io/misc/simple_transform_simulation.html">log-transform
fails</a>.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(0)
<span class="org-variable-name">N</span> = 100
<span class="org-variable-name">onehot</span> = np.zeros((2 * N, 2))
<span class="org-variable-name">onehot</span>[:N, 0] = 1
<span class="org-variable-name">onehot</span>[N:, 1] = 1
<span class="org-variable-name">s</span> = onehot.dot(np.array([1e5, 2e5]))
<span class="org-variable-name">mu</span> = 1e-5

<span class="org-variable-name">llrs</span> = []
<span class="org-variable-name">pvals</span> = []
<span class="org-keyword">for</span> trial <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1000):
  <span class="org-variable-name">x</span> = np.random.poisson(lam=s * mu)
  <span class="org-variable-name">mu0</span> = x.<span class="org-builtin">sum</span>() / s.<span class="org-builtin">sum</span>()
  <span class="org-variable-name">mu1</span> = x[:N].<span class="org-builtin">sum</span>() / s[:N].<span class="org-builtin">sum</span>()
  <span class="org-variable-name">mu2</span> = x[N:].<span class="org-builtin">sum</span>() / s[N:].<span class="org-builtin">sum</span>()
  <span class="org-variable-name">llr</span> = st.poisson(mu=mu0).logpmf(x).<span class="org-builtin">sum</span>() - st.poisson(mu=onehot.dot(np.array([mu1, mu2]))).logpmf(x).<span class="org-builtin">sum</span>()
  llrs.append(llr)
<span class="org-variable-name">llrs</span> = np.array(llrs)
</pre>
</div>

<p>
Look at the QQ plot.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(st.chi2(1).ppf(np.linspace(0, 1, llrs.shape[0])), np.sort(-2 * llrs), c=<span class="org-string">'k'</span>, s=2)
plt.plot([0, 11], [0, 11], c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">':'</span>)
plt.xlabel(<span class="org-string">'Expected chi-square'</span>)
plt.ylabel(<span class="org-string">'Observed chi-square'</span>)
</pre>
</div>

<pre class="example">
Text(0, 0.5, 'Observed chi-square')

</pre>

<div class="figure">
<p><img src="figure/deconv-de.org/idealized-qq.png" alt="idealized-qq.png">
</p>
</div>

<p>
Look at the histogram of p-values.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.hist(st.chi2(1).sf(-2 * llrs), np.linspace(0, 1, 11), density=<span class="org-constant">True</span>, color=<span class="org-string">'black'</span>)
plt.axhline(y=1, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.xlabel(<span class="org-string">'P-value'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconv-de.org/idealized.png" alt="idealized.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org7b05d73" class="outline-3">
<h3 id="org7b05d73">Realistic null simulation</h3>
<div class="outline-text-3" id="text-org7b05d73">
<p>
Take a real homogeneous data set (sorted cells from Zheng et al. 2017), and
randomly partition samples into two groups.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cd8</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>)
<span class="org-variable-name">s</span> = cd8.<span class="org-builtin">sum</span>(axis=1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(0)
<span class="org-variable-name">z</span> = np.random.uniform(size=cd8.shape[0]) &lt; 0.5
z.mean()
</pre>
</div>

<pre class="example">
0.5071015770398668

</pre>
</div>

<div id="outline-container-org31e6fce" class="outline-4">
<h4 id="org31e6fce">Point-mass deconvolution</h4>
<div class="outline-text-4" id="text-org31e6fce">
<p>
Deconvolve gene expression assuming \(g\) is a point mass (counts are
marginally Poisson distributed). Naively, this should work because the data
are informative about the mean.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">de</span>(x, s, z):
  <span class="org-doc">"""Return LRT p-value</span>

<span class="org-doc">  x - counts (n,)</span>
<span class="org-doc">  s - size factors (n,)</span>
<span class="org-doc">  z - boolean group assignment (n,)</span>

<span class="org-doc">  """</span>
  <span class="org-variable-name">mu0</span> = x.<span class="org-builtin">sum</span>() / s.<span class="org-builtin">sum</span>()
  <span class="org-variable-name">mu1</span> = x[z].<span class="org-builtin">sum</span>() / s[z].<span class="org-builtin">sum</span>()
  <span class="org-variable-name">mu2</span> = x[~z].<span class="org-builtin">sum</span>() / s[~z].<span class="org-builtin">sum</span>()
  <span class="org-variable-name">onehot</span> = pd.get_dummies(z)
  <span class="org-variable-name">llr</span> = st.poisson(mu=mu0).logpmf(x).<span class="org-builtin">sum</span>() - st.poisson(mu=onehot.dot(np.array([mu1, mu2]))).logpmf(x).<span class="org-builtin">sum</span>()
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'mu0'</span>: mu0, <span class="org-string">'mu1'</span>: mu1, <span class="org-string">'mu2'</span>: mu2, <span class="org-string">'llr'</span>: llr, <span class="org-string">'p'</span>: st.chi2(1).sf(-2 * llr)})
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">de_res</span> = cd8.<span class="org-builtin">apply</span>(de, axis=0, args=(s, z)).T
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.hist(de_res.loc[:,<span class="org-string">'p'</span>], np.linspace(0, 1, 11), density=<span class="org-constant">True</span>, color=<span class="org-string">'black'</span>)
plt.axhline(y=1, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.xlabel(<span class="org-string">'P-value'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconv-de.org/zheng-null.png" alt="zheng-null.png">
</p>
</div>

<p>
Report the proportion of genes for which the null model fit better than the
alternative model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">(de_res[<span class="org-string">'llr'</span>] &gt; 0).mean()
</pre>
</div>

<pre class="example">
0.6529284164859002

</pre>

<p>
Look at the gene where the null model did best against the alternative.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgeea08c4"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">k</span> = de_res[<span class="org-string">'llr'</span>].idxmax()
<span class="org-variable-name">x</span> = cd8.loc[:,k]
gene_info.loc[k]
</pre>
</div>

<pre class="example">
chr             hs15
start       69745123
end         69748255
name           RPLP1
strand             +
source    H. sapiens
Name: ENSG00000137818, dtype: object
</pre>

<p>
Look at the deconvolved distributions for this gene. Superimpose the Poisson
point-mass estimates for the two groups (dotted lines), and the Gamma mean
estimate (dashed line).
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(6, 4)
ax[0].hist(x[z], bins=np.arange(x.<span class="org-builtin">max</span>() + 1), alpha=0.6, color=<span class="org-string">'k'</span>, label=<span class="org-string">'Group 1'</span>)
ax[0].hist(x[~z], bins=np.arange(x.<span class="org-builtin">max</span>() + 1), alpha=0.6, color=<span class="org-string">'r'</span>, label=<span class="org-string">'Group 2'</span>)
ax[0].legend(frameon=<span class="org-constant">False</span>)
ax[0].set_xlabel(<span class="org-string">'Num mols'</span>)
ax[0].set_ylabel(<span class="org-string">'Num cells'</span>)

<span class="org-keyword">for</span> c, m <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(plt.get_cmap(<span class="org-string">'Dark2'</span>).colors, [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>]):
  ax[1].plot(*<span class="org-builtin">getattr</span>(scmodes.deconvolve, f<span class="org-string">'fit_{m.lower()}'</span>)(x, s), color=c, lw=1, label=m)

ax[1].axvline(x=scqtl.simple.fit_nb(x, s)[0], c=plt.get_cmap(<span class="org-string">'Dark2'</span>).colors[0], ls=<span class="org-string">'--'</span>, lw=1)
ax[1].axvline(x=de_res.loc[k, <span class="org-string">'mu1'</span>], c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
ax[1].axvline(x=de_res.loc[k, <span class="org-string">'mu2'</span>], c=<span class="org-string">'r'</span>, ls=<span class="org-string">':'</span>, lw=1)

ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
ax[1].legend(frameon=<span class="org-constant">False</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconv-de.org/cd8-rplp1.png" alt="cd8-rplp1.png">
</p>
</div>

<p>
Check whether also allowing dispersions to vary across groups fixes this
gene.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">llr</span> = scqtl.simple.fit_nb(x, s)[-1] - (scqtl.simple.fit_nb(x[z], s[z])[-1] + scqtl.simple.fit_nb(x[~z], s[~z])[-1])
llr, st.chi2(2).sf(-2 * llr)
</pre>
</div>

<pre class="example">
(-2.176938147920737, 0.11338817750598136)

</pre>
</div>
</div>

<div id="outline-container-org262be24" class="outline-4">
<h4 id="org262be24">Gamma deconvolution</h4>
<div class="outline-text-4" id="text-org262be24">
<p>
Deconvolve gene expression assuming group-specific Gammas under the
alternative.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">fit_nb_collapse</span>(x, s):
  <span class="org-doc">"""Return NB or Poisson solution, whichever is more sensible"""</span>
  <span class="org-keyword">try</span>:
    <span class="org-variable-name">res</span> = scqtl.simple.fit_nb(x, s)
  <span class="org-keyword">except</span>:
    <span class="org-comment-delimiter"># </span><span class="org-comment">Failure to converge</span>
    <span class="org-variable-name">res</span> = scqtl.simple.fit_pois(x, s)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Converged to solution with large inverse overdispersion, so log likelihood</span>
  <span class="org-comment-delimiter"># </span><span class="org-comment">is nonsense</span>
  <span class="org-keyword">if</span> res[-1] &gt; 0:
    <span class="org-variable-name">res</span> = scqtl.simple.fit_pois(x, s)
  <span class="org-keyword">return</span> res

<span class="org-keyword">def</span> <span class="org-function-name">de_nb</span>(x, s, z):
  <span class="org-doc">"""Return LRT p-value</span>

<span class="org-doc">  x - counts (n,)</span>
<span class="org-doc">  s - size factors (n,)</span>
<span class="org-doc">  z - boolean group assignment (n,)</span>

<span class="org-doc">  """</span>
  <span class="org-variable-name">res0</span> = fit_nb_collapse(x, s)
  <span class="org-variable-name">res1</span> = fit_nb_collapse(x[z], s[z])
  <span class="org-variable-name">res2</span> = fit_nb_collapse(x[~z], s[~z])
  <span class="org-variable-name">llr</span> = res0[-1] - (res1[-1] + res2[-1])
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'mu0'</span>: res0[0],
                    <span class="org-string">'mu1'</span>: res1[0],
                    <span class="org-string">'mu2'</span>: res2[0],
                    <span class="org-string">'llr'</span>: llr,
                    <span class="org-string">'p'</span>: st.chi2(2).sf(-2 * llr)})
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">de_nb_res</span> = cd8.sample(n=100, axis=1, random_state=1).<span class="org-builtin">apply</span>(de_nb, axis=0, args=(s, z)).T
</pre>
</div>

<p>
Look at the histogram of p-values.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.hist(de_nb_res[<span class="org-string">'p'</span>], np.linspace(0, 1, 11), density=<span class="org-constant">True</span>, color=<span class="org-string">'black'</span>)
plt.axhline(y=1, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.xlabel(<span class="org-string">'P-value'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/deconv-de.org/zheng-cd8-nb.png" alt="zheng-cd8-nb.png">
</p>
</div>

<p>
Look at the QQ plot.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(st.chi2(2).ppf(np.linspace(0, 1, de_nb_res.shape[0])), np.sort(-2 * de_nb_res[<span class="org-string">'llr'</span>]), c=<span class="org-string">'k'</span>, s=2)
plt.plot([0, 11], [0, 11], c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">':'</span>)
plt.xlabel(<span class="org-string">'Expected chi-square'</span>)
plt.ylabel(<span class="org-string">'Observed chi-square'</span>)
</pre>
</div>

<pre class="example">
Text(0, 0.5, 'Observed chi-square')

</pre>

<div class="figure">
<p><img src="figure/deconv-de.org/zheng-cd8-nb-qq.png" alt="zheng-cd8-nb-qq.png">
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-orgb04ebae" class="outline-3">
<h3 id="orgb04ebae">Follow up</h3>
<div class="outline-text-3" id="text-orgb04ebae">
<p>
Joyce Hsiao
<a href="https://stephenslab.github.io/dsc-log-fold-change/eval_initial_type1.html">reports</a>
that commonly used count-based methods (deSeq2, edgeR) do not successfully
control Type 1 error.
</p>

<p>
Why do we see a relatively simple count-based approach adequately
controlling Type 1 error here?
</p>

<ol class="org-ol">
<li>We are comparing much larger groups</li>
<li>We are not shrinking dispersion parameters across genes</li>
</ol>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-05-23 Thu 16:18</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
