<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-03-21 Thu 19:42 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Comparison of expression deconvolution approaches</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Comparison of expression deconvolution approaches</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org7ba1eb0">Introduction</a></li>
<li><a href="#orga732197">Setup</a></li>
<li><a href="#orgb2252bb">Methods</a>
<ul>
<li><a href="#org4b51f6e">Distribution deconvolution</a></li>
</ul>
</li>
<li><a href="#org1b135ab">Results</a>
<ul>
<li><a href="#orgd37e090">Homogeneous cell populations</a></li>
<li><a href="#org439c6e2">Synthetic cell mixtures</a></li>
<li><a href="#org0c2cae0">Real cell mixtures</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org7ba1eb0" class="outline-2">
<h2 id="org7ba1eb0">Introduction</h2>
<div class="outline-text-2" id="text-org7ba1eb0">
<p>
The fundamental question in modeling scRNA-seq count data for a single gene
\(j\) is what distributional assumptions are made on the relative abundances
\(\lambda_{1j}, \ldots, \lambda_{nj} \sim g_j(\cdot)\).
</p>

<p>
Here, we investigate the tradoeff between model complexity/flexibility and
generalization for different choices of \(g_j\) in real data.
</p>
</div>
</div>

<div id="outline-container-orga732197" class="outline-2">
<h2 id="orga732197">Setup</h2>
<div class="outline-text-2" id="text-orga732197">
<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])

<span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> gzip
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.io
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp

<span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
<span class="org-keyword">import</span> rpy2.robjects.numpy2ri

rpy2.robjects.pandas2ri.activate()
rpy2.robjects.numpy2ri.activate()

<span class="org-variable-name">descend</span> = rpy2.robjects.packages.importr(<span class="org-string">'descend'</span>)
<span class="org-variable-name">ashr</span> = rpy2.robjects.packages.importr(<span class="org-string">'ashr'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgb2252bb" class="outline-2">
<h2 id="orgb2252bb">Methods</h2>
<div class="outline-text-2" id="text-orgb2252bb">
</div>
<div id="outline-container-org4b51f6e" class="outline-3">
<h3 id="org4b51f6e">Distribution deconvolution</h3>
<div class="outline-text-3" id="text-org4b51f6e">
<p>
The general form of distribution deconvolution for scRNA-seq is:
</p>

<p>
\[ x_{ij} \sim \mathrm{Poisson}(\exp(\mathbf{z}_i' \mathbf{b}_j) \lambda_{ij}) \]
</p>

<p>
\[ \lambda_{ij} \sim g_j(\cdot) \]
</p>

<p>
where:
</p>

<ul class="org-ul">
<li>\(x_{ij}\) is the count of molecules of gene \(j\) in cell \(i\)</li>
<li>\(\mathbf{z}_i\) is a \(q\)-vector of covariates for cell \(i\)</li>
<li>\(\mathbf{b}_j\) is a \(q\)-vector of confounding effects on gene \(j\)</li>
<li>\(\lambda_{ij}\) is proportional to the relative abundance of gene \(j\)
in cell \(i\)</li>
</ul>

<p>
The primary inference goal is to recover \(g_j\). A secondary goal could be
to recover \(\lambda_{ij}\). We can trade off flexibility and complexity of
\(g_j\) for ease of implementation and speed.
</p>

<ol class="org-ol">
<li><b>NB:</b> \(g_j = \mathrm{Gamma}(\cdot)\). Implemented in <code>scqtl</code>.</li>
<li><b>ZINB:</b> \(g_j = \pi_j \delta_0(\cdot) + (1 - \pi_j)
      \mathrm{Gamma}(\cdot)\) Implemented in <code>scqtl</code>.</li>
<li><b>Unimodal:</b> \(g_j = \sum_k \pi_k \mathrm{Uniform}(\lambda_0, a_{jk})\),
where \(\lambda_0\) is the mode (jointly estimated from the
data). Implemented in <code>ashR</code>.</li>
<li><b>Spline:</b> \(g_j = \exp(\mathbf{Q}'\alpha - \phi(\alpha))\), where
\(\mathbf{Q}\) is a \(5 \times K\) discretized degree 5 cubic
spline. Implemented in <code>DESCEND</code>.</li>
<li><b>Nonparametric:</b> \(g_j = \sum_k \pi_k \mathrm{Uniform}(a_k, b_k)\), where
\(a_k, b_k\) define a uniform grid. Implemented in <code>ashR</code>.</li>
</ol>

<p>
In order to evaluate methods on their ability to estimate \(g_j\), we hold
out samples in a cross-validation setting, and score methods on the log
likelihood of the held out data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">nb_nll</span>(x, mean, inv_disp):
  <span class="org-keyword">return</span> (x * np.log(mean / inv_disp) -
          x * np.log(1 + mean / inv_disp) -
          inv_disp * np.log(1 + mean / inv_disp) +
          sp.gammaln(x + inv_disp) -
          sp.gammaln(inv_disp) -
          sp.gammaln(x + 1))

<span class="org-keyword">def</span> <span class="org-function-name">score_nb</span>(x_train, x_test):
  <span class="org-keyword">import</span> scqtl
  <span class="org-variable-name">onehot</span> = np.ones((x_train.shape[0], 1))
  <span class="org-variable-name">size_factor</span> = x_train.<span class="org-builtin">sum</span>(axis=0).reshape(-1, 1)
  <span class="org-variable-name">design</span> = np.zeros((num_samples, 1))
  log_mu, log_phi, *<span class="org-variable-name">_</span> = scqtl.tf.fit(
    umi=x_train.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=design.astype(np.float32),
    size_factor=size_factor.astype(np.float32),
    learning_rate=1e-3,
    max_epochs=20000,
    verbose=<span class="org-constant">True</span>)
  <span class="org-keyword">return</span> nb_nll(x_test, np.exp(log_mu), np.exp(-log_phi)).<span class="org-builtin">sum</span>()

<span class="org-keyword">def</span> <span class="org-function-name">softplus</span>(x):
  <span class="org-keyword">if</span> x &gt; 30:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> np.log(1 + np.exp(x))

<span class="org-keyword">def</span> <span class="org-function-name">zinb_nll</span>(x, mean, inv_disp, logodds):
  <span class="org-variable-name">case_zero</span> = -softplus(-logodds) + softplus(nb_llik(x, mean, inv_disp) - logodds)
  <span class="org-variable-name">case_non_zero</span> = -softplus(logodds) + nb_llik(x, mean, inv_disp)
  <span class="org-keyword">return</span> np.where(x &lt; 1, case_zero, case_non_zero)

<span class="org-keyword">def</span> <span class="org-function-name">score_zinb</span>(x_train, x_test):
  <span class="org-keyword">import</span> scqtl
  <span class="org-variable-name">onehot</span> = np.ones((x_train.shape[0], 1))
  <span class="org-variable-name">size_factor</span> = x_train.<span class="org-builtin">sum</span>(axis=0).reshape(-1, 1)
  <span class="org-variable-name">design</span> = np.zeros((num_samples, 1))
  <span class="org-variable-name">init</span> = scqtl.tf.fit(
    umi=x_train.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=design.astype(np.float32),
    size_factor=size_factor.astype(np.float32),
    learning_rate=1e-3,
    max_epochs=20000,
    verbose=<span class="org-constant">True</span>)
  <span class="org-variable-name">log_mu</span>, <span class="org-variable-name">log_phi</span>, <span class="org-variable-name">logodds</span>, <span class="org-variable-name">nb_llik</span>, <span class="org-variable-name">zinb_llik</span> = scqtl.tf.fit(
    umi=umi.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=design.astype(np.float32),
    size_factor=size_factor.astype(np.float32),
    learning_rate=1e-3,
    max_epochs=20000,
    warm_start=init[:3],
    verbose=<span class="org-constant">True</span>)
  <span class="org-keyword">return</span> zinb_nll(x_test, np.exp(logodds), np.exp(-log_phi), logodds).<span class="org-builtin">sum</span>()

<span class="org-keyword">def</span> <span class="org-function-name">score_ash</span>(x_train, x_test):
  <span class="org-keyword">pass</span>
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org1b135ab" class="outline-2">
<h2 id="org1b135ab">Results</h2>
<div class="outline-text-2" id="text-org1b135ab">
</div>
<div id="outline-container-orgd37e090" class="outline-3">
<h3 id="orgd37e090">Homogeneous cell populations</h3>
<div class="outline-text-3" id="text-orgd37e090">
<p>
Use sorted cells from Zheng et al. 2016 as a baseline.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">cd8_cytotoxic_t_cells</span>(min_detect=0.25):
  <span class="org-variable-name">counts</span> = scipy.io.mmread(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cd8+_cytotoxic_t_cells/filtered_matrices_mex/hg19/matrix.mtx.gz'</span>).tocsr()
  <span class="org-variable-name">counts</span> = counts[((counts &gt; 0).mean(axis=1) &gt;= min_detect).A.ravel()].T.A.astype(np.<span class="org-builtin">int</span>)
  <span class="org-keyword">return</span> counts

<span class="org-keyword">def</span> <span class="org-function-name">cd19_b_cells</span>(min_detect=0.25):
  <span class="org-variable-name">counts</span> = scipy.io.mmread(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cd19+_b_cells/filtered_matrices_mex/hg19/matrix.mtx.gz'</span>).tocsr()
  <span class="org-variable-name">counts</span> = counts[((counts &gt; 0).mean(axis=1) &gt;= min_detect).A.ravel()].T.A.astype(np.<span class="org-builtin">int</span>)
  <span class="org-keyword">return</span> counts
</pre>
</div>
</div>
</div>

<div id="outline-container-org439c6e2" class="outline-3">
<h3 id="org439c6e2">Synthetic cell mixtures</h3>
</div>

<div id="outline-container-org0c2cae0" class="outline-3">
<h3 id="org0c2cae0">Real cell mixtures</h3>
<div class="outline-text-3" id="text-org0c2cae0">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">pbmcs_68k</span>(min_detect=0.25):
  <span class="org-variable-name">counts</span> = scipy.io.mmread(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/68k_pbmcs/filtered_matrices_mex/hg19/matrix.mtx.gz'</span>).tocsr()
  <span class="org-variable-name">counts</span> = counts[((counts &gt; 0).mean(axis=1) &gt;= min_detect).A.ravel()].T.A.astype(np.<span class="org-builtin">int</span>)
  <span class="org-keyword">return</span> counts

<span class="org-keyword">def</span> <span class="org-function-name">cortex</span>():
  <span class="org-variable-name">counts</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/zeisel-2015/GSE60361_C1-3005-Expression.txt.gz'</span>, index_col=0)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Follow scVI here</span>
  <span class="org-keyword">return</span> counts.loc[counts.var(axis=1).sort_values(ascending=<span class="org-constant">False</span>).head(n=500).index].values.T
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-03-21 Thu 19:42</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
