<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-09-01 Tue 23:45 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Negative binomial measurement model</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Negative binomial measurement model</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org040c2a7">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#orgf3cc87c">Results</a>
<ul>
<li><a href="#org72a0851">VBEM algorithm for Gamma expression model</a></li>
<li><a href="#org1a493b0">Simulation</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org040c2a7" class="outline-2">
<h2 id="org040c2a7">Introduction</h2>
<div class="outline-text-2" id="text-org040c2a7">
<p>
We, and Svensson 2020, found some evidence for overdispersion in control
scRNA-seq data. Here, we estimate to what extent that overdispersion could be
explained by an overdispersed measurement model, using the key fact that the
measurement overdispersion is described by a single parameter common across
all genes. We specifically consider combining an NB measurement model with a
Gamma expression model \(
  \DeclareMathOperator\Pois{Poisson}
  \DeclareMathOperator\Gam{Gamma}
  \DeclareMathOperator\NB{NB}
  \newcommand\const{\mathrm{const}}
  \newcommand\lnb{l_{\mathrm{NB}}}
  \newcommand\E[1]{\left\langle #1 \right\rangle}
  \)
</p>

\begin{align}
  x_{ij} \mid s_i, \lambda_{ij}, u_{ij} &\sim \NB(s_i \lambda_{ij}, \theta)\\
  \lambda_{ij} \mid a_j, b_j &\sim \Gam(a_j, b_j),
\end{align}

<p>
where the NB distribution is parameterized by mean and dispersion. 
</p>
</div>
</div>

<div id="outline-container-orgf1021bd" class="outline-2">
<h2 id="setup"><a id="orgf1021bd"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgf3cc87c" class="outline-2">
<h2 id="orgf3cc87c">Results</h2>
<div class="outline-text-2" id="text-orgf3cc87c">
</div>
<div id="outline-container-org72a0851" class="outline-3">
<h3 id="org72a0851">VBEM algorithm for Gamma expression model</h3>
<div class="outline-text-3" id="text-org72a0851">
<p>
To estimate \(a_1, \ldots, a_p, b_1, \ldots, b_p, \theta\) from observed
data, we use a VBEM algorithm. First, introduce latent variables \(u_{ij}\)
</p>

\begin{align}
  x_{ij} \mid s_i, \lambda_{ij}, u_{ij} &\sim \Pois(s_i \lambda_{ij} u_{ij})\\
  u_{ij} \mid \theta &\sim \Gam(\theta^{-1}, \theta^{-1})\\
  \lambda_{ij} \mid a_j, b_j &\sim \Gam(a_j, b_j),
\end{align}

<p>
where the Gamma distribution is parameterized by shape and rate. It is
straightforward to show that marginalizing over \(u_{ij}\) yields the
original NB-Gamma compound model of interest. The log joint
</p>

\begin{multline}
  \ln p(x_{ij} \mid \lambda_{ij}, u_{ij}, a_j, b_j, \theta) = x_{ij} \ln (s_i \lambda_{ij} u_{ij}) - s_i \lambda_{ij} u_{ij} - \ln\Gamma(x_{ij} + 1)\\
  + (a_j - 1) \ln \lambda_{ij} - b_j \lambda_{ij} + a_j \ln b_j - \ln\Gamma(a_j) + (\theta^{-1} - 1) \ln u_{ij} - \theta^{-1} u_{ij} + \theta^{-1}\ln\theta^{-1} - \ln\Gamma(\theta^{-1}),
\end{multline}

<p>
and the posteriors
</p>

\begin{align}
  \ln p(\lambda_{ij} \mid x_{ij}, u_{ij}, a_j, b_j) &= (x_{ij} + a_j - 1) \ln \lambda_{ij} - (s_i u_{ij} + b_j) \lambda_{ij} + \const\\
  &= \Gam(x_{ij} + a_j, s_i u_{ij} + b_j)\\
  \ln p(u_{ij} \mid x_{ij}, \lambda_{ij}, a_j, b_j) &= (x_{ij} + \theta^{-1} - 1) \ln \lambda_{ij} - (s_i \lambda_{ij} + b_j) u_{ij} + \const\\
  &= \Gam(x_{ij} + \theta^{-1}, s_i \lambda_{ij} + b_j).
\end{align}

<p>
However, the required expectations for an EM algorithm that directly
maximizes the likelihood are non-analytic. To side-step this problem,
introduce a variational approximation
</p>

\begin{align}
  q &= \prod_{i,j} q(\lambda_{ij}) q(u_{ij})\\
  q^*(\lambda_{ij}) &\propto \exp((x_{ij} + a_j - 1) \ln \lambda_{ij} - (s_i \E{u_{ij}} + b_j) \lambda_{ij})\\
  &= \Gam(x_{ij} + a_j, s_i \E{u_{ij}} + b_j)\\
  &\triangleq \Gam(\alpha_{ij}, \beta_{ij})\\
  q^*(u_{ij}) &\propto \exp((x_{ij} + \theta^{-1} - 1) \ln u_{ij} - (s_i \E{\lambda_{ij}} + b_j) u_{ij})\\
  &= \Gam(x_{ij} + \theta^{-1}, s_i \E{\lambda_{ij}} + \theta^{-1})\\
  &\triangleq \Gam(\gamma_{ij}, \delta_{ij}).
\end{align}

<p>
The evidence lower bound
</p>

\begin{multline}
  \ell = \sum_{i, j} \left[ (x_{ij} + a_j - \alpha_{ij}) \E{\ln \lambda_{ij}} - (b_j - \beta_{ij}) \E{\lambda_{ij}} + (x_{ij} + \theta^{-1} - \gamma_{ij}) \E{\ln u_{ij}} - (\theta^{-1} - \delta_{ij}) \E{u_{ij}} - s_i \E{\lambda_{ij}} \E{u_{ij}}\right.\\
    + \left. a_j \ln b_j + \theta^{-1}\ln\theta^{-1} - \alpha_{ij} \ln \beta_{ij} - \gamma_{ij} \ln \delta_{ij} - \ln\Gamma(a_j) - \ln\Gamma(\theta^{-1}) + \ln\Gamma(\alpha_{ij}) + \ln\Gamma(\gamma_{ij})\right] + \const,
\end{multline}

<p>
where
</p>

\begin{align}
  \E{\lambda_{ij}} &= \alpha_{ij} / \beta_{ij}\\
  \E{\ln\lambda_{ij}} &= \psi(\alpha_{ij}) - \ln(\beta_{ij})\\
  \E{u_{ij}} &= \gamma_{ij} / \delta_{ij}\\
  \E{\ln u_{ij}} &= \psi(\gamma_{ij}) - \ln(\delta_{ij}),
\end{align}

<p>
and \(\psi\) denotes the digamma function. Then, we have analytic M step update
</p>

\begin{align}
  \frac{\partial \ell}{\partial b_j} &= \sum_{i} \frac{a_j}{b_j} - \E{\lambda_{ij}} = 0\\
  b_j &:= \frac{n a_j}{\sum_i \E{\lambda_{ij}}}
\end{align}

<p>
and Newton-Raphson partial M step updates
</p>

\begin{align}
  \eta &\triangleq \theta^{-1}\\
  \frac{\partial \ell}{\partial \eta} &= \sum_{i, j} 1 + \E{\ln u_{ij}} - \E{u_{ij}} - \psi(\eta)\\
  \frac{\partial^2 \ell}{\partial \eta^2} &= -n p \psi^{(1)}(\eta)\\
  \frac{\partial \ell}{\partial a_j} &= \sum_i \E{\ln\lambda_{ij}} + \ln b_j - \psi(a_j)\\
  \frac{\partial^2 \ell}{\partial a_j^2} &=  -n \psi^{(1)}(a_j),
\end{align}

<p>
where \(\psi^{(1)}\) denotes the trigamma function.
</p>
</div>
</div>

<div id="outline-container-org1a493b0" class="outline-3">
<h3 id="org1a493b0">Simulation</h3>
<div class="outline-text-3" id="text-org1a493b0">
<p>
Implement a simplified DSC.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">simulate_nb_gamma</span>(n, p, theta, seed=0):
  np.random.seed(seed)
  <span class="org-variable-name">log_mean</span> = np.random.uniform(low=-12, high=-8, size=(1, p))
  <span class="org-variable-name">log_disp</span> = np.random.uniform(low=-6, high=0, size=(1, p))
  <span class="org-variable-name">s</span> = 1e5 * np.ones((n, 1))
  <span class="org-variable-name">lam</span> = st.gamma(a=np.exp(-log_disp), scale=np.exp(log_mean + log_disp)).rvs(size=(n, p))
  <span class="org-variable-name">u</span> = st.gamma(a=1 / theta, scale=theta).rvs(size=(n, p))
  <span class="org-variable-name">x</span> = st.poisson(s * lam * u).rvs()
  <span class="org-keyword">return</span> x, s, lam, u, log_mean, -log_disp, theta

<span class="org-keyword">def</span> <span class="org-function-name">evaluate_ebnbm_gamma</span>(n, p, n_trials=1):
  <span class="org-variable-name">res</span> = <span class="org-builtin">dict</span>()
  <span class="org-keyword">for</span> log_theta <span class="org-keyword">in</span> (-2,):
    <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_trials):
      <span class="org-variable-name">x</span>, <span class="org-variable-name">s</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">log_mean</span>, <span class="org-variable-name">log_inv_disp</span>, <span class="org-variable-name">_</span> = simulate_nb_gamma(n, p, theta=np.exp(log_theta), seed=i)
      log_mean_hat, log_inv_disp_hat, log_inv_meas_disp_hat, *<span class="org-variable-name">_</span> = scmodes.ebnbm.ebnbm_gamma(x, s, extrapolate=<span class="org-constant">False</span>)
      <span class="org-variable-name">res</span>[(log_theta, i)] = pd.DataFrame({<span class="org-string">'log_mean'</span>: log_mean.ravel(),
                                          <span class="org-string">'log_inv_disp'</span>: log_inv_disp.ravel(),
                                          <span class="org-string">'log_mean_hat'</span>: log_mean_hat.ravel(),
                                          <span class="org-string">'log_inv_disp_hat'</span>: log_inv_disp_hat.ravel()})
      <span class="org-comment-delimiter"># </span><span class="org-comment">res[(log_theta, i)]['log_inv_meas_disp_hat'] = log_inv_meas_disp_hat</span>
  <span class="org-variable-name">res</span> = (pd.concat(res)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'log_theta'</span>, <span class="org-string">'level_1'</span>: <span class="org-string">'trial'</span>, <span class="org-string">'level_2'</span>: <span class="org-string">'gene'</span>}, axis=1))
  <span class="org-keyword">return</span> res
</pre>
</div>

<p>
Simulate an example. Fit the model, first fixing the hyperparameters to the
ground truth values.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span>, <span class="org-variable-name">s</span>, <span class="org-variable-name">lam</span>, <span class="org-variable-name">u</span>, <span class="org-variable-name">log_mean</span>, <span class="org-variable-name">log_inv_disp</span>, <span class="org-variable-name">theta</span> = simulate_nb_gamma(n=1000, p=5, theta=0.2, seed=1)
<span class="org-variable-name">log_mean_hat</span>, <span class="org-variable-name">log_inv_disp_hat</span>, <span class="org-variable-name">log_meas_disp_hat</span>, <span class="org-variable-name">alpha</span>, <span class="org-variable-name">beta</span>, <span class="org-variable-name">gamma</span>, <span class="org-variable-name">delta</span>, <span class="org-variable-name">elbo</span> = scmodes.ebnbm.ebnbm_gamma(
  x,
  s,
  init=np.hstack([np.exp(log_inv_disp).ravel(), np.exp(-log_mean + log_inv_disp).ravel(), theta]),
  tol=1e-5,
  extrapolate=<span class="org-constant">False</span>,
  verbose=<span class="org-constant">True</span>,
  fix_theta=<span class="org-constant">True</span>)

<span class="org-variable-name">y</span> = np.arange(x[:,0].<span class="org-builtin">max</span>() + 1)

<span class="org-comment-delimiter"># </span><span class="org-comment">Poisson measurement =&gt; NB observation</span>
<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">pmf</span>[<span class="org-string">'Poisson'</span>] = st.nbinom(n=np.exp(log_inv_disp[:,0]), p=1 / (1 + (s * np.exp(log_mean[:,0] - log_inv_disp[:,0])))).pmf(y).mean(axis=0)

<span class="org-comment-delimiter"># </span><span class="org-comment">NB measurement =&gt; Monte Carlo integral</span>
<span class="org-variable-name">n_samples</span> = 1000
<span class="org-variable-name">G</span> = st.gamma(a=np.exp(log_inv_disp[:,0]), scale=np.exp(log_mean[:,0] - log_inv_disp[:,0]))
<span class="org-variable-name">temp</span> = G.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\theta$ = {theta:.2g})'</span>] = st.nbinom(n=1 / theta, p=1 / (1 + s[0] * temp * theta)).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">Ghat</span> = st.gamma(a=np.exp(log_inv_disp_hat[:,0]), scale=np.exp(log_mean_hat[:,0] - log_inv_disp_hat[:,0]))
<span class="org-variable-name">temp</span> = Ghat.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\hat\theta$ = {np.exp(log_meas_disp_hat):.2g})'</span>] = st.nbinom(n=np.exp(-log_meas_disp_hat), p=1 / (1 + s[0] * temp * np.exp(log_meas_disp_hat))).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">cdf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">grid</span> = np.linspace(0, (x / s).<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdf</span>[<span class="org-string">'Estimated'</span>] = Ghat.cdf(grid).ravel()
<span class="org-variable-name">cdf</span>[<span class="org-string">'Ground truth'</span>] = G.cdf(grid).ravel()
</pre>
</div>

<p>
Make sure we didn&rsquo;t mess up the parameterization.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4.5, 4)

ax[0].hist(x[:,0], bins=y, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[0].plot(y + .5, pmf[k], c=cm(i), lw=1, label=k)
ax[0].legend(title=<span class="org-string">'Measurement model'</span>, frameon=<span class="org-constant">False</span>)
ax[0].set_xlim(0, x[:,0].<span class="org-builtin">max</span>())
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)

<span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Paired'</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(cdf):
  ax[1].plot(grid, cdf[k], c=cm(i), lw=1, label=k)
ax[1].legend(title=<span class="org-string">'Expression model'</span>, frameon=<span class="org-constant">False</span>)
ax[1].set_xlim(0, 1.5e-4)
ax[1].set_xticks([0, 5e-5, 1e-4, 1.5e-4])
ax[1].set_xlabel(<span class="org-string">'True gene expression level'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex.png" alt="sim-ex.png">
</p>
</div>

<p>
Compare the true posterior against the variational approximation.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2)
fig.set_size_inches(5, 2.5)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_aspect(<span class="org-string">'equal'</span>, adjustable=<span class="org-string">'datalim'</span>)

ax[0].set_xscale(<span class="org-string">'log'</span>)
ax[0].set_yscale(<span class="org-string">'log'</span>)
ax[0].scatter(((x + np.exp(log_inv_disp)) / (s * u + np.exp(log_inv_disp - log_mean))).ravel(), (alpha / beta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[0].set_xlabel(r<span class="org-string">'$\mathrm{E}[\lambda \mid x, u]$'</span>)
ax[0].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[\lambda]$'</span>)

ax[1].scatter(((x + theta) / (s * lam + theta)), (gamma / delta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[1].set_xlabel(r<span class="org-string">'$\mathrm{E}[u \mid x, \mu]$'</span>)
ax[1].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[u]$'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex2.png" alt="sim-ex2.png">
</p>
</div>

<p>
Now, fit the model initialized at the ground truth.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_mean_hat</span>, <span class="org-variable-name">log_inv_disp_hat</span>, <span class="org-variable-name">log_meas_disp_hat</span>, <span class="org-variable-name">alpha</span>, <span class="org-variable-name">beta</span>, <span class="org-variable-name">gamma</span>, <span class="org-variable-name">delta</span>, <span class="org-variable-name">elbo</span> = scmodes.ebnbm.ebnbm_gamma(
  x,
  s,
  init=np.hstack([np.exp(log_inv_disp).ravel(), np.exp(-log_mean + log_inv_disp).ravel(), theta]),
  tol=1e-3, 
  max_iters=20000,
  extrapolate=<span class="org-constant">False</span>,
  verbose=<span class="org-constant">True</span>,
  fix_theta=<span class="org-constant">False</span>)

<span class="org-variable-name">y</span> = np.arange(x[:,0].<span class="org-builtin">max</span>() + 1)

<span class="org-comment-delimiter"># </span><span class="org-comment">Poisson measurement =&gt; NB observation</span>
<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">pmf</span>[<span class="org-string">'Poisson'</span>] = st.nbinom(n=np.exp(log_inv_disp[:,0]), p=1 / (1 + (s * np.exp(log_mean[:,0] - log_inv_disp[:,0])))).pmf(y).mean(axis=0)

<span class="org-comment-delimiter"># </span><span class="org-comment">NB measurement =&gt; Monte Carlo integral</span>
<span class="org-variable-name">n_samples</span> = 1000
<span class="org-variable-name">G</span> = st.gamma(a=np.exp(log_inv_disp[:,0]), scale=np.exp(log_mean[:,0] - log_inv_disp[:,0]))
<span class="org-variable-name">temp</span> = G.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\theta$ = {theta:.2g})'</span>] = st.nbinom(n=1 / theta, p=1 / (1 + s[0] * temp * theta)).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">Ghat</span> = st.gamma(a=np.exp(log_inv_disp_hat[:,0]), scale=np.exp(log_mean_hat[:,0] - log_inv_disp_hat[:,0]))
<span class="org-variable-name">temp</span> = Ghat.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\hat\theta$ = {np.exp(log_meas_disp_hat):.2g})'</span>] = st.nbinom(n=np.exp(-log_meas_disp_hat), p=1 / (1 + s[0] * temp * np.exp(log_meas_disp_hat))).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">cdf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">grid</span> = np.linspace(0, (x / s).<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdf</span>[<span class="org-string">'Estimated'</span>] = Ghat.cdf(grid).ravel()
<span class="org-variable-name">cdf</span>[<span class="org-string">'Ground truth'</span>] = G.cdf(grid).ravel()
</pre>
</div>

<p>
Plot the fitted models against the observed data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4.5, 4)

ax[0].hist(x[:,0], bins=y, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[0].plot(y + .5, pmf[k], c=cm(i), lw=1, label=k)
ax[0].legend(title=<span class="org-string">'Measurement model'</span>, frameon=<span class="org-constant">False</span>)
ax[0].set_xlim(0, x[:,0].<span class="org-builtin">max</span>())
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)

<span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Paired'</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(cdf):
  ax[1].plot(grid, cdf[k], c=cm(i), lw=1, label=k)
ax[1].legend(title=<span class="org-string">'Expression model'</span>, frameon=<span class="org-constant">False</span>)
ax[1].set_xlim(0, 1.5e-4)
ax[1].set_xticks([0, 5e-5, 1e-4, 1.5e-4])
ax[1].set_xlabel(<span class="org-string">'True gene expression level'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex3.png" alt="sim-ex3.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">y</span> = np.arange(x[:,1].<span class="org-builtin">max</span>() + 1)

<span class="org-comment-delimiter"># </span><span class="org-comment">Poisson measurement =&gt; NB observation</span>
<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">pmf</span>[<span class="org-string">'Poisson'</span>] = st.nbinom(n=np.exp(log_inv_disp[:,1]), p=1 / (1 + (s * np.exp(log_mean[:,1] - log_inv_disp[:,1])))).pmf(y).mean(axis=0)

<span class="org-comment-delimiter"># </span><span class="org-comment">NB measurement =&gt; Monte Carlo integral</span>
<span class="org-variable-name">n_samples</span> = 1000
<span class="org-variable-name">G</span> = st.gamma(a=np.exp(log_inv_disp[:,1]), scale=np.exp(log_mean[:,1] - log_inv_disp[:,1]))
<span class="org-variable-name">temp</span> = G.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\theta$ = {theta:.2g})'</span>] = st.nbinom(n=1 / theta, p=1 / (1 + s[0] * temp * theta)).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">Ghat</span> = st.gamma(a=np.exp(log_inv_disp_hat[:,1]), scale=np.exp(log_mean_hat[:,1] - log_inv_disp_hat[:,1]))
<span class="org-variable-name">temp</span> = Ghat.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\hat\theta$ = {np.exp(log_meas_disp_hat):.2g})'</span>] = st.nbinom(n=np.exp(-log_meas_disp_hat), p=1 / (1 + s[0] * temp * np.exp(log_meas_disp_hat))).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">cdf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">grid</span> = np.linspace(0, (x / s).<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdf</span>[<span class="org-string">'Estimated'</span>] = Ghat.cdf(grid).ravel()
<span class="org-variable-name">cdf</span>[<span class="org-string">'Ground truth'</span>] = G.cdf(grid).ravel()

<span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4.5, 4)

ax[0].hist(x[:,1], bins=y, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[0].plot(y + .5, pmf[k], c=cm(i), lw=1, label=k)
ax[0].legend(title=<span class="org-string">'Measurement model'</span>, frameon=<span class="org-constant">False</span>)
ax[0].set_xlim(0, y.<span class="org-builtin">max</span>())
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)

<span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Paired'</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(cdf):
  ax[1].plot(grid, cdf[k], c=cm(i), lw=1, label=k)
ax[1].legend(title=<span class="org-string">'Expression model'</span>, frameon=<span class="org-constant">False</span>)
ax[1].set_xlabel(<span class="org-string">'True gene expression level'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex3-2.png" alt="sim-ex3-2.png">
</p>
</div>

<p>
Compare the true posterior against the variational approximation.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2)
fig.set_size_inches(5, 2.5)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_aspect(<span class="org-string">'equal'</span>, adjustable=<span class="org-string">'datalim'</span>)

ax[0].set_xscale(<span class="org-string">'log'</span>)
ax[0].set_yscale(<span class="org-string">'log'</span>)
ax[0].scatter(((x + np.exp(log_inv_disp)) / (s * u + np.exp(log_inv_disp - log_mean))).ravel(), (alpha / beta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[0].set_xlabel(r<span class="org-string">'$\mathrm{E}[\lambda \mid x, u]$'</span>)
ax[0].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[\lambda]$'</span>)

ax[1].scatter(((x + theta) / (s * lam + theta)), (gamma / delta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[1].set_xlabel(r<span class="org-string">'$\mathrm{E}[u \mid x, \mu]$'</span>)
ax[1].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[u]$'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex4.png" alt="sim-ex4.png">
</p>
</div>

<p>
Compare the estimated expression models against the ground truth.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2)
fig.set_size_inches(5, 2.5)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_aspect(<span class="org-string">'equal'</span>, <span class="org-string">'datalim'</span>)
ax[0].scatter(log_mean, log_mean_hat, s=4, c=<span class="org-string">'k'</span>)
<span class="org-variable-name">lim</span> = ax[0].get_xlim()
ax[0].plot(lim, lim, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
ax[0].set_xlabel(<span class="org-string">'$\ln(\mu)$'</span>)
ax[0].set_ylabel(<span class="org-string">'$\ln(\hat\mu)$'</span>)
ax[1].scatter(log_inv_disp, log_inv_disp_hat, s=4, c=<span class="org-string">'k'</span>)
<span class="org-variable-name">lim</span> = ax[1].get_xlim()
ax[1].plot(lim, lim, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
ax[1].set_xlabel(<span class="org-string">'$\ln(\phi)$'</span>)
ax[1].set_ylabel(<span class="org-string">'$\ln(\hat\phi)$'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex5.png" alt="sim-ex5.png">
</p>
</div>

<p>
Now initialize the expression models at the ground truth, and start the
measurement dispersion from close to zero.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">init</span> = np.hstack([np.exp(log_inv_disp).ravel(), np.exp(-log_mean + log_inv_disp).ravel(), 1e-3])
<span class="org-variable-name">log_mean_hat</span>, <span class="org-variable-name">log_inv_disp_hat</span>, <span class="org-variable-name">log_meas_disp_hat</span>, <span class="org-variable-name">alpha</span>, <span class="org-variable-name">beta</span>, <span class="org-variable-name">gamma</span>, <span class="org-variable-name">delta</span>, <span class="org-variable-name">elbo</span> = scmodes.ebnbm.ebnbm_gamma(
  x,
  s,
  init=init,
  tol=1e-3, 
  extrapolate=<span class="org-constant">False</span>,
  verbose=<span class="org-constant">True</span>,
  fix_theta=<span class="org-constant">False</span>)

<span class="org-variable-name">y</span> = np.arange(x[:,0].<span class="org-builtin">max</span>() + 1)

<span class="org-comment-delimiter"># </span><span class="org-comment">Poisson measurement =&gt; NB observation</span>
<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">pmf</span>[<span class="org-string">'Poisson'</span>] = st.nbinom(n=np.exp(log_inv_disp[:,0]), p=1 / (1 + (s * np.exp(log_mean[:,0] - log_inv_disp[:,0])))).pmf(y).mean(axis=0)

<span class="org-comment-delimiter"># </span><span class="org-comment">NB measurement =&gt; Monte Carlo integral</span>
<span class="org-variable-name">n_samples</span> = 1000
<span class="org-variable-name">G</span> = st.gamma(a=np.exp(log_inv_disp[:,0]), scale=np.exp(log_mean[:,0] - log_inv_disp[:,0]))
<span class="org-variable-name">temp</span> = G.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\theta$ = {theta:.2g})'</span>] = st.nbinom(n=1 / theta, p=1 / (1 + s[0] * temp * theta)).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">Ghat</span> = st.gamma(a=np.exp(log_inv_disp_hat[:,0]), scale=np.exp(log_mean_hat[:,0] - log_inv_disp_hat[:,0]))
<span class="org-variable-name">temp</span> = Ghat.rvs(size=(n_samples, y.shape[0], 1))
<span class="org-variable-name">pmf</span>[rf<span class="org-string">'NB ($\hat\theta$ = {np.exp(log_meas_disp_hat):.2g})'</span>] = st.nbinom(n=np.exp(-log_meas_disp_hat), p=1 / (1 + s[0] * temp * np.exp(log_meas_disp_hat))).pmf(y.reshape(-1, 1)).mean(axis=0)

<span class="org-variable-name">cdf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">grid</span> = np.linspace(0, (x / s).<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdf</span>[<span class="org-string">'Estimated'</span>] = Ghat.cdf(grid).ravel()
<span class="org-variable-name">cdf</span>[<span class="org-string">'Ground truth'</span>] = G.cdf(grid).ravel()
</pre>
</div>

<p>
Plot the fit.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4.5, 4)

ax[0].hist(x[:,0], bins=y, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[0].plot(y + .5, pmf[k], c=cm(i), lw=1, label=k)
ax[0].legend(title=<span class="org-string">'Measurement model'</span>, frameon=<span class="org-constant">False</span>)
ax[0].set_xlim(0, x[:,0].<span class="org-builtin">max</span>())
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)

<span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Paired'</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(cdf):
  ax[1].plot(grid, cdf[k], c=cm(i), lw=1, label=k)
ax[1].legend(title=<span class="org-string">'Expression model'</span>, frameon=<span class="org-constant">False</span>)
ax[1].set_xlim(0, 1.5e-4)
ax[1].set_xticks([0, 5e-5, 1e-4, 1.5e-4])
ax[1].set_xlabel(<span class="org-string">'True gene expression level'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex6.png" alt="sim-ex6.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2)
fig.set_size_inches(5, 2.5)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_aspect(<span class="org-string">'equal'</span>, <span class="org-string">'datalim'</span>)
ax[0].scatter(log_mean, log_mean_hat, s=4, c=<span class="org-string">'k'</span>)
<span class="org-variable-name">lim</span> = ax[0].get_xlim()
ax[0].plot(lim, lim, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
ax[0].set_xlabel(<span class="org-string">'$\ln(\mu)$'</span>)
ax[0].set_ylabel(<span class="org-string">'$\ln(\hat\mu)$'</span>)
ax[1].scatter(log_inv_disp, log_inv_disp_hat, s=4, c=<span class="org-string">'k'</span>)
<span class="org-variable-name">lim</span> = ax[1].get_xlim()
ax[1].plot(lim, lim, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
ax[1].set_xlabel(<span class="org-string">'$\ln(\phi)$'</span>)
ax[1].set_ylabel(<span class="org-string">'$\ln(\hat\phi)$'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex7.png" alt="sim-ex7.png">
</p>
</div>

<p>
Plot the line search for \(\eta\).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
<span class="org-variable-name">grid</span> = np.linspace(0, 100, 1000)[1:]
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">pm</span> = (x + 1 / theta) / (s * lam + 1 / theta)
<span class="org-variable-name">plm</span> = sp.digamma(x + 1 / theta) - np.log(s * lam + 1 / theta)
plt.plot(grid, [-(a * np.log(a) + a * plm - a * pm - sp.gammaln(a)).<span class="org-builtin">sum</span>() <span class="org-keyword">for</span> a <span class="org-keyword">in</span> grid], lw=1, c=cm(0), label=<span class="org-string">'Ground truth'</span>)
<span class="org-variable-name">pm</span> = gamma / delta
<span class="org-variable-name">plm</span> = sp.digamma(gamma) - np.log(delta)
plt.plot(grid, [-(a * np.log(a) + a * plm - a * pm - sp.gammaln(a)).<span class="org-builtin">sum</span>() <span class="org-keyword">for</span> a <span class="org-keyword">in</span> grid], lw=1, c=cm(1), label=<span class="org-string">'Var. approx.'</span>)
plt.axvline(x=5, lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.legend()
plt.xlabel(r<span class="org-string">'$\eta$'</span>)
plt.ylabel(<span class="org-string">'Loss'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex-eta.png" alt="sim-ex-eta.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2)
fig.set_size_inches(5, 2.5)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_aspect(<span class="org-string">'equal'</span>, adjustable=<span class="org-string">'datalim'</span>)

ax[0].set_xscale(<span class="org-string">'log'</span>)
ax[0].set_yscale(<span class="org-string">'log'</span>)
ax[0].scatter(((x + np.exp(log_inv_disp)) / (s * u + np.exp(log_inv_disp - log_mean))).ravel(), (alpha / beta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[0].set_xlabel(r<span class="org-string">'$\mathrm{E}[\lambda \mid x, u]$'</span>)
ax[0].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[\lambda]$'</span>)

ax[1].scatter(((x + theta) / (s * lam + theta)), (gamma / delta).ravel(), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[1].set_xlabel(r<span class="org-string">'$\mathrm{E}[u \mid x, \mu]$'</span>)
ax[1].set_ylabel(r<span class="org-string">'$\mathrm{E}_q[u]$'</span>)

fig.tight_layout()

</pre>
</div>


<div class="figure">
<p><img src="figure/ebnbm.org/sim-ex8.png" alt="sim-ex8.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-09-01 Tue 23:45</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
