<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-02-09 Sun 23:00 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Manuscript figures</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Manuscript figures</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#setup">Setup</a></li>
<li><a href="#orge46b8cc">Figure 1</a></li>
<li><a href="#orgc60948c">Figure 2</a></li>
</ul>
</div>
</div>

<div id="outline-container-org39f1bb1" class="outline-2">
<h2 id="setup"><a id="org39f1bb1"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scmodes
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> sqlite3
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'svg'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orge46b8cc" class="outline-2">
<h2 id="orge46b8cc">Figure 1</h2>
<div class="outline-text-2" id="text-orge46b8cc">
<p>
Draw an example of our approach to testing GOF using real data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cytotoxic_t</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene</span> = <span class="org-string">'ENSG00000109475'</span>
<span class="org-variable-name">x</span> = cytotoxic_t.loc[:,gene]
<span class="org-variable-name">s</span> = cytotoxic_t.<span class="org-builtin">sum</span>(axis=1)
<span class="org-variable-name">y</span> = np.arange(x.<span class="org-builtin">max</span>() + 1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">rpp</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
np.random.seed(1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gamma_res</span> = scmodes.ebpm.ebpm_gamma(x, s)
<span class="org-variable-name">rpp</span>[<span class="org-string">'Gamma'</span>] = scmodes.benchmark.gof._rpp(
  scmodes.benchmark.gof._zig_cdf(x - 1, size=s, log_mu=gamma_res[0], log_phi=-gamma_res[1]),
  scmodes.benchmark.gof._zig_pmf(x, size=s, log_mu=gamma_res[0], log_phi=-gamma_res[1]))
<span class="org-variable-name">pmf</span>[<span class="org-string">'Gamma'</span>] = np.array([scmodes.benchmark.gof._zig_pmf(k, size=s, log_mu=gamma_res[0], log_phi=-gamma_res[1]).mean()
                         <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">point_gamma_res</span> = scmodes.ebpm.ebpm_point_gamma(x, s)
<span class="org-variable-name">rpp</span>[<span class="org-string">'ZIG'</span>] = scmodes.benchmark.gof._rpp(
  scmodes.benchmark.gof._zig_cdf(x - 1, size=s, log_mu=point_gamma_res[0], log_phi=-point_gamma_res[1], logodds=point_gamma_res[2]),
  scmodes.benchmark.gof._zig_pmf(x, size=s, log_mu=point_gamma_res[0], log_phi=-point_gamma_res[1], logodds=point_gamma_res[2]))
<span class="org-variable-name">pmf</span>[<span class="org-string">'ZIG'</span>] = np.array([scmodes.benchmark.gof._zig_pmf(k, size=s, log_mu=point_gamma_res[0], log_phi=-point_gamma_res[1], logodds=point_gamma_res[2]).mean()
                       <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">zief_res</span> = scmodes.ebpm.ebpm_point_expfam(x, s)
<span class="org-variable-name">rpp</span>[<span class="org-string">'ZIEF'</span>] = scmodes.benchmark.gof._rpp(
  scmodes.benchmark.gof._point_expfam_cdf(x.values.ravel() - 1, size=s, res=zief_res),
  scmodes.benchmark.gof._point_expfam_pmf(x.values.ravel(), size=s, res=zief_res))
<span class="org-comment-delimiter"># </span><span class="org-comment">We need np.full here because _point_expfam_pmf does not broadcast</span>
<span class="org-variable-name">pmf</span>[<span class="org-string">'ZIEF'</span>] = np.array([scmodes.benchmark.gof._point_expfam_pmf(np.full(x.shape, k), size=s, res=zief_res).mean()
                        <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">unimodal_res</span> = scmodes.ebpm.ebpm_unimodal(x, s)
<span class="org-variable-name">rpp</span>[<span class="org-string">'Unimodal'</span>] = scmodes.benchmark.gof._rpp(
  scmodes.benchmark.gof._ash_cdf(x - 1, s=s, fit=unimodal_res),
  scmodes.benchmark.gof._ash_pmf(x, s=s, fit=unimodal_res))
<span class="org-comment-delimiter"># </span><span class="org-comment">It is simpler to compute this here than to mess with the ash_data object</span>
<span class="org-variable-name">g</span> = np.array(unimodal_res.rx2(<span class="org-string">'fitted_g'</span>))
<span class="org-variable-name">a</span> = np.fmin(g[1], g[2])
<span class="org-variable-name">b</span> = np.fmax(g[1], g[2])
<span class="org-variable-name">comp_dens_conv</span> = np.array([((st.gamma(a=k + 1, scale=1 / s.values.reshape(-1, 1)).cdf(b.reshape(1, -1)) - st.gamma(a=k + 1, scale=1 / s.values.reshape(-1, 1)).cdf(a.reshape(1, -1))) / np.outer(s, b - a)).mean(axis=0) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
<span class="org-variable-name">comp_dens_conv</span>[:,0] = st.poisson(mu=s.values.reshape(-1, 1) * b[0]).pmf(y).mean(axis=0)
<span class="org-variable-name">pmf</span>[<span class="org-string">'Unimodal'</span>] = comp_dens_conv @ g[0]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">npmle_res</span> = scmodes.ebpm.ebpm_npmle(x, s)
<span class="org-variable-name">rpp</span>[<span class="org-string">'NPMLE'</span>] = scmodes.benchmark.gof._rpp(
  scmodes.benchmark.gof._ash_cdf(x - 1, s=s, fit=npmle_res),
  scmodes.benchmark.gof._ash_pmf(x, s=s, fit=npmle_res))
<span class="org-variable-name">g</span> = np.array(npmle_res.rx2(<span class="org-string">'fitted_g'</span>))
<span class="org-variable-name">a</span> = np.fmin(g[1], g[2])
<span class="org-variable-name">b</span> = np.fmax(g[1], g[2])
<span class="org-variable-name">comp_dens_conv</span> = np.array([((st.gamma(a=k + 1, scale=1 / s.values.reshape(-1, 1)).cdf(b.reshape(1, -1)) - st.gamma(a=k + 1, scale=1 / s.values.reshape(-1, 1)).cdf(a.reshape(1, -1))) / np.outer(s, b - a)).mean(axis=0) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
<span class="org-variable-name">pmf</span>[<span class="org-string">'NPMLE'</span>] = comp_dens_conv @ g[0]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4, 4)
ax[0].hist(x, bins=y, color=<span class="org-string">'k'</span>, label=<span class="org-constant">None</span>)
<span class="org-keyword">for</span> c, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(plt.get_cmap(<span class="org-string">'Dark2'</span>).colors, [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>]):
  ax[0].plot(y + .5, x.shape[0] * pmf[k], color=c, lw=1, label=k)
ax[0].legend(frameon=<span class="org-constant">False</span>)
ax[0].set_title(gene_info.loc[gene, <span class="org-string">'name'</span>])
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)

<span class="org-keyword">for</span> c, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(plt.get_cmap(<span class="org-string">'Dark2'</span>).colors, [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>]):
  ax[1].plot(*<span class="org-builtin">getattr</span>(scmodes.deconvolve, f<span class="org-string">'fit_{k.lower()}'</span>)(x, s), color=c, lw=1, label=k)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
ax[1].legend(frameon=<span class="org-constant">False</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/cytotoxic_t-rpl34.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">grid</span> = np.linspace(0, 1, x.shape[0] + 1)[1:]
plt.xscale(<span class="org-string">'log'</span>)
plt.yscale(<span class="org-string">'log'</span>)
<span class="org-keyword">for</span> c, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(plt.get_cmap(<span class="org-string">'Dark2'</span>).colors, [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>, <span class="org-string">'ZIEF'</span>, <span class="org-string">'NPMLE'</span>]):
  <span class="org-keyword">if</span> k <span class="org-keyword">in</span> rpp:
    plt.plot(grid, np.sort(rpp[k]), color=c, lw=1, marker=<span class="org-constant">None</span>, label=k)
plt.plot([1e-5, 1], [1e-5, 1], lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'0.5'</span>)
plt.legend(frameon=<span class="org-constant">False</span>)
plt.xlim(1e-5, 1)
plt.ylim(1e-5, 1)
plt.xlabel(<span class="org-string">'Theoretical quantiles'</span>)
plt.ylabel(<span class="org-string">'Randomized quantiles'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/cytotoxic_t-rpl34-rpp.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<p>
Draw the histogram of GOF test \(p\)-values.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/mlcb.org/gamma-gof-hist.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<p>
Draw examples of multi-modal gene expression in iPSCs
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">ipsc_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/deconv-generalization/ipsc-gpu.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">lrt</span> = st.chi2(1).sf(-2 * (ipsc_res[<span class="org-string">'gamma'</span>] - ipsc_res[<span class="org-string">'point_gamma'</span>]))
<span class="org-variable-name">query</span> = <span class="org-builtin">list</span>(ipsc_res[lrt &lt; .05 / lrt.shape[0]].index)

<span class="org-comment-delimiter"># </span><span class="org-comment">SKP1: top eQTL</span>
query.append(<span class="org-string">'ENSG00000113558'</span>)
<span class="org-variable-name">ipsc_counts</span> = scmodes.dataset.ipsc(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>,
                                   query=query, return_df=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">annotation</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, index_col=0, header=<span class="org-constant">None</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">annotation</span> = annotation.loc[keep_samples.values.ravel()]
<span class="org-variable-name">s</span> = annotation[<span class="org-string">'mol_hs'</span>]
<span class="org-variable-name">donor_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/20130606_sample_info.txt'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">ipsc_log_mu</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">ipsc_log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">ipsc_logodds</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-logodds.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-keyword">del</span> ipsc_log_mu[<span class="org-string">'NA18498'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">point_gamma_cdf</span>(grid, log_mu, log_phi, logodds):
  <span class="org-variable-name">res</span> = st.gamma(a=np.exp(-log_phi), scale=np.exp(log_mu + log_phi)).cdf(grid)
  <span class="org-variable-name">res</span> *= sp.expit(-logodds)
  <span class="org-variable-name">res</span> += sp.expit(logodds)
  <span class="org-keyword">return</span> res

<span class="org-variable-name">lam</span> = ipsc_counts.values / s.values.reshape(-1, 1)
<span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000)
<span class="org-variable-name">cdfs</span> = np.array([[point_gamma_cdf(grid,
                                  ipsc_log_mu.loc[j, k],
                                  ipsc_log_phi.loc[j, k],
                                  ipsc_logodds.loc[j, k])
                  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> ipsc_log_mu]
                 <span class="org-keyword">for</span> j <span class="org-keyword">in</span> ipsc_counts.columns])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">colors</span> = {k: <span class="org-string">'r'</span> <span class="org-keyword">if</span> v == <span class="org-string">'male'</span> <span class="org-keyword">else</span> <span class="org-string">'b'</span> <span class="org-keyword">for</span> k, v <span class="org-keyword">in</span>
          donor_info.set_index(<span class="org-string">'Sample'</span>)
          .<span class="org-builtin">filter</span>(items=<span class="org-builtin">list</span>(ipsc_log_mu.columns), axis=0)
          [<span class="org-string">'Gender'</span>]
          .iteritems()}

</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">k</span> = <span class="org-string">'ENSG00000129824'</span>
<span class="org-variable-name">x</span> = ipsc_counts.loc[:,k].values
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4, 4)

ax[0].hist(x, bins=np.arange(x.<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_title(gene_info.loc[k, <span class="org-string">'name'</span>])
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)

<span class="org-keyword">for</span> f, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(cdfs[2], ipsc_log_mu.columns):
  ax[1].plot(grid, f, lw=1, alpha=0.2, c=colors.get(k, <span class="org-string">'k'</span>))
ax[1].set_xlim(0, 2e-4)
ax[1].set_xticks(np.linspace(0, 2e-4, 3))
<span class="org-variable-name">dummy</span> = [plt.Line2D([0], [0], c=<span class="org-string">'r'</span>), plt.Line2D([0], [0], c=<span class="org-string">'b'</span>)]
ax[1].legend(dummy, [<span class="org-string">'Male'</span>, <span class="org-string">'Female'</span>], title=<span class="org-string">'Sex'</span>, frameon=<span class="org-constant">False</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/ipsc-rps4y1.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<p>
Include a gene with an eQTL to illustrate how marginally "multi-modal" might
not be the correct question to ask.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean_qtl_stats</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/zinb/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/browser/browser.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">geno</span> = pd.read_sql(<span class="org-string">'select * from mean_qtl_geno where gene = ?'</span>, con=conn, params=(<span class="org-string">'ENSG00000113558'</span>,)).set_index(<span class="org-string">'ind'</span>)[<span class="org-string">'value'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
<span class="org-variable-name">k</span> = <span class="org-string">'ENSG00000113558'</span>
<span class="org-variable-name">x</span> = ipsc_counts.loc[:,k].values
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4, 4)

ax[0].hist(x, bins=np.arange(x.<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_title(gene_info.loc[k, <span class="org-string">'name'</span>])
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)

<span class="org-keyword">for</span> f, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(cdfs[1], ipsc_log_mu.columns):
  ax[1].plot(grid, f, lw=1, alpha=0.3, c=cm(<span class="org-builtin">int</span>(geno[k])))
<span class="org-variable-name">dummy</span> = [plt.Line2D([0], [0], c=cm(i)) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(3)]
<span class="org-comment-delimiter"># </span><span class="org-comment">https://www.ncbi.nlm.nih.gov/snp/?term=rs13356194</span>
ax[1].legend(dummy, [<span class="org-string">'CC'</span>, <span class="org-string">'CT'</span>, <span class="org-string">'TT'</span>], title=mean_qtl_stats.loc[<span class="org-string">'ENSG00000113558'</span>, <span class="org-string">'id'</span>].split(<span class="org-string">'.'</span>)[0], frameon=<span class="org-constant">False</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/ipsc-skp1.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<p>
Draw examples of bimodal genes detected in an in silico mixture.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">_mix_10x</span>(k1, k2, min_detect=0.01, return_y=<span class="org-constant">False</span>):
  <span class="org-variable-name">x1</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k1}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x2</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k2}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(x1, x2, min_detect=min_detect)
  <span class="org-keyword">if</span> return_y:
    <span class="org-keyword">return</span> x, y
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, **kwargs)

<span class="org-variable-name">mix_x</span>, <span class="org-variable-name">mix_y</span> = _cd8_cd19_mix(return_y=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sig0</span> = gof_res[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>].groupby(<span class="org-string">'dataset'</span>).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]]).reset_index(drop=<span class="org-constant">True</span>)
<span class="org-variable-name">nsig1</span> = gof_res[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>].groupby(<span class="org-string">'dataset'</span>).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &gt;= 0.05 / x.shape[0]]).reset_index(drop=<span class="org-constant">True</span>)
<span class="org-variable-name">examples</span> = sig0.merge(nsig1, on=[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>], suffixes=[<span class="org-string">'_gamma'</span>, <span class="org-string">'_zig'</span>])
<span class="org-keyword">del</span> examples[<span class="org-string">'method_gamma'</span>]
<span class="org-keyword">del</span> examples[<span class="org-string">'method_zig'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 2)
fig.set_size_inches(4, 4)
<span class="org-keyword">for</span> a, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), examples.sort_values(<span class="org-string">'p_gamma'</span>)[<span class="org-string">'gene'</span>].head(n=4)):
  <span class="org-keyword">for</span> i, (l, c) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'B cell'</span>, <span class="org-string">'T cell'</span>], [<span class="org-string">'r'</span>, <span class="org-string">'k'</span>])):
    <span class="org-variable-name">x</span> = mix_x.loc[mix_y == i,k]
    a.hist(x, bins=np.arange(x.<span class="org-builtin">max</span>() + 1), color=c, label=l, alpha=0.5)
    a.set_title(gene_info.loc[k, <span class="org-string">'name'</span>])
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a[0].set_ylabel(<span class="org-string">'Number of cells'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a[-1].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0,0].legend(frameon=<span class="org-constant">False</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/cytotoxic_t-b_cells-examples.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-orgc60948c" class="outline-2">
<h2 id="orgc60948c">Figure 2</h2>
<div class="outline-text-2" id="text-orgc60948c">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">data</span> = [<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, <span class="org-string">'ipsc'</span>, <span class="org-string">'cytotoxic_t-b_cells'</span>, <span class="org-string">'cytotoxic_t-naive_t'</span>, <span class="org-string">'pbmcs_68k'</span>]
<span class="org-variable-name">methods</span> = [<span class="org-string">'oracle'</span>, <span class="org-string">'ebpm_point'</span>, <span class="org-string">'wnmf'</span>, <span class="org-string">'wglmpca'</span>]
<span class="org-variable-name">imputation_res</span> = (pd.concat({(d, m): pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/imputation/{d}-{m}.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
                                 <span class="org-keyword">for</span> d <span class="org-keyword">in</span> data <span class="org-keyword">for</span> m <span class="org-keyword">in</span> methods <span class="org-keyword">if</span> d != <span class="org-string">'naive_t'</span>})
                  .reset_index(level=0)
                  .reset_index(drop=<span class="org-constant">True</span>)
                  .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1))

plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 3, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(5, 4)
<span class="org-keyword">for</span> a, d, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), data, [<span class="org-string">'T cells'</span>, <span class="org-string">'B cells'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>]):
  <span class="org-variable-name">g</span> = imputation_res[imputation_res[<span class="org-string">'dataset'</span>] == d]
  <span class="org-variable-name">base</span> = g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'ebpm_point'</span>]
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(methods[2:]):
    <span class="org-variable-name">y</span> = g.loc[g[<span class="org-string">'method'</span>] == m]
    a.scatter(x + np.random.normal(size=y.shape[0], scale=0.1),
              1e5 * (base[<span class="org-string">'loss'</span>].values - y[<span class="org-string">'loss'</span>].values), s=2, c=<span class="org-string">'k'</span>, zorder=4)
  a.axhline(y=0, c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">':'</span>)
  a.set_title(t)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax[-1]:
  a.set_xticks(np.arange(2))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> methods[2:]], rotation=90)
  a.set_xlabel(<span class="org-string">'Method'</span>)
<span class="org-variable-name">a</span> = fig.add_subplot(111, frameon=<span class="org-constant">False</span>)
a.tick_params(labelcolor=<span class="org-string">'none'</span>, top=<span class="org-constant">False</span>, bottom=<span class="org-constant">False</span>, left=<span class="org-constant">False</span>, right=<span class="org-constant">False</span>)
a.set_ylabel(<span class="org-string">'Improvement in log likelihood\nover noiseless rank one assumption'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="figure/figures.org/imputation-bench.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-02-09 Sun 23:00</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
