<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-11-03 Sun 23:16 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Goodness of fit of deconvolved distributions</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Goodness of fit of deconvolved distributions</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org283b08a">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org9458460">Methods</a>
<ul>
<li><a href="#orgfbb6acb">Test for goodness of fit</a></li>
<li><a href="#org1e959d4">Poisson-unimodal goodness of fit</a></li>
<li><a href="#orgfd5c235">Data</a></li>
</ul>
</li>
<li><a href="#org2a494f2">Results</a>
<ul>
<li><a href="#simulation">Simulation</a></li>
<li><a href="#gof">Test for goodness of fit</a></li>
<li><a href="#examples">Examples</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org283b08a" class="outline-2">
<h2 id="org283b08a">Introduction</h2>
<div class="outline-text-2" id="text-org283b08a">
<p>
We <a href="deconvolution.html">previously estimated</a> the out-of-sample log likelihood as a measure of the
generalization performance of distribution deconvolution methods for
scRNA-seq. However, this is not the simplest way to support our key results:
</p>

<ol class="org-ol">
<li>For spike-ins (control experiments), the data do not depart from a
unimodal distribution for biological variation (meaning, neither technical
variation nor biological variation involve a point mass on zero).</li>
<li>For most genes, in most data sets, a Gamma assumption appears sufficient
(meaning, the data do not significantly depart from the fitted model).</li>
</ol>

<p>
Here, we directly test for the goodness of fit of the estimated distribution
(as in <a href="https://dx.doi.org/10.1371/journal.pgen.1008045">Sarkar et al. 2019</a>).
</p>
</div>
</div>

<div id="outline-container-org4b66c4c" class="outline-2">
<h2 id="setup"><a id="org4b66c4c"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython" id="orgde1bf22"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scanpy
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> os
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scqtl
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org9458460" class="outline-2">
<h2 id="org9458460">Methods</h2>
<div class="outline-text-2" id="text-org9458460">
</div>
<div id="outline-container-orgfbb6acb" class="outline-3">
<h3 id="orgfbb6acb">Test for goodness of fit</h3>
<div class="outline-text-3" id="text-orgfbb6acb">
<p>
The key idea underlying our test for goodness of fit is the fact that if
\(x_{i} \sim F(\cdot)\), then \(F(x_{i}) \sim \operatorname{Uniform}(0,
  1)\). To test for goodness of fit of an estimated \(\hat{F}\) to the data
\(x_{1}, \ldots, x_{n}\), we apply the Kolmogorov-Smirnov (KS) test to test
whether the values \(\hat{F}(x_1), \ldots, \hat{F}(x_n)\) are uniformly
distributed. (This test is slightly conservative because it uses the same
data to estimate \(\hat{F}\).)
</p>

<p>
Here, we have to modify this simple procedure to account for the fact that
our data are discrete counts, so \(F\) is not continuous. To address this
issue, we used randomized quantiles (<a href="https://www.jstor.org/stable/1390802">Dunn 1996</a>): we sample one random value
per observation \(u_{i} \mid x_{i} \sim \mathrm{Uniform}(\hat{F}(x_i - 1),
  \hat{F}(x_i))\). These have the property that if \(x_i \sim F\) then \(u_i
  \sim \mathrm{Uniform}(0, 1)\).
</p>

<p>
In our model, each observed UMI count \(x_{i}\) comes from a different
distribution \(F_{i}\):
</p>

<p>
\[ F_i(x_i) = \sum_{k=0}^{x_i} \int_0^\infty \operatorname{Poisson}(k; s_i
  \lambda_i) g(d\lambda_i) \]
</p>

<p>
We therefore draw \(u_{i} \mid x_{i} \sim
  \mathrm{Uniform}(\hat{F}_{i}(x_{i} - 1), \hat{F}_{i}(x_{i}))\). Then, we
apply the KS test to whether the randomized quantiles \(u_{i}\) are uniformly
distributed.
</p>
</div>
</div>

<div id="outline-container-org1e959d4" class="outline-3">
<h3 id="org1e959d4">Poisson-unimodal goodness of fit</h3>
<div class="outline-text-3" id="text-org1e959d4">
<p>
We parameterized the Poisson-unimodal model as:
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(s_i \lambda_i)\\
  \lambda_i &\sim \sum_k w_k \operatorname{Uniform}(\lambda_0, a_k)
\end{align*}

<p>
where we abuse notation to allow \(a_k < \lambda_0\). To test for goodness
of fit, we need the PMF and CDF of \(x_i\), marginalized over
\(\lambda_i\). <a href="https://lsun.github.io/truncash/diagnostic_plot.html#ash:_(t)_likelihood,_uniform_mixture_prior">These
are analytic for certain choices of likelihood and prior</a>, e.g. normal
likelihood and mixture of uniforms prior.
</p>

<p>
Simulate a continuous example.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">bhat</span> = pd.Series(np.random.normal(loc=0.1, size=1000))
<span class="org-variable-name">fitn</span> = ashr.ash_workhorse(bhat, 1, mixcompdist=<span class="org-string">'uniform'</span>, output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
<span class="org-comment-delimiter"># </span><span class="org-comment">Important: Normal-uniform convolution CDF is analytic</span>
<span class="org-variable-name">F</span> = np.array(fitn.rx2(<span class="org-string">'fitted_g'</span>).rx2(<span class="org-string">'pi'</span>)).dot(np.array(ashr.comp_cdf_conv(fitn.rx2(<span class="org-string">'fitted_g'</span>), fitn.rx2(<span class="org-string">'data'</span>))))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">st.kstest(F, <span class="org-string">'uniform'</span>)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.03101662170773789, pvalue=0.28605255994768175)

</pre>

<p>
For the Poisson-unimodal model, we have (for one component):
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(s_i \lambda_i)\\
  \lambda_i &\sim \operatorname{Uniform}(a, b)
\end{align*}

<p>
and:
</p>

\begin{align*}
  \Pr(x_i = x) &= \int_0^\infty \operatorname{Poisson}(x_i; s_i \lambda_i) \operatorname{Uniform}(a, b) d\lambda_i\\
  &= \frac{1}{b - a} \int_a^b \operatorname{Poisson}(x_i; s_i \lambda_i) d\lambda_i\\
  &= \frac{1}{s_i (b - a)} \int_a^b \frac{s_i^{x_i + 1}}{\Gamma(x_i + 1)} \lambda_i^{(x_i + 1) - 1}\exp(-s_i\lambda_i) d\lambda_i\\
  &= \frac{1}{s_i (b - a)} \left( F_\Gamma(b; x_i + 1, s_i) - F_\Gamma(a, x_i + 1, s_i) \right)\\
  \Pr(x_i \leq x) &= \sum_{k=0}^{x} \frac{1}{s_i (b - a)} \left( F_\Gamma(b; k + 1, s_i) - F_\Gamma(a, k + 1, s_i) \right)\\
\end{align*}

<p>
where \(F_\Gamma(\cdot; \alpha, \beta)\) denotes the CDF of the Gamma
distribution with shape \(\alpha\) and rate \(\beta\). The marginal PMF is
analytic. Computing the CDF for each data point, for each component gives a
matrix of values \(\mathbf{F} = [F_{ik}]\). Then, the marginal CDF of the
data is given by \(\mathbf{F}\mathbf{w}\).
</p>

<p>
Simulate a simple discrete example.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">y</span> = pd.Series(np.random.poisson(lam=10, size=1000))
<span class="org-variable-name">s</span> = pd.Series(np.ones(y.shape))
<span class="org-variable-name">fitp</span> = ashr.ash_workhorse(
  np.zeros(y.shape), 1,
  lik=ashr.lik_pois(y=y, scale=s, link=<span class="org-string">'identity'</span>),
  mixcompdist=<span class="org-string">'halfuniform'</span>,
  output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
scmodes.benchmark.gof._gof(y,
  cdf=scmodes.benchmark.gof._ash_cdf,
  pmf=scmodes.benchmark.gof._ash_pmf, fit=fitp, s=s)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.018791616085501672, pvalue=0.8718125394494075)

</pre>

<p>
Simulate a Poisson example with varying size factors.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mu</span> = 1e-3
<span class="org-variable-name">s</span> = pd.Series(np.random.poisson(lam=1000, size=10000))
<span class="org-variable-name">y</span> = pd.Series(np.random.poisson(lam=s * mu))
<span class="org-variable-name">fitp</span> = ashr.ash_workhorse(
  np.zeros(y.shape), 1,
  lik=ashr.lik_pois(y=y, scale=s, link=<span class="org-string">'identity'</span>),
  mixcompdist=<span class="org-string">'halfuniform'</span>,
  output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
scmodes.benchmark.gof._gof(y,
  cdf=scmodes.benchmark.gof._ash_cdf,
  pmf=scmodes.benchmark.gof._ash_pmf, fit=fitp, s=s)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.008062709224521236, pvalue=0.533974089244596)

</pre>
</div>
</div>

<div id="outline-container-orgfd5c235" class="outline-3">
<h3 id="orgfd5c235">Data</h3>
<div class="outline-text-3" id="text-orgfd5c235">
<div class="org-src-container">
<pre class="src src-ipython" id="org5c09812"><span class="org-keyword">def</span> <span class="org-function-name">read_chromium</span>(sample):
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/svensson_chromium_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[x.obs[<span class="org-string">'sample'</span>] == sample]
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_dropseq</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/macosko_dropseq_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_indrops</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/klein_indrops_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_gemcode</span>():
  <span class="org-variable-name">x</span> = scanpy.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/zheng_gemcode_control.h5ad'</span>)
  scanpy.pp.filter_genes(x, min_counts=1)
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_c1</span>(min_detect=1000):
  <span class="org-variable-name">x</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/ercc-counts.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
  <span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, header=<span class="org-constant">None</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Throw out samples with too few spike-in molecules detected</span>
  <span class="org-keyword">return</span> x.loc[:,np.logical_and((x.<span class="org-builtin">sum</span>(axis=0) &gt; min_detect).values, keep_samples.values.ravel())].T

<span class="org-keyword">def</span> <span class="org-function-name">_mix_10x</span>(k1, k2, min_detect=0.05, return_y=<span class="org-constant">False</span>, p=500):
  <span class="org-variable-name">x1</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k1}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x2</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k2}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(x1, x2, min_detect=min_detect)
  <span class="org-variable-name">x</span> = x.sample(n=p, axis=1, random_state=1).reset_index(drop=<span class="org-constant">True</span>)
  <span class="org-keyword">if</span> return_y:
    <span class="org-keyword">return</span> x, y
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_cyto_naive_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_read_10x</span>(k):
  <span class="org-keyword">return</span> scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k}/filtered_matrices_mex/hg19/'</span>,
                                  return_df=<span class="org-constant">True</span>, min_detect=0.05, p=500)

<span class="org-variable-name">data</span> = {
  <span class="org-string">'dropseq'</span>: read_dropseq,
  <span class="org-string">'indrops'</span>: read_indrops,
  <span class="org-string">'chromium1'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20311'</span>),
  <span class="org-string">'chromium2'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20312'</span>),
  <span class="org-string">'gemcode'</span>: read_gemcode,
  <span class="org-string">'c1'</span>: read_c1,
  <span class="org-string">'cytotoxic_t'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'cytotoxic_t'</span>),
  <span class="org-string">'b_cells'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'b_cells'</span>),
  <span class="org-string">'ipsc'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>, p=500, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'cytotoxic_t-b_cells'</span>: _cd8_cd19_mix,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: _cyto_naive_mix,
  <span class="org-string">'pbmcs_68k'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'fresh_68k_pbmc_donor_a'</span>),
  <span class="org-comment-delimiter"># </span><span class="org-comment">Needed for Poisson-unimodal GOF</span>
  <span class="org-string">'pbmcs_68k_1'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'fresh_68k_pbmc_donor_a'</span>).iloc[:,:250],
  <span class="org-string">'pbmcs_68k_2'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'fresh_68k_pbmc_donor_a'</span>).iloc[:,250:],
}
</pre>
</div>

<p>
Report the dimensions of each data set. To reduce computational burden (for
<code>ashr</code>), look at no more than 500 genes per data set (randomly chosen).
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame([data[k]().shape <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data],
             columns=[<span class="org-string">'num_cells'</span>, <span class="org-string">'num_genes'</span>],
             index=data.keys())
</pre>
</div>

<pre class="example">
num_cells  num_genes
dropseq                     84         81
indrops                    953        103
chromium1                 2000         88
chromium2                 2000         88
gemcode                   1015         91
c1                        5002         92
cytotoxic_t              10209        500
b_cells                  10085        500
ipsc                      5597        500
cytotoxic_t-b_cells      20294        500
cytotoxic_t-naive_t      20688        500
pbmcs_68k                68579        500
</pre>
</div>
</div>
</div>

<div id="outline-container-org2a494f2" class="outline-2">
<h2 id="org2a494f2">Results</h2>
<div class="outline-text-2" id="text-org2a494f2">
</div>
<div id="outline-container-orgbc2ab18" class="outline-3">
<h3 id="simulation"><a id="orgbc2ab18"></a>Simulation</h3>
<div class="outline-text-3" id="text-simulation">
<p>
Draw data following the Poisson-point Gamma distribution:
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(s_i \lambda_i)\\
  \lambda_i &\sim \pi \delta_0(\cdot) + (1 - \pi) \operatorname{Gamma}(1/\phi, 1/(\mu\phi))
\end{align*}

<p>
Draw simulation parameters \(s_i=10^5, \ln\mu, \ln\phi\) from typical values
(Sarkar et al. 2019).
</p>

\begin{align*}
  \ln\mu &\sim \operatorname{Uniform}(-12, -8)\\
  \ln\phi &\sim \operatorname{Uniform}(-6, 0)
\end{align*}

<p>
Then, test for goodness of fit of Gamma, point-Gamma, and unimodal, each
convolved with Poisson.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> scmodes.benchmark.gof

<span class="org-keyword">def</span> <span class="org-function-name">trial</span>(num_samples, logodds, seed, verbose=<span class="org-constant">False</span>):
  <span class="org-variable-name">data</span>, <span class="org-variable-name">_</span> = scqtl.simulation.simulate(num_samples=num_samples, logodds=logodds, seed=seed)
  <span class="org-variable-name">x</span> = data[:,0]
  <span class="org-variable-name">s</span> = data[:,1]
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: scqtl.simple returns mu, 1/phi</span>
  <span class="org-variable-name">fit0</span> = scqtl.simple.fit_nb(x, s)
  <span class="org-variable-name">res0</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit0[0]), log_phi=-np.log(fit0[1]))
  <span class="org-variable-name">fit1</span> = scqtl.simple.fit_zinb(x, s)
  <span class="org-variable-name">res1</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit1[0]), log_phi=-np.log(fit1[1]), logodds=fit1[2])

  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: this returns the gene name as the first return value</span>
  <span class="org-variable-name">res2</span> = scmodes.benchmark.gof._gof_unimodal(<span class="org-string">'gene'</span>, pd.Series(x), pd.Series(s))
  <span class="org-keyword">return</span> res0[1], res1[1], res2[2]

<span class="org-keyword">def</span> <span class="org-function-name">evaluate</span>(num_samples, num_trials):
  <span class="org-variable-name">result</span> = []
  <span class="org-keyword">for</span> logodds <span class="org-keyword">in</span> np.linspace(-5, 0, 10):
    <span class="org-keyword">for</span> seed <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_trials):
      <span class="org-variable-name">gamma_res</span>, <span class="org-variable-name">zig_res</span>, <span class="org-variable-name">unimodal_res</span> = trial(num_samples, logodds, seed)
      result.append([logodds, seed, gamma_res, zig_res, unimodal_res])
  <span class="org-variable-name">result</span> = pd.DataFrame(result)
  <span class="org-variable-name">result.columns</span> = [<span class="org-string">'logodds'</span>, <span class="org-string">'trial'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>]
  <span class="org-keyword">return</span> result
</pre>
</div>

<p>
Run the simulation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_res</span> = evaluate(num_samples=1000, num_trials=10)
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">sim_res.to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/simulation.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/simulation.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>, index_col=0)
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-keyword">for</span> i, (k, l) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>], [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>])):
  <span class="org-variable-name">jitter</span> = np.random.normal(scale=0.01, size=sim_res.shape[0])
  plt.scatter(sim_res[<span class="org-string">'logodds'</span>] + jitter, -np.log10(sim_res[k] + 1e-100), s=2, c=np.atleast_2d(cm(i)), label=l)
plt.axhline(y=0, c=<span class="org-string">'k'</span>, lw=1)
plt.axhline(y=-np.log10(.05), c=<span class="org-string">'0.8'</span>, ls=<span class="org-string">'--'</span>, lw=1)
plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4)
plt.xticks(np.linspace(-5, 0, 10), [f<span class="org-string">'{x:.2g}'</span> <span class="org-keyword">for</span> x <span class="org-keyword">in</span> sp.expit(np.linspace(-5, 0, 10))], rotation=90)
plt.xlabel(<span class="org-string">'$\pi$'</span>)
plt.ylabel(<span class="org-string">'Goodness of fit $-\log_{10}(p)$'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/sim.png" alt="sim.png">
</p>
</div>

<p>
Plot a zoomed-in version of the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-keyword">for</span> i, (k, l) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>], [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>])):
  <span class="org-variable-name">jitter</span> = np.random.normal(scale=0.01, size=sim_res.shape[0])
  plt.scatter(sim_res[<span class="org-string">'logodds'</span>] + jitter, -np.log10(sim_res[k] + 1e-100), s=2, c=np.atleast_2d(cm(i)), label=l)
plt.axhline(y=0, c=<span class="org-string">'k'</span>, lw=1)
plt.axhline(y=-np.log10(.05), c=<span class="org-string">'0.8'</span>, ls=<span class="org-string">'--'</span>, lw=1)
plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4)
plt.xticks(np.linspace(-5, 0, 10), [f<span class="org-string">'{x:.2g}'</span> <span class="org-keyword">for</span> x <span class="org-keyword">in</span> sp.expit(np.linspace(-5, 0, 10))], rotation=90)
plt.xlabel(<span class="org-string">'$\pi$'</span>)
plt.ylim(0, 10)
plt.ylabel(<span class="org-string">'Goodness of fit $-\log_{10}(p)$'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/sim-inset.png" alt="sim-inset.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org9d1e8de" class="outline-3">
<h3 id="gof"><a id="org9d1e8de"></a>Test for goodness of fit</h3>
<div class="outline-text-3" id="text-gof">
</div>
<div id="outline-container-org88f2899" class="outline-4">
<h4 id="org88f2899">Run the test</h4>
<div class="outline-text-4" id="text-org88f2899">
<p>
Run the GPU-based methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --job-name=gof --time=12:00:00 -a 6-11
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
import os
&lt;&lt;data&gt;&gt;
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[task]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_gof(x, methods=['gamma', 'zig'])</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-gpu.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Run the CPU methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -n1 -c28 --exclusive --job-name=gof --time=24:00:00 -a 12-13
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
import multiprocessing as mp
import os
&lt;&lt;data&gt;&gt;
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">with mp.Pool(maxtasksperchild=20) as pool:</span>
<span class="org-sh-heredoc">  # Important: this needs to be done after initializing the pool to avoid</span>
<span class="org-sh-heredoc">  # memory duplication</span>
<span class="org-sh-heredoc">  x = data[task]()</span>
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_gof(x, pool=pool, methods=['unimodal'])</span>
<span class="org-sh-heredoc">  res.index = x.columns</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-unimodal.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh">rsync -au /scratch/midway2/aksarkar/modes/gof/ /project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = []
<span class="org-keyword">for</span> m <span class="org-keyword">in</span> [<span class="org-string">'gpu'</span>, <span class="org-string">'unimodal'</span>]:
  <span class="org-variable-name">res</span> = <span class="org-builtin">dict</span>()
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data:
    <span class="org-variable-name">f</span> = f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/{k}-{m}.txt.gz'</span>
    <span class="org-keyword">if</span> os.path.exists(f):
      <span class="org-comment-delimiter"># </span><span class="org-comment">Hacks</span>
      <span class="org-keyword">if</span> k <span class="org-keyword">in</span> (<span class="org-string">'pbmcs_68k_1'</span>, <span class="org-string">'pbmcs_68k_2'</span>):
        <span class="org-variable-name">k</span> = <span class="org-string">'pbmcs_68k'</span>
      <span class="org-variable-name">res</span>[k] = pd.read_csv(f, sep=<span class="org-string">'\t'</span>, index_col=0)
      <span class="org-keyword">if</span> <span class="org-string">'gene.1'</span> <span class="org-keyword">in</span> res[k].columns:
        <span class="org-keyword">del</span> res[k][<span class="org-string">'gene.1'</span>]
      <span class="org-keyword">if</span> m == <span class="org-string">'unimodal'</span>:
        <span class="org-variable-name">res</span>[k] = res[k].reset_index(drop=(res[k].index.name != <span class="org-string">'gene'</span>))
  gof_res.append(pd.concat(res, sort=<span class="org-constant">True</span>)
                 .reset_index(level=0)
                 .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1))
<span class="org-variable-name">gof_res</span> = pd.concat(gof_res).reset_index(drop=<span class="org-constant">True</span>)
</pre>
</div>

<p>
Write out the post-processed results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">gof_res.to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<p>
Read the post-processed results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>
</div>
</div>

<div id="outline-container-orga840f27" class="outline-4">
<h4 id="orga840f27">Control data sets</h4>
<div class="outline-text-4" id="text-orga840f27">
<p>
Report the number of genes which depart from the null that the data for the
gene follows the fitted distribution (after Bonferroni correction at level
0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">control</span> = <span class="org-builtin">list</span>(data.keys())[:5]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">(gof_res[gof_res[<span class="org-string">'dataset'</span>].isin(control)]
 .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).<span class="org-builtin">sum</span>())
 .reset_index()
 .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))
</pre>
</div>

<pre class="example">
0
method    gamma unimodal zig
dataset
chromium1     3        0   3
chromium2     0        0   0
dropseq       0        0   0
gemcode      20        0  21
indrops       0        0   0
</pre>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted Gamma
distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-gamma-gof-hist.png" alt="control-gamma-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted point-Gamma
distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-zig-gof-hist.png" alt="control-zig-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted unimodal
distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'unimodal'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-unimodal-gof-hist.png" alt="control-unimodal-gof-hist.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orge4150fe" class="outline-4">
<h4 id="orge4150fe">Biological data sets</h4>
<div class="outline-text-4" id="text-orge4150fe">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">non_control</span> = <span class="org-builtin">list</span>(data.keys())[6:12]
</pre>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted Gamma
distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/gamma-gof-hist.png" alt="gamma-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted
point-Gamma distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/zig-gof-hist.png" alt="zig-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted
point-Gamma distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'unimodal'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/unimodal-gof-hist.png" alt="unimodal-gof-hist.png">
</p>
</div>

<p>
Report the number of genes which depart from the null that the data for the
gene follows the fitted distribution (after Bonferroni correction at level
0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython">(gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(<span class="org-builtin">list</span>(data.keys())[6:])]
 .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).<span class="org-builtin">sum</span>())
 .reset_index()
 .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))
</pre>
</div>

<pre class="example">
0
method              gamma unimodal zig
dataset
b_cells                 4        1   3
cytotoxic_t             3        0   1
cytotoxic_t-b_cells    11        9  11
cytotoxic_t-naive_t     3        4   4
ipsc                   22        2  23
pbmcs_68k              25        7  19
</pre>

<p>
Report the fraction of genes which depart from the null that the data for
the gene follows the fitted distribution (after Bonferroni correction at
level 0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython">(gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(<span class="org-builtin">list</span>(data.keys())[6:])]
 .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).mean())
 .reset_index()
 .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))
</pre>
</div>

<pre class="example">
0
method               gamma unimodal    zig
dataset
b_cells              0.008    0.002  0.006
cytotoxic_t          0.006    0.000  0.002
cytotoxic_t-b_cells  0.022    0.018  0.022
cytotoxic_t-naive_t  0.006    0.008  0.008
ipsc                 0.044    0.004  0.046
pbmcs_68k            0.050    0.028  0.038
</pre>

<p>
<b>Remark</b> We previously fit point-Gamma distributions to the iPSC data
(Sarkar et al. 2019), and <a href="https://jdblischak.github.io/singlecell-qtl/zinb.html#org17af587">reported</a>:
</p>

<blockquote>
<p>
We tested the goodness of fit for each individual and each gene, and
rejected the null that the model fit the data for only 60 of 537,658
individual-gene combinations (0.01%) after Bonferroni correction (\(p < 9
    \times 10^{−8}\))
</p>
</blockquote>

<p>
The key difference between the approach taken here and the previous result
is that here we do not account for the fact that the cells are derived from
multiple donors. The differing genetic backgrounds of the donor individuals
means both the mean and variance of gene expression vary, such that the
marginal distribution is not well-described by a single Gamma distribution.
</p>
</div>
</div>
</div>

<div id="outline-container-org8d99ad6" class="outline-3">
<h3 id="examples"><a id="org8d99ad6"></a>Examples</h3>
<div class="outline-text-3" id="text-examples">
<p>
Look at genes which significantly depart from Gamma, but not from
point-Gamma.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sig0</span> = gof_res[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>].groupby(<span class="org-string">'dataset'</span>).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]]).reset_index(drop=<span class="org-constant">True</span>)
<span class="org-variable-name">nsig1</span> = gof_res[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>].groupby(<span class="org-string">'dataset'</span>).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &gt;= 0.05 / x.shape[0]]).reset_index(drop=<span class="org-constant">True</span>)
<span class="org-variable-name">examples</span> = sig0.merge(nsig1, on=[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>], suffixes=[<span class="org-string">'_gamma'</span>, <span class="org-string">'_zig'</span>])
<span class="org-keyword">del</span> examples[<span class="org-string">'method_gamma'</span>]
<span class="org-keyword">del</span> examples[<span class="org-string">'method_zig'</span>]
examples.sort_values(<span class="org-string">'p_gamma'</span>).head()
</pre>
</div>

<pre class="example">
dataset             gene  stat_gamma        p_gamma  stat_zig  \
6   cytotoxic_t-b_cells  ENSG00000204287    0.181818   0.000000e+00  0.011718
11  cytotoxic_t-b_cells  ENSG00000231389    0.095940  1.125366e-162  0.010051
17  cytotoxic_t-b_cells  ENSG00000105369    0.093518  1.383684e-154  0.014097
16  cytotoxic_t-b_cells  ENSG00000007312    0.072012   7.766159e-92  0.006933
7   cytotoxic_t-b_cells  ENSG00000198502    0.066750   5.775193e-79  0.008084

p_zig
6   0.007598
11  0.033140
17  0.000628
16  0.283539
7   0.140896
</pre>

<p>
The top examples come from the T cell/B cell mixture.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mix_x</span>, <span class="org-variable-name">mix_y</span> = _cd8_cd19_mix(return_y=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 2)
fig.set_size_inches(4, 4)
<span class="org-keyword">for</span> a, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), examples.sort_values(<span class="org-string">'p_gamma'</span>)[<span class="org-string">'gene'</span>].head(n=4)):
  <span class="org-keyword">for</span> i, (l, c) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'B cell'</span>, <span class="org-string">'T cell'</span>], [<span class="org-string">'r'</span>, <span class="org-string">'k'</span>])):
    <span class="org-variable-name">x</span> = mix_x.loc[mix_y == i,k]
    a.hist(x, bins=np.arange(x.<span class="org-builtin">max</span>() + 1), color=c, label=l, alpha=0.5)
    a.set_title(gene_info.loc[k, <span class="org-string">'name'</span>])
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a[0].set_ylabel(<span class="org-string">'Number of cells'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a[-1].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0,0].legend(frameon=<span class="org-constant">False</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/examples.png" alt="examples.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-11-03 Sun 23:16</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
