<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-09-24 Tue 12:22 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Goodness of fit of deconvolved distributions</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Goodness of fit of deconvolved distributions</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org283b08a">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org9458460">Methods</a>
<ul>
<li><a href="#org2274d9e">Test for goodness of fit</a></li>
<li><a href="#orgfd5c235">Data</a></li>
</ul>
</li>
<li><a href="#org2a494f2">Results</a>
<ul>
<li><a href="#simulation">Simulation</a></li>
<li><a href="#gof">Test for goodness of fit</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org283b08a" class="outline-2">
<h2 id="org283b08a">Introduction</h2>
<div class="outline-text-2" id="text-org283b08a">
<p>
We <a href="deconvolution.html">previously estimated</a> the out-of-sample log likelihood as a measure of the
generalization performance of distribution deconvolution methods for
scRNA-seq. However, this is not the simplest way to support some of our key
results:
</p>

<ol class="org-ol">
<li>For most genes, in most data sets, a Gamma assumption appears sufficient
(meaning, more complicated distributional assumptions do not have better
generalization performance).</li>
<li>When Gamma is insufficient, it is generally not the case that assuming a
point mass on zero improves generalization (meaning, departures from Gamma
do not involve excess zeros).</li>
</ol>

<p>
Here, we directly test for the goodness of fit of the estimated distribution
(as in <a href="https://dx.doi.org/10.1371/journal.pgen.1008045">Sarkar et al. 2019</a>).
</p>
</div>
</div>

<div id="outline-container-org4b66c4c" class="outline-2">
<h2 id="setup"><a id="org4b66c4c"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython" id="orgde1bf22"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scqtl
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org9458460" class="outline-2">
<h2 id="org9458460">Methods</h2>
<div class="outline-text-2" id="text-org9458460">
</div>
<div id="outline-container-org2274d9e" class="outline-3">
<h3 id="org2274d9e">Test for goodness of fit</h3>
<div class="outline-text-3" id="text-org2274d9e">
<p>
The key idea underlying our test for goodness of fit is the fact that if
\(x_{i} \sim F(\cdot)\), then \(F(x_{i}) \sim \operatorname{Uniform}(0,
  1)\). To test for goodness of fit of an estimated \(\hat{F}\) to the data
\(x_{1}, \ldots, x_{n}\), we apply the Kolmogorov-Smirnov (KS) test to test
whether the values \(\hat{F}(x_1), \ldots, \hat{F}(x_n)\) are uniformly
distributed. (This test is slightly conservative because it uses the same
data to estimate \(\hat{F}\).)
</p>

<p>
Here, we have to modify this simple procedure to account for the fact that
our data are discrete counts, so \(F\) is not continuous. To address this
issue, we used randomized quantiles (<a href="https://www.jstor.org/stable/1390802">Dunn 1996</a>): we sample one random value
per observation \(u_{i} \mid x_{i} \sim \mathrm{Uniform}(\hat{F}(x_i - 1),
  \hat{F}(x_i))\). These have the property that if \(x_i \sim F\) then \(u_i
  \sim \mathrm{Uniform}(0, 1)\).
</p>

<p>
In our model, each observed UMI count \(x_{i}\) comes from a different
distribution \(F_{i}\):
</p>

<p>
\[ F_i(x_i) = \sum_{k=0}^{x_i} \int_0^\infty \operatorname{Poisson}(k; s_i
  \lambda_i) g(d\lambda_i) \]
</p>

<p>
We therefore draw \(u_{i} \mid x_{i} \sim
  \mathrm{Uniform}(\hat{F}_{i}(x_{i} - 1), \hat{F}_{i}(x_{i}))\). Then, we
apply the KS test to whether the randomized quantiles \(u_{i}\) are uniformly
distributed.
</p>
</div>
</div>

<div id="outline-container-orgfd5c235" class="outline-3">
<h3 id="orgfd5c235">Data</h3>
<div class="outline-text-3" id="text-orgfd5c235">
<div class="org-src-container">
<pre class="src src-ipython" id="org5c09812"><span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>():
  <span class="org-variable-name">t_cells</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">b_cells</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/b_cells/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-keyword">return</span> scmodes.dataset.synthetic_mix(t_cells, b_cells, min_detect=0.25)[0]

<span class="org-keyword">def</span> <span class="org-function-name">_cyto_naive_mix</span>():
  <span class="org-variable-name">cyto</span> = scmodes.dataset.read_10x(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">naive</span> = scmodes.dataset.read_10x(prefix=<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/naive_t/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-keyword">return</span> scmodes.dataset.synthetic_mix(cyto, naive, min_detect=0.25)[0]

<span class="org-variable-name">data</span> = {
  <span class="org-string">'cytotoxic_t'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/cytotoxic_t/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'b_cells'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/b_cells/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'ipsc'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>, n=100, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'cytotoxic_t-b_cells'</span>: _cd8_cd19_mix,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: _cyto_naive_mix,
  <span class="org-string">'pbmcs_68k'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/fresh_68k_pbmc_donor_a/filtered_matrices_mex/hg19'</span>, return_df=<span class="org-constant">True</span>),
}
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2a494f2" class="outline-2">
<h2 id="org2a494f2">Results</h2>
<div class="outline-text-2" id="text-org2a494f2">
</div>
<div id="outline-container-orgbc2ab18" class="outline-3">
<h3 id="simulation"><a id="orgbc2ab18"></a>Simulation</h3>
<div class="outline-text-3" id="text-simulation">
<p>
Draw data following the Poisson-point Gamma distribution:
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(s_i \lambda_i)\\
  \lambda_i &\sim \pi \delta_0(\cdot) + (1 - \pi) \operatorname{Gamma}(1/\phi, 1/(\mu\phi))
\end{align*}

<p>
Draw simulation parameters \(s_i=10^5, \ln\mu, \ln\phi\) from typical values
(Sarkar et al. 2019).
</p>

\begin{align*}
  \ln\mu &\sim \operatorname{Uniform}(-12, -8)\\
  \ln\phi &\sim \operatorname{Uniform}(-6, 0)
\end{align*}

<p>
Then, test for goodness of fit of Poisson-Gamma and Poisson-point Gamma.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> scmodes.benchmark.gof

<span class="org-keyword">def</span> <span class="org-function-name">trial</span>(num_samples, logodds, seed, verbose=<span class="org-constant">False</span>):
  <span class="org-variable-name">data</span>, <span class="org-variable-name">_</span> = scqtl.simulation.simulate(num_samples=num_samples, logodds=logodds, seed=seed)
  <span class="org-variable-name">x</span> = data[:,0]
  <span class="org-variable-name">s</span> = data[:,1]
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: scqtl.simple returns mu, 1/phi</span>
  <span class="org-variable-name">fit0</span> = scqtl.simple.fit_nb(x, s)
  <span class="org-variable-name">res0</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit0[0]), log_phi=-np.log(fit0[1]))
  <span class="org-variable-name">fit1</span> = scqtl.simple.fit_zinb(x, s)
  <span class="org-variable-name">res1</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit1[0]), log_phi=-np.log(fit1[1]), logodds=fit1[2])
  <span class="org-keyword">if</span> verbose:
    <span class="org-keyword">print</span>(fit0, fit1)
  <span class="org-keyword">return</span> res0[1], res1[1]

<span class="org-keyword">def</span> <span class="org-function-name">evaluate</span>(num_samples, num_trials):
  <span class="org-variable-name">result</span> = []
  <span class="org-keyword">for</span> logodds <span class="org-keyword">in</span> np.linspace(-5, 0, 10):
    <span class="org-keyword">for</span> seed <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_trials):
      <span class="org-variable-name">gamma_res</span>, <span class="org-variable-name">zig_res</span> = trial(num_samples, logodds, seed)
      result.append([logodds, seed, gamma_res, zig_res])
  <span class="org-variable-name">result</span> = pd.DataFrame(result)
  <span class="org-variable-name">result.columns</span> = [<span class="org-string">'logodds'</span>, <span class="org-string">'trial'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>]
  <span class="org-keyword">return</span> result
</pre>
</div>

<p>
Run the simulation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_res</span> = evaluate(num_samples=1000, num_trials=10)
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-keyword">for</span> i, (k, l) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>], [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>])):
  <span class="org-variable-name">jitter</span> = np.random.normal(scale=0.01, size=sim_res.shape[0])
  plt.scatter(sim_res[<span class="org-string">'logodds'</span>] + jitter, -np.log10(sim_res[k]), s=2, c=np.atleast_2d(cm(i)), label=l)
plt.axhline(y=0, c=<span class="org-string">'k'</span>, lw=1)
plt.axhline(y=-np.log10(.05 / sim_res.shape[0]), c=<span class="org-string">'0.8'</span>, ls=<span class="org-string">':'</span>, lw=1)
plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4)
plt.xticks(np.linspace(-5, 0, 10), [f<span class="org-string">'{x:.2g}'</span> <span class="org-keyword">for</span> x <span class="org-keyword">in</span> sp.expit(np.linspace(-5, 0, 10))], rotation=90)
plt.xlabel(<span class="org-string">'$\pi$'</span>)
plt.ylabel(<span class="org-string">'Goodness of fit $-\log_{10}(p)$'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/sim.png" alt="sim.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org32b9e57" class="outline-3">
<h3 id="gof"><a id="org32b9e57"></a>Test for goodness of fit</h3>
<div class="outline-text-3" id="text-gof">
<p>
Run the GPU-based methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --time=4:00:00 -a 1-5
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
import os
&lt;&lt;data&gt;&gt;
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[task]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_gof(x, methods=['gamma', 'zig'])</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-gpu.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Run the CPU methods
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -n1 -c28 --exclusive --time=12:00:00 -a 0
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
import multiprocessing as mp
import os
&lt;&lt;data&gt;&gt;
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">with mp.Pool(maxtasksperchild=20) as pool:</span>
<span class="org-sh-heredoc">  # Important: this needs to be done after initializing the pool to avoid</span>
<span class="org-sh-heredoc">  # memory duplication</span>
<span class="org-sh-heredoc">  x = data[task]()</span>
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_gof(x, pool=pool, methods=['unimodal'])</span>
<span class="org-sh-heredoc">  res.index = x.columns</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-unimodal.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh">rsync -au /scratch/midway2/aksarkar/modes/gof/ /project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = (pd.concat([pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/{k}-gpu.txt.gz'</span>,
                                 sep=<span class="org-string">'\t'</span>, index_col=0) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data],
                     keys=data.keys())
           .reset_index(level=0)
           .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=<span class="org-string">'columns'</span>))
</pre>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted Gamma
distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 3, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(6, 4)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=10, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a[-1].set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/gamma-gof-hist.png" alt="gamma-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted
point-Gamma distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 3, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(6, 4)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>],
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>, <span class="org-string">'p'</span>], bins=10, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a[-1].set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/zig-gof-hist.png" alt="zig-gof-hist.png">
</p>
</div>

<p>
Report the number of genes which depart from the null that the data for the
gene follows the fitted distribution (after Bonferroni correction at level
0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython">gof_res.groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>]).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).<span class="org-builtin">sum</span>()).reset_index().pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>)
</pre>
</div>

<pre class="example">
0
method              gamma  zig
dataset
b_cells                12   10
cytotoxic_t             7    6
cytotoxic_t-b_cells    35   24
cytotoxic_t-naive_t    29   28
ipsc                    2    2
pbmcs_68k             108  109
</pre>

<p>
Report the fraction of genes which depart from the null that the data for
the gene follows the fitted distribution (after Bonferroni correction at
level 0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython">gof_res.groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>]).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).mean()).reset_index().pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>)
</pre>
</div>

<pre class="example">
0
method                  gamma       zig
dataset
b_cells              0.032000  0.026667
cytotoxic_t          0.015184  0.013015
cytotoxic_t-b_cells  0.086634  0.059406
cytotoxic_t-naive_t  0.069212  0.066826
ipsc                 0.037736  0.037736
pbmcs_68k            0.267327  0.269802
</pre>

<p>
<b>Remark</b> We previously fit point-Gamma distributions to the iPSC data
(Sarkar et al. 2019), and <a href="https://jdblischak.github.io/singlecell-qtl/zinb.html#org17af587">reported</a>:
</p>

<blockquote>
<p>
We tested the goodness of fit for each individual and each gene, and
rejected the null that the model fit the data for only 60 of 537,658
individual-gene combinations (0.01%) after Bonferroni correction (\(p < 9
   \times 10^{âˆ’8}\))
</p>
</blockquote>

<p>
The key difference between the approach taken here and the previous result
is that here we do not account for the fact that the cells are derived from
multiple donors. The differing genetic backgrounds of the donor individuals
means both the mean and variance of gene expression vary, such that the
marginal distribution is not well-described by a single Gamma distribution.
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-09-24 Tue 12:22</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
