<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-07-25 Sat 12:06 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Goodness of fit of deconvolved distributions</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Goodness of fit of deconvolved distributions</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org283b08a">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org9458460">Methods</a>
<ul>
<li><a href="#org53f4e0a">Test for goodness of fit</a></li>
<li><a href="#org1e959d4">Poisson-unimodal goodness of fit</a></li>
<li><a href="#orgfd5c235">Data</a></li>
</ul>
</li>
<li><a href="#org2a494f2">Results</a>
<ul>
<li><a href="#simulation">Simulation</a></li>
<li><a href="#gof">Test for goodness of fit</a></li>
<li><a href="#org77d2646">Application to control data sets</a></li>
<li><a href="#orgf807e8b">Application to biological data sets</a></li>
<li><a href="#browser">Browser</a></li>
<li><a href="#examples">Real data examples</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org283b08a" class="outline-2">
<h2 id="org283b08a">Introduction</h2>
<div class="outline-text-2" id="text-org283b08a">
<p>
We <a href="deconvolution.html">previously estimated</a> the out-of-sample log likelihood as a measure of the
generalization performance of distribution deconvolution methods for
scRNA-seq. However, this is not the simplest way to support our key results:
</p>

<ol class="org-ol">
<li>For control experiments, the data do not significantly depart from a
unimodal distribution for expression variation (meaning, neither
sequencing variation nor expression variation involve a point mass on
zero).</li>
<li>For most genes, in most data sets, the data do not significantly depart
from a Gamma assumption.</li>
<li>When the data do significantly depart from a Gamma assumption, they often
do not depart from a unimodal assumption (meaning, departures from Gamma
are often not due to excess zeros).</li>
</ol>

<p>
Here, we directly test for the goodness of fit of the estimated distributions
(as in <a href="https://dx.doi.org/10.1371/journal.pgen.1008045">Sarkar et al. 2019</a>).
</p>
</div>
</div>

<div id="outline-container-org4b66c4c" class="outline-2">
<h2 id="setup"><a id="org4b66c4c"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython" id="orgde1bf22"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scanpy <span class="org-keyword">as</span> sc
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> os
<span class="org-keyword">import</span> rpy2.robjects
<span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> scipy.io <span class="org-keyword">as</span> si
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scqtl
<span class="org-keyword">import</span> sqlite3

<span class="org-variable-name">ashr</span> = rpy2.robjects.packages.importr(<span class="org-string">'ashr'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org9458460" class="outline-2">
<h2 id="org9458460">Methods</h2>
<div class="outline-text-2" id="text-org9458460">
</div>
<div id="outline-container-org53f4e0a" class="outline-3">
<h3 id="org53f4e0a">Test for goodness of fit</h3>
<div class="outline-text-3" id="text-org53f4e0a">
<p>
The key idea underlying our test for goodness of fit is the fact that if
\(x_{i} \sim F(\cdot)\), then \(F(x_{i}) \sim \operatorname{Uniform}(0,
  1)\). To test for goodness of fit of an estimated \(\hat{F}\) to the data
\(x_{1}, \ldots, x_{n}\), we apply the Kolmogorov-Smirnov (KS) test to test
whether the values \(\hat{F}(x_1), \ldots, \hat{F}(x_n)\) are uniformly
distributed. (This test is slightly conservative because it uses the same
data to estimate \(\hat{F}\).)
</p>

<p>
Here, we have to modify this simple procedure to account for the fact that
our data are discrete counts, so \(F\) is not continuous. To address this
issue, we used randomized quantiles (<a href="https://www.jstor.org/stable/1390802">Dunn 1996</a>): we sample one random value
per observation \(u_{i} \mid x_{i} \sim \mathrm{Uniform}(\hat{F}(x_i - 1),
  \hat{F}(x_i))\). These have the property that if \(x_i \sim F\) then \(u_i
  \sim \mathrm{Uniform}(0, 1)\).
</p>

<p>
In our model, each observed UMI count \(x_{i}\) comes from a different
distribution \(F_{i}\):
</p>

<p>
\[ F_i(x_i) = \sum_{k=0}^{x_i} \int_0^\infty \operatorname{Poisson}(k; x_i^+
  \lambda_i) g(d\lambda_i) \]
</p>

<p>
We therefore draw \(u_{i} \mid x_{i} \sim
  \mathrm{Uniform}(\hat{F}_{i}(x_{i} - 1), \hat{F}_{i}(x_{i}))\). Then, we
apply the KS test to whether the randomized quantiles \(u_{i}\) are uniformly
distributed.
</p>
</div>
</div>

<div id="outline-container-org1e959d4" class="outline-3">
<h3 id="org1e959d4">Poisson-unimodal goodness of fit</h3>
<div class="outline-text-3" id="text-org1e959d4">
<p>
We parameterized the Poisson-unimodal model as:
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(x_i^+ \lambda_i)\\
  \lambda_i &\sim \sum_k w_k \operatorname{Uniform}(\lambda_0, a_k)
\end{align*}

<p>
where we abuse notation to allow \(a_k < \lambda_0\). To test for goodness
of fit, we need the PMF and CDF of \(x_i\), marginalized over
\(\lambda_i\). <a href="https://lsun.github.io/truncash/diagnostic_plot.html#ash:_(t)_likelihood,_uniform_mixture_prior">These
are analytic for certain choices of likelihood and prior</a>, e.g. normal
likelihood and mixture of uniforms prior.
</p>

<p>
Simulate a continuous example.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">bhat</span> = pd.Series(np.random.normal(loc=0.1, size=1000))
<span class="org-variable-name">fitn</span> = ashr.ash_workhorse(bhat, 1, mixcompdist=<span class="org-string">'uniform'</span>, output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
<span class="org-comment-delimiter"># </span><span class="org-comment">Important: Normal-uniform convolution CDF is analytic</span>
<span class="org-variable-name">F</span> = np.array(fitn.rx2(<span class="org-string">'fitted_g'</span>).rx2(<span class="org-string">'pi'</span>)).dot(np.array(ashr.comp_cdf_conv(fitn.rx2(<span class="org-string">'fitted_g'</span>), fitn.rx2(<span class="org-string">'data'</span>))))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">st.kstest(F, <span class="org-string">'uniform'</span>)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.03101662170773789, pvalue=0.28605255994768175)

</pre>

<p>
For the Poisson-unimodal model, we have (for one component):
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(x_i^+ \lambda_i)\\
  \lambda_i &\sim \operatorname{Uniform}(a, b)
\end{align*}

<p>
and:
</p>

\begin{align*}
  \Pr(x_i = x) &= \int_0^\infty \operatorname{Poisson}(x; x_i^+ \lambda_i) \operatorname{Uniform}(a, b)\,d\lambda_i\\
  &= \frac{1}{b - a} \int_a^b \operatorname{Poisson}(x; x_i^+ \lambda_i)\,d\lambda_i\\
  &= \frac{1}{x_i^+ (b - a)} \int_a^b \frac{(x_i^+)^{x + 1}}{\Gamma(x + 1)} \lambda_i^{(x + 1) - 1}\exp(-x_i^+ \lambda_i)\,d\lambda_i\\
  &= \frac{1}{x_i^+ (b - a)} \left( F_\Gamma(b; x + 1, x_i^+) - F_\Gamma(a; x + 1, x_i^+) \right)\\
  \Pr(x_i \leq x) &= \sum_{k=0}^{x} \frac{1}{x_i^+ (b - a)} \left( F_\Gamma(b; k + 1, x_i^+) - F_\Gamma(a; k + 1, x_i^+) \right)
\end{align*}

<p>
where \(F_\Gamma(\cdot; \alpha, \beta)\) denotes the CDF of the Gamma
distribution with shape \(\alpha\) and rate \(\beta\). The marginal PMF is
analytic. Computing the CDF for each data point, for each component gives a
matrix of values \(\mathbf{F} = [F_{ik}]\). Then, the marginal CDF of the
data is given by \(\mathbf{F}\mathbf{w}\).
</p>

<p>
Simulate a simple discrete example.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">y</span> = pd.Series(np.random.poisson(lam=10, size=1000))
<span class="org-variable-name">s</span> = pd.Series(np.ones(y.shape))
<span class="org-variable-name">fitp</span> = ashr.ash_workhorse(
  np.zeros(y.shape), 1,
  lik=ashr.lik_pois(y=y, scale=s, link=<span class="org-string">'identity'</span>),
  mixcompdist=<span class="org-string">'halfuniform'</span>,
  output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
scmodes.benchmark.gof._gof(y,
  cdf=scmodes.benchmark.gof._ash_cdf,
  pmf=scmodes.benchmark.gof._ash_pmf, fit=fitp, s=s)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.018791616085501672, pvalue=0.8718125394494075)

</pre>

<p>
Simulate a Poisson example with varying size factors.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mu</span> = 1e-3
<span class="org-variable-name">s</span> = pd.Series(np.random.poisson(lam=1000, size=10000))
<span class="org-variable-name">y</span> = pd.Series(np.random.poisson(lam=s * mu))
<span class="org-variable-name">fitp</span> = ashr.ash_workhorse(
  np.zeros(y.shape), 1,
  lik=ashr.lik_pois(y=y, scale=s, link=<span class="org-string">'identity'</span>),
  mixcompdist=<span class="org-string">'halfuniform'</span>,
  output=pd.Series([<span class="org-string">'loglik'</span>, <span class="org-string">'fitted_g'</span>, <span class="org-string">'data'</span>]))
scmodes.benchmark.gof._gof(y,
  cdf=scmodes.benchmark.gof._ash_cdf,
  pmf=scmodes.benchmark.gof._ash_pmf, fit=fitp, s=s)
</pre>
</div>

<pre class="example">
KstestResult(statistic=0.008062709224521236, pvalue=0.533974089244596)

</pre>

<p>
An alternative parameterization is to assume the log link. For one component,
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(x_i^+ \lambda_i)\\
  \theta_i = \ln \lambda_i &\sim \operatorname{Uniform}(a, b)
\end{align*}

<p>
and
</p>

\begin{align*}
  \Pr(x_i = x) &= \frac{1}{b - a} \int_a^b \operatorname{Poisson}(x; \exp(\ln x_i^+ + \theta_i))\,d\theta_i\\
  &= \frac{1}{b - a} \int_a^b \frac{(x_i^+)^x}{\Gamma(x + 1)} \exp(\theta_i)^{x}\exp(-x_i^+ \exp(\theta_i))\,d\theta_i\\
  &= \frac{1}{x (b - a)} \int_{\exp(a)}^{\exp(b)} \frac{(x_i^+)^x}{\Gamma(x)} \lambda_i^{x - 1}\exp(-x_i^+ \lambda_i)\,d\lambda_i\\
  &= \frac{1}{x (b - a)} \left( F_\Gamma(\exp(b); x, x_i^+) - F_\Gamma(\exp(a); x, x_i^+) \right)\\
  \Pr(x_i \leq x) &= \sum_{k=0}^{x} \frac{1}{k (b - a)} \left( F_\Gamma(\exp(b); k, x_i^+) - F_\Gamma(\exp(a); k, x_i^+) \right)
\end{align*}
</div>
</div>

<div id="outline-container-orgfd5c235" class="outline-3">
<h3 id="orgfd5c235">Data</h3>
<div class="outline-text-3" id="text-orgfd5c235">
<div class="org-src-container">
<pre class="src src-ipython" id="org5c09812"><span class="org-keyword">def</span> <span class="org-function-name">read_chromium</span>(sample):
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/svensson_chromium_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[x.obs[<span class="org-string">'sample'</span>] == sample]
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_dropseq</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/macosko_dropseq_control.h5ad'</span>)
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_indrops</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/klein_indrops_control.h5ad'</span>)
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_gemcode</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/zheng_gemcode_control.h5ad'</span>)
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-keyword">return</span> pd.DataFrame(x.X.A, index=x.obs.index, columns=x.var.index)

<span class="org-keyword">def</span> <span class="org-function-name">read_c1</span>(min_detect=1000):
  <span class="org-variable-name">x</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/ercc-counts.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
  <span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, header=<span class="org-constant">None</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Throw out samples with too few spike-in molecules detected</span>
  <span class="org-keyword">return</span> x.loc[:,np.logical_and((x.<span class="org-builtin">sum</span>(axis=0) &gt; min_detect).values, keep_samples.values.ravel())].T

<span class="org-keyword">def</span> <span class="org-function-name">_mix_10x</span>(k1, k2, min_detect=0.01, return_y=<span class="org-constant">False</span>):
  <span class="org-variable-name">x1</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k1}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x2</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k2}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(x1, x2, min_detect=min_detect)
  <span class="org-keyword">if</span> return_y:
    <span class="org-keyword">return</span> x, y
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_cyto_naive_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_read_10x</span>(k, return_df=<span class="org-constant">True</span>, min_detect=0.01):
  <span class="org-keyword">return</span> scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k}/filtered_matrices_mex/hg19/'</span>,
                                  return_adata=<span class="org-keyword">not</span> return_df, return_df=return_df, min_detect=min_detect)

<span class="org-keyword">def</span> <span class="org-function-name">read_liver</span>():
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/human-cell-atlas/liver-caudate-lobe/liver-caudate-lobe.h5ad'</span>)
  <span class="org-variable-name">query</span> = x.obs[<span class="org-string">'donor'</span>] == <span class="org-string">'P3TLH'</span>
  <span class="org-variable-name">y</span> = x[query]
  sc.pp.filter_genes(y, min_cells=.01 * y.shape[0])
  <span class="org-keyword">return</span> y

<span class="org-keyword">def</span> <span class="org-function-name">read_kidney</span>():
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/modes/kidney.h5ad'</span>)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Supp Table 2+3 lists donor+biopsy information</span>
  <span class="org-variable-name">donor_pass</span> = <span class="org-builtin">set</span>([
    <span class="org-comment-delimiter"># </span><span class="org-comment">'4602STDY6949184',</span>
    <span class="org-string">'4602STDY6949185'</span>,
    <span class="org-comment-delimiter"># </span><span class="org-comment">'4602STDY6949187',</span>
    <span class="org-comment-delimiter"># </span><span class="org-comment">'4602STDY6949188',</span>
  ])
  <span class="org-variable-name">query</span> = np.array([t.split(<span class="org-string">'_'</span>)[0] <span class="org-keyword">in</span> donor_pass <span class="org-keyword">for</span> t <span class="org-keyword">in</span> x.obs.index])
  <span class="org-variable-name">y</span> = x[query]
  sc.pp.filter_genes(y, min_cells=.01 * y.shape[0])
  <span class="org-keyword">return</span> y

<span class="org-keyword">def</span> <span class="org-function-name">read_brain</span>():
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/gtex-droncseq/gtex-droncseq.h5ad'</span>)
  sc.pp.filter_genes(x, min_counts=.01 * x.shape[0])
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">read_retina</span>():
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/human-cell-atlas/adult-retina/adult-retina.h5ad'</span>)  
  <span class="org-variable-name">query</span> = x.obs[<span class="org-string">'donor_organism.provenance.document_id'</span>] == <span class="org-string">'427c0a62-9baf-42ab-a3a3-f48d10544280'</span>
  <span class="org-variable-name">y</span> = x[query]
  sc.pp.filter_genes(y, min_cells=.01 * y.shape[0])
  <span class="org-keyword">return</span> y

<span class="org-variable-name">data</span> = {
  <span class="org-string">'dropseq'</span>: read_dropseq,
  <span class="org-string">'indrops'</span>: read_indrops,
  <span class="org-string">'chromium1'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20311'</span>),
  <span class="org-string">'chromium2'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20312'</span>),
  <span class="org-string">'gemcode'</span>: read_gemcode,
  <span class="org-string">'c1'</span>: read_c1,
  <span class="org-string">'cytotoxic_t'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'cytotoxic_t'</span>),
  <span class="org-string">'b_cells'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'b_cells'</span>),
  <span class="org-string">'ipsc'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>, return_df=<span class="org-constant">True</span>),
  <span class="org-string">'cytotoxic_t-b_cells'</span>: _cd8_cd19_mix,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: _cyto_naive_mix,
  <span class="org-string">'pbmcs_68k'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'fresh_68k_pbmc_donor_a'</span>, return_df=<span class="org-constant">False</span>),
  <span class="org-string">'liver-caudate-lobe'</span>: <span class="org-keyword">lambda</span>: read_liver(),
  <span class="org-string">'kidney'</span>: <span class="org-keyword">lambda</span>: read_kidney(),
  <span class="org-string">'brain'</span>: <span class="org-keyword">lambda</span>: read_brain(),
  <span class="org-string">'retina'</span>: <span class="org-keyword">lambda</span>: read_retina()
}
</pre>
</div>

<p>
Report the dimensions of each data set.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame([data[k]().shape <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data],
             columns=[<span class="org-string">'num_cells'</span>, <span class="org-string">'num_genes'</span>],
             index=data.keys())
</pre>
</div>

<pre class="example">
num_cells  num_genes
dropseq                     84         81
indrops                    953        103
chromium1                 2000         88
chromium2                 2000         88
gemcode                   1015         91
c1                        5002         92
cytotoxic_t              10209       6530
b_cells                  10085       6417
ipsc                      5597       9957
cytotoxic_t-b_cells      20294       6647
cytotoxic_t-naive_t      20688       6246
pbmcs_68k                68579       6502
liver-caudate-lobe        3127      10646
kidney                    4876       8463
brain                    14963      11744
</pre>
</div>
</div>
</div>

<div id="outline-container-org2a494f2" class="outline-2">
<h2 id="org2a494f2">Results</h2>
<div class="outline-text-2" id="text-org2a494f2">
</div>
<div id="outline-container-orgbc2ab18" class="outline-3">
<h3 id="simulation"><a id="orgbc2ab18"></a>Simulation</h3>
<div class="outline-text-3" id="text-simulation">
<p>
Draw data following the Poisson-point Gamma distribution:
</p>

\begin{align*}
  x_i &\sim \operatorname{Poisson}(x_i^+ \lambda_i)\\
  \lambda_i &\sim \pi \delta_0(\cdot) + (1 - \pi) \operatorname{Gamma}(1/\phi, 1/(\mu\phi))
\end{align*}

<p>
Draw simulation parameters \(x_i^+=10^5, \ln\mu, \ln\phi\) from typical values
(Sarkar et al. 2019).
</p>

\begin{align*}
  \ln\mu &\sim \operatorname{Uniform}(-12, -8)\\
  \ln\phi &\sim \operatorname{Uniform}(-6, 0)
\end{align*}

<p>
Then, test for goodness of fit of Gamma, point-Gamma, and unimodal, each
convolved with Poisson.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> scmodes.benchmark.gof

<span class="org-keyword">def</span> <span class="org-function-name">trial</span>(num_samples, logodds, seed, verbose=<span class="org-constant">False</span>):
  <span class="org-variable-name">data</span>, <span class="org-variable-name">_</span> = scqtl.simulation.simulate(num_samples=num_samples, logodds=logodds, seed=seed)
  <span class="org-variable-name">x</span> = data[:,0]
  <span class="org-variable-name">s</span> = data[:,1]
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: scqtl.simple returns mu, 1/phi</span>
  <span class="org-variable-name">fit0</span> = scqtl.simple.fit_nb(x, s)
  <span class="org-variable-name">res0</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit0[0]), log_phi=-np.log(fit0[1]))
  <span class="org-variable-name">fit1</span> = scqtl.simple.fit_zinb(x, s)
  <span class="org-variable-name">res1</span> = scmodes.benchmark.gof._gof(x, cdf=scmodes.benchmark.gof._zig_cdf,
                                    pmf=scmodes.benchmark.gof._zig_pmf, size=s,
                                    log_mu=np.log(fit1[0]), log_phi=-np.log(fit1[1]), logodds=fit1[2])

  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: this returns the gene name as the first return value</span>
  <span class="org-variable-name">res2</span> = scmodes.benchmark.gof._gof_unimodal(<span class="org-string">'gene'</span>, pd.Series(x), pd.Series(s))
  <span class="org-keyword">return</span> res0[1], res1[1], res2[2]

<span class="org-keyword">def</span> <span class="org-function-name">evaluate</span>(num_samples, num_trials):
  <span class="org-variable-name">result</span> = []
  <span class="org-keyword">for</span> logodds <span class="org-keyword">in</span> np.linspace(-5, 0, 10):
    <span class="org-keyword">for</span> seed <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_trials):
      <span class="org-variable-name">gamma_res</span>, <span class="org-variable-name">zig_res</span>, <span class="org-variable-name">unimodal_res</span> = trial(num_samples, logodds, seed)
      result.append([logodds, seed, gamma_res, zig_res, unimodal_res])
  <span class="org-variable-name">result</span> = pd.DataFrame(result)
  <span class="org-variable-name">result.columns</span> = [<span class="org-string">'logodds'</span>, <span class="org-string">'trial'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>]
  <span class="org-keyword">return</span> result
</pre>
</div>

<p>
Run the simulation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_res</span> = evaluate(num_samples=1000, num_trials=10)
</pre>
</div>

<p>
Write out the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">sim_res.to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/simulation.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/simulation.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>, index_col=0)
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-keyword">for</span> i, (k, l) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>], [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>])):
  <span class="org-variable-name">jitter</span> = np.random.normal(scale=0.01, size=sim_res.shape[0])
  plt.scatter(sim_res[<span class="org-string">'logodds'</span>] + jitter, -np.log10(sim_res[k] + 1e-100), s=2, c=np.atleast_2d(cm(i)), label=l)
plt.axhline(y=0, c=<span class="org-string">'k'</span>, lw=1)
plt.axhline(y=-np.log10(.05), c=<span class="org-string">'0.8'</span>, ls=<span class="org-string">'--'</span>, lw=1)
plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4)
plt.xticks(np.linspace(-5, 0, 10), [f<span class="org-string">'{x:.2g}'</span> <span class="org-keyword">for</span> x <span class="org-keyword">in</span> sp.expit(np.linspace(-5, 0, 10))], rotation=90)
plt.xlabel(<span class="org-string">'$\pi$'</span>)
plt.ylabel(<span class="org-string">'Goodness of fit $-\log_{10}(p)$'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/sim.png" alt="sim.png">
</p>
</div>

<p>
Plot a zoomed-in version of the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-keyword">for</span> i, (k, l) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([<span class="org-string">'gamma'</span>, <span class="org-string">'zig'</span>, <span class="org-string">'unimodal'</span>], [<span class="org-string">'Gamma'</span>, <span class="org-string">'ZIG'</span>, <span class="org-string">'Unimodal'</span>])):
  <span class="org-variable-name">jitter</span> = np.random.normal(scale=0.01, size=sim_res.shape[0])
  plt.scatter(sim_res[<span class="org-string">'logodds'</span>] + jitter, -np.log10(sim_res[k] + 1e-100), s=2, c=np.atleast_2d(cm(i)), label=l)
plt.axhline(y=0, c=<span class="org-string">'k'</span>, lw=1)
plt.axhline(y=-np.log10(.05), c=<span class="org-string">'0.8'</span>, ls=<span class="org-string">'--'</span>, lw=1)
plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4)
plt.xticks(np.linspace(-5, 0, 10), [f<span class="org-string">'{x:.2g}'</span> <span class="org-keyword">for</span> x <span class="org-keyword">in</span> sp.expit(np.linspace(-5, 0, 10))], rotation=90)
plt.xlabel(<span class="org-string">'$\pi$'</span>)
plt.ylim(0, 10)
plt.ylabel(<span class="org-string">'Goodness of fit $-\log_{10}(p)$'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/sim-inset.png" alt="sim-inset.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org9465c85" class="outline-3">
<h3 id="gof"><a id="org9465c85"></a>Test for goodness of fit</h3>
<div class="outline-text-3" id="text-gof">
<p>
Run the GPU-based methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --job-name=gof --time=1:00:00 -a 15
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[task]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_gof(x, methods=['gamma', 'zig'])</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-gpu.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh">rsync -au /scratch/midway2/aksarkar/modes/gof/ /project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = []
<span class="org-keyword">for</span> m <span class="org-keyword">in</span> [<span class="org-string">'point'</span>, <span class="org-string">'gpu'</span>, <span class="org-string">'unimodal'</span>]:
  <span class="org-variable-name">res</span> = <span class="org-builtin">dict</span>()
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data:
    <span class="org-variable-name">f</span> = f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/{k}-{m}.txt.gz'</span>
    <span class="org-keyword">if</span> os.path.exists(f):
      <span class="org-comment-delimiter"># </span><span class="org-comment">Hacks</span>
      <span class="org-variable-name">res</span>[k] = pd.read_csv(f, sep=<span class="org-string">'\t'</span>, index_col=0)
      <span class="org-keyword">if</span> <span class="org-string">'ID'</span> <span class="org-keyword">in</span> res[k].columns:
        <span class="org-variable-name">res</span>[k] = res[k].rename({<span class="org-string">'ID'</span>: <span class="org-string">'gene'</span>}, axis=1)
      <span class="org-keyword">if</span> res[k].index.name == <span class="org-string">'ID'</span>:
        <span class="org-keyword">del</span> res[k][<span class="org-string">'gene'</span>]
        res[k]<span class="org-variable-name">.index.name</span> = <span class="org-string">'gene'</span>
      <span class="org-keyword">if</span> <span class="org-string">'gene.1'</span> <span class="org-keyword">in</span> res[k].columns:
        <span class="org-keyword">del</span> res[k][<span class="org-string">'gene.1'</span>]
      <span class="org-keyword">if</span> m == <span class="org-string">'unimodal'</span>:
        <span class="org-variable-name">res</span>[k] = res[k].reset_index(drop=(res[k].index.name != <span class="org-string">'gene'</span>))
  gof_res.append(pd.concat(res, sort=<span class="org-constant">True</span>)
                 .reset_index(level=0)
                 .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1))
<span class="org-variable-name">gof_res</span> = pd.concat(gof_res).reset_index(drop=<span class="org-constant">True</span>)
</pre>
</div>

<p>
Write out the post-processed results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">gof_res.to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<p>
Read the post-processed results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gof_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/gof/gof.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<p>
To reduce computational burden, run <code>ashr</code> only on genes which
significantly departed from a Gamma assumption on expression variation in
each dataset. (This is implicitly assuming that <code>ashr</code> will fit data which
did not depart from a Gamma assumption, because the Gamma distribution is
unimodal; however, in practice this can sometimes fail due to a yet to be
tracked down bug.)
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sig</span> = (gof_res[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>]
       .groupby(<span class="org-string">'dataset'</span>)
       .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]]))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">for</span> k <span class="org-keyword">in</span> data:
  <span class="org-variable-name">x</span> = data[k]()
  <span class="org-keyword">if</span> <span class="org-keyword">not</span> <span class="org-builtin">isinstance</span>(x, anndata.AnnData):
    <span class="org-keyword">if</span> k <span class="org-keyword">in</span> sig.index:
      <span class="org-variable-name">s</span> = x.<span class="org-builtin">sum</span>(axis=1)
      <span class="org-comment-delimiter"># </span><span class="org-comment">Important: we need to keep around the size factors computed using all the</span>
      <span class="org-comment-delimiter"># </span><span class="org-comment">genes</span>
      <span class="org-variable-name">t</span> = anndata.AnnData(x.loc[:,sig.loc[k, <span class="org-string">'gene'</span>]].values, var=sig.loc[k, <span class="org-string">'gene'</span>].to_frame().set_index(<span class="org-string">'gene'</span>))
      <span class="org-variable-name">t.obs</span> = s.to_frame().rename({0: <span class="org-string">'size'</span>}, axis=1)
      t.write(f<span class="org-string">'/scratch/midway2/aksarkar/modes/unimodal-data/{k}.h5ad'</span>)
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">s</span> = x.X.tocsc().<span class="org-builtin">sum</span>(axis=1).A.ravel()
    <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = s
    <span class="org-variable-name">x.obs</span> = x.obs.rename({0: <span class="org-string">'barcode'</span>}, axis=1)
    <span class="org-keyword">if</span> x.var.shape[1] == 2:
      <span class="org-variable-name">x.var.columns</span> = [<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>]
    <span class="org-keyword">else</span>:
      <span class="org-variable-name">x.var</span> = x.var.rename({<span class="org-string">'featurekey'</span>: <span class="org-string">'gene'</span>, <span class="org-string">'featurename'</span>: <span class="org-string">'name'</span>}, axis=1)
    <span class="org-variable-name">x.var</span> = x.var.set_index(<span class="org-string">'gene'</span>)
    x[:,x.var.index.isin(sig.loc[k, <span class="org-string">'gene'</span>])].write(f<span class="org-string">'/scratch/midway2/aksarkar/modes/unimodal-data/{k}.h5ad'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -n1 -c28 --exclusive --job-name=gof --time=12:00:00 -a 10
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import anndata</span>
<span class="org-sh-heredoc">import multiprocessing as mp</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">import scipy.sparse as ss</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">with mp.Pool(maxtasksperchild=20) as pool:</span>
<span class="org-sh-heredoc">  # Important: this needs to be done after initializing the pool to avoid</span>
<span class="org-sh-heredoc">  # memory duplication</span>
<span class="org-sh-heredoc">  x = anndata.read_h5ad(f'/scratch/midway2/aksarkar/modes/unimodal-data/{task}.h5ad')</span>
<span class="org-sh-heredoc">  if ss.isspmatrix(x.X):</span>
<span class="org-sh-heredoc">    y = x.X.A</span>
<span class="org-sh-heredoc">  else:</span>
<span class="org-sh-heredoc">    y = x.X</span>
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_gof(pd.DataFrame(y, index=x.obs.index, columns=x.var.index), </span>
<span class="org-sh-heredoc">                                       s=x.obs['size'], </span>
<span class="org-sh-heredoc">                                       pool=pool, </span>
<span class="org-sh-heredoc">                                       methods=['unimodal'])</span>
<span class="org-sh-heredoc">  res.index = x.var.index</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-unimodal.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Run goodness of fit tests for a point mass assumption on control data sets
only.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --mem=4G --job-name=gof --time=1:00:00 -a 0-4
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">tasks = list(data.keys())</span>
<span class="org-sh-heredoc">task = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[task]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_gof(x, methods=['point'])</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/gof/{task}-point.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org77d2646" class="outline-3">
<h3 id="org77d2646">Application to control data sets</h3>
<div class="outline-text-3" id="text-org77d2646">
<p>
Report the number of genes which depart from the null that the data for the
gene follows the fitted distribution (after Bonferroni correction at level
0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">control</span> = <span class="org-builtin">list</span>(data.keys())[:5]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">(gof_res[gof_res[<span class="org-string">'dataset'</span>].isin(control)]
 .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]).<span class="org-builtin">sum</span>())
 .reset_index()
 .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))
</pre>
</div>

<pre class="example">
0
method    gamma point unimodal zig
dataset
chromium1     7     9        0   7
chromium2     6    14        0   6
dropseq       0    65        0   0
gemcode      22    28        0  22
indrops       0     2        0   0
</pre>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted Gamma
distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-gamma-gof-hist.png" alt="control-gamma-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted
point-Gamma distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax.T:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-zig-gof-hist.png" alt="control-zig-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted unimodal
distributions for each control dataset.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 5, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 2)
<span class="org-keyword">for</span> a, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                     gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'unimodal'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(k)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'$p$-value'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/control-unimodal-gof-hist.png" alt="control-unimodal-gof-hist.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgf807e8b" class="outline-3">
<h3 id="orgf807e8b">Application to biological data sets</h3>
<div class="outline-text-3" id="text-orgf807e8b">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">non_control</span> = <span class="org-builtin">list</span>(data.keys())[6:]
<span class="org-variable-name">labels</span> = [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/\nB cell'</span>, <span class="org-string">'Cytotoxic/\nnaive T'</span>,
          <span class="org-string">'PBMC'</span>, <span class="org-string">'Liver'</span>, <span class="org-string">'Kidney'</span>, <span class="org-string">'Brain'</span>, <span class="org-string">'Retina'</span>]
</pre>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted Gamma
distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 10, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        labels,
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'gamma'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-variable-name">a</span> = fig.add_subplot(111, xticks=[], yticks=[], frameon=<span class="org-constant">False</span>)
a.set_xlabel(<span class="org-string">'$p$-value'</span>, labelpad=16)
fig.tight_layout(w_pad=0)
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/gamma-gof-hist.png" alt="gamma-gof-hist.png">
</p>
</div>

<p>
Plot the histogram of goodness of fit test \(p\)-values for the fitted
point-Gamma distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 10, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, t, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(),
                        labels,
                        gof_res.groupby(<span class="org-string">'dataset'</span>)):
  a.hist(g.loc[g[<span class="org-string">'method'</span>] == <span class="org-string">'zig'</span>, <span class="org-string">'p'</span>], bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
  a.axhline(y=1, c=<span class="org-string">'k'</span>, ls=<span class="org-string">':'</span>, lw=1)
  a.set_xlim([0, 1])
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
<span class="org-variable-name">a</span> = fig.add_subplot(111, xticks=[], yticks=[], frameon=<span class="org-constant">False</span>)
a.set_xlabel(<span class="org-string">'$p$-value'</span>, labelpad=16)
fig.tight_layout(w_pad=0)
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/zig-gof-hist.png" alt="zig-gof-hist.png">
</p>
</div>

<p>
Report the number of genes which depart from the fitted distribution (after
Bonferroni correction at level 0.05). Note the unimodal assumption was
tested only for genes which significantly depart from Gamma.
</p>

<div class="org-src-container">
<pre class="src src-ipython">(gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(non_control)]
 .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &lt;= 0.05 / x.shape[0]).<span class="org-builtin">sum</span>())
 .reset_index()
 .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))[0]
</pre>
</div>

<pre class="example">
method               gamma  unimodal   zig
dataset
b_cells                 53        10    45
brain                  399        63   400
cytotoxic_t             35         0    34
cytotoxic_t-b_cells   1077        19  1070
cytotoxic_t-naive_t    945        20   951
ipsc                  2251         6  2244
kidney                3487      1815  2934
liver-caudate-lobe    5365      4796  4559
pbmcs_68k              422        22   409
retina                 226       133   202
</pre>

<p>
Report the fraction of genes which <i>do not</i> depart from the null that the
data for the gene follows the fitted distribution (after Bonferroni
correction at level 0.05).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">frac</span> = (gof_res.loc[gof_res[<span class="org-string">'dataset'</span>].isin(non_control)]
        .groupby([<span class="org-string">'dataset'</span>, <span class="org-string">'method'</span>])
        .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'p'</span>] &gt;= 0.05 / x.shape[0]).mean())
        .reset_index()
        .pivot(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>))[0]
frac
</pre>
</div>

<pre class="example">
method                  gamma  unimodal       zig
dataset
b_cells              0.991741  0.811321  0.992987
brain                0.966025  0.842105  0.965940
cytotoxic_t          0.994640  1.000000  0.994793
cytotoxic_t-b_cells  0.837972  0.982358  0.839025
cytotoxic_t-naive_t  0.848703  0.978836  0.847743
ipsc                 0.773928  0.997335  0.774631
kidney               0.587971  0.479495  0.653314
liver-caudate-lobe   0.496055  0.106058  0.571764
pbmcs_68k            0.935097  0.947867  0.937096
retina               0.977506  0.411504  0.979894
</pre>

<p>
<b>Remark</b> We previously fit point-Gamma distributions to the iPSC data
(Sarkar et al. 2019), and <a href="https://jdblischak.github.io/singlecell-qtl/zinb.html#org17af587">reported</a>:
</p>

<blockquote>
<p>
We tested the goodness of fit for each individual and each gene, and
rejected the null that the model fit the data for only 60 of 537,658
individual-gene combinations (0.01%) after Bonferroni correction (\(p < 9
    \times 10^{8}\))
</p>
</blockquote>

<p>
The key difference between the approach taken here and the previous result
is that here we do not account for the fact that the cells are derived from
multiple donors. The differing genetic backgrounds of the donor individuals
means both the mean and variance of gene expression vary, such that the
marginal distribution is not well-described by a single Gamma distribution.
</p>

<p>
Report the fraction of genes which depart from a Gamma assumption, but not
a point-Gamma assumption on expression variation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">query</span> = (gof_res[gof_res[<span class="org-string">'dataset'</span>].isin(non_control)]
         .pivot_table(index=[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>], columns=[<span class="org-string">'method'</span>], values=[<span class="org-string">'p'</span>, <span class="org-string">'stat'</span>]))
(query
 .groupby(level=0)
 .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: np.logical_and(x[<span class="org-string">'p'</span>][<span class="org-string">'gamma'</span>] &lt; .05 / x.shape[0],
                                 x[<span class="org-string">'p'</span>][<span class="org-string">'zig'</span>] &gt;= .05 / x.shape[0]).<span class="org-builtin">sum</span>()
        / (x[<span class="org-string">'p'</span>][<span class="org-string">'gamma'</span>] &lt; .05 / x.shape[0]).<span class="org-builtin">sum</span>()))
</pre>
</div>

<pre class="example">
dataset
b_cells                0.150943
brain                  0.082707
cytotoxic_t            0.057143
cytotoxic_t-b_cells    0.053853
cytotoxic_t-naive_t    0.069370
ipsc                   0.055531
kidney                 0.158876
liver-caudate-lobe     0.155079
pbmcs_68k              0.099526
retina                 0.230088
dtype: float64
</pre>
</div>
</div>

<div id="outline-container-org276689d" class="outline-3">
<h3 id="browser"><a id="org276689d"></a>Browser</h3>
<div class="outline-text-3" id="text-browser">
<p>
Read gene metadata.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<p>
Populate the database with the top 100 genes departing from each fitted
distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sig</span> = (gof_res
       .loc[gof_res[<span class="org-string">'method'</span>] == <span class="org-string">'unimodal'</span>]
       .groupby(<span class="org-string">'dataset'</span>)
       .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.loc[x[<span class="org-string">'p'</span>] &lt; 0.05 / x.shape[0]].sort_values(<span class="org-string">'p'</span>).head(n=100))
       .reset_index(drop=<span class="org-constant">True</span>)
       .merge(gene_info, on=<span class="org-string">'gene'</span>, how=<span class="org-string">'left'</span>)
       [[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'method'</span>, <span class="org-string">'stat'</span>, <span class="org-string">'p'</span>]])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">keys</span> = []
<span class="org-variable-name">counts</span> = []
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> sig.groupby([<span class="org-string">'dataset'</span>]):
  <span class="org-variable-name">x</span> = data[k]()
  <span class="org-variable-name">query</span> = g[<span class="org-string">'gene'</span>].unique()
  <span class="org-keyword">if</span> <span class="org-keyword">not</span> <span class="org-builtin">isinstance</span>(x, anndata.AnnData):
    counts.append(x[query])
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">key</span> = [t <span class="org-keyword">for</span> t <span class="org-keyword">in</span> (<span class="org-string">'gene'</span>, <span class="org-string">'ID'</span>, <span class="org-string">'featurekey'</span>, 0) <span class="org-keyword">if</span> t <span class="org-keyword">in</span> x.var.columns][0]
    counts.append(pd.DataFrame(x[:,x.var[key].isin(query)].X.A, index=x.obs.index, columns=query))
  keys.append(k)

<span class="org-keyword">def</span> <span class="org-function-name">_melt</span>(y):
  <span class="org-variable-name">y.index.name</span> = <span class="org-string">'sample'</span>
  <span class="org-keyword">return</span> (y
          .reset_index()
          .melt(id_vars=<span class="org-string">'sample'</span>, var_name=<span class="org-string">'gene'</span>, value_name=<span class="org-string">'count'</span>))

<span class="org-variable-name">count_data</span> = pd.concat([_melt(y) <span class="org-keyword">for</span> y <span class="org-keyword">in</span> counts], keys=keys).reset_index(level=0).rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/browser/browser.db'</span>) <span class="org-keyword">as</span> conn:
  sig.to_sql(<span class="org-string">'genes'</span>, conn, if_exists=<span class="org-string">'replace'</span>, index=<span class="org-constant">False</span>)
  conn.execute(<span class="org-string">'create index genes_idx on genes(dataset, gene);'</span>)
  count_data.to_sql(<span class="org-string">'counts'</span>, conn, if_exists=<span class="org-string">'replace'</span>, index=<span class="org-constant">False</span>)
  conn.execute(<span class="org-string">'create index counts_idx on counts(dataset, gene);'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-org2542000" class="outline-3">
<h3 id="examples"><a id="org2542000"></a>Real data examples</h3>
<div class="outline-text-3" id="text-examples">
<p>
Get examples of genes which significantly depart from the estimated unimodal
distribution, but are clearly unimodal upon inspection.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.read_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/modes/unimodal-data/b_cells.h5ad'</span>)
</pre>
</div>

<p>
Write out the data to RDS.
</p>

<div class="org-src-container">
<pre class="src src-ipython">rpy2.robjects.r[<span class="org-string">'saveRDS'</span>](
  pd.DataFrame(dat.X, columns=dat.var.index, index=dat.obs.index)
  .reset_index()
  .rename({<span class="org-string">'index'</span>: <span class="org-string">'sample'</span>}, axis=1)
  .melt(id_vars=<span class="org-string">'sample'</span>, var_name=<span class="org-string">'gene'</span>, value_name=<span class="org-string">'count'</span>)
  .merge(dat.obs, left_on=<span class="org-string">'sample'</span>, right_index=<span class="org-constant">True</span>)
  .sort_values([<span class="org-string">'gene'</span>, <span class="org-string">'sample'</span>])
  .reset_index(drop=<span class="org-constant">True</span>),
  <span class="org-string">'b-cell-data.Rds'</span>)
</pre>
</div>

<pre class="example">
rpy2.rinterface.NULL

</pre>

<p>
<i>RPS27</i> departs from Gamma, point-Gamma, and unimodal.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(6, 4)

ax[0].hist(dat.X[:,0], bins=np.arange(dat.X[:,0].<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)
ax[0].set_title(gene_info.loc[dat.var.index[0], <span class="org-string">'name'</span>])

<span class="org-keyword">for</span> i, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>((<span class="org-string">'gamma'</span>, <span class="org-string">'unimodal'</span>)):
  ax[1].plot(*<span class="org-builtin">getattr</span>(scmodes.deconvolve, f<span class="org-string">'fit_{m}'</span>)(dat.X[:,0], dat.obs[<span class="org-string">'size'</span>]), c=cm(i), label=m.capitalize())
ax[1].legend(frameon=<span class="org-constant">False</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954.png" alt="ENSG00000177954.png">
</p>
</div>

<p>
Look at the distribution of the MLE \(\hat\lambda_i = x_i / s_i\)
(intuitively, the quantity which we will be shrinking).
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.hist(dat.X[:,0] / dat.obs[<span class="org-string">'size'</span>], bins=50, color=<span class="org-string">'k'</span>)
plt.xlabel(<span class="org-string">'Relative expression'</span>)
plt.ylabel(<span class="org-string">'Number of cells'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954-naive.png" alt="ENSG00000177954-naive.png">
</p>
</div>

<p>
Look at the randomized quantiles for the fitted unimodal distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = scmodes.ebpm.ebpm_unimodal(dat.X[:,0], dat.obs[<span class="org-string">'size'</span>].values)
<span class="org-variable-name">F</span> = scmodes.benchmark.gof._ash_cdf(dat.X[:,0] - 1, fit=res, s=dat.obs[<span class="org-string">'size'</span>])
<span class="org-variable-name">f</span> = scmodes.benchmark.gof._ash_pmf(dat.X[:,0], fit=res, s=dat.obs[<span class="org-string">'size'</span>])
np.random.seed(0)
<span class="org-variable-name">rpp</span> = F + np.random.uniform(size=F.shape[0]) * f
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.hist(rpp, bins=np.linspace(0, 1, 11), color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
plt.axhline(y=1, ls=<span class="org-string">':'</span>, lw=1, c=<span class="org-string">'k'</span>)
plt.xlim(0, 1)
plt.xlabel(<span class="org-string">'$p$-value'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954-rpp.png" alt="ENSG00000177954-rpp.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.xscale(<span class="org-string">'log'</span>)
plt.yscale(<span class="org-string">'log'</span>)
plt.scatter(np.linspace(0, 1, rpp.shape[0] + 2)[1:-1], np.sort(rpp), s=1, c=<span class="org-string">'k'</span>)
plt.plot([1e-5, 1], [1e-5, 1], lw=1, ls=<span class="org-string">':'</span>, c=<span class="org-string">'r'</span>)
plt.xlabel(<span class="org-string">'Theoretical quantile'</span>)
plt.ylabel(<span class="org-string">'Randomized quantile'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954-diagnostic.png" alt="ENSG00000177954-diagnostic.png">
</p>
</div>

<p>
Fit models with finer grids.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> step <span class="org-keyword">in</span> np.log(np.linspace(1, 2, 11)[1:]):
  <span class="org-variable-name">res</span>[step] = scmodes.ebpm.ebpm_unimodal(
    dat.X[:,0],
    dat.obs[<span class="org-string">'size'</span>],
    mixsd=np.exp(np.arange(np.log(1e-8), np.log(lam.<span class="org-builtin">max</span>()), step=step)))
</pre>
</div>

<p>
Report the log likelihood of the fitted models against the (log) stepsize of the grid.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.Series({k: np.array(res[k].rx2(<span class="org-string">'loglik'</span>))[0] <span class="org-keyword">for</span> k <span class="org-keyword">in</span> res})
</pre>
</div>

<pre class="example">
0.095310   -27916.693078
0.182322   -27916.691828
0.262364   -27916.847090
0.336472   -27917.009390
0.405465   -27916.711257
0.470004   -27917.364408
0.530628   -27917.541988
0.587787   -27917.067727
0.641854   -27917.669822
0.693147   -27918.161042
dtype: float64
</pre>

<p>
Test the fitted models for goodness of fit, and report the KS test statistic
and \(p\)-value.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
pd.Series({
  k: scmodes.benchmark.gof._gof(
    dat.X[:,0],
    cdf=scmodes.benchmark.gof._ash_cdf,
    pmf=scmodes.benchmark.gof._ash_pmf,
    s=dat.obs[<span class="org-string">'size'</span>],
    fit=res[k])
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> res})
</pre>
</div>

<pre class="example">
0.095310     (0.03343639916233043, 3.219041220233783e-10)
0.182322     (0.03349651214950944, 2.968120165176161e-10)
0.262364     (0.03281679008032623, 7.367539493197478e-10)
0.336472      (0.03312928499093126, 4.86187521991659e-10)
0.405465     (0.03260696496636967, 9.717889023225659e-10)
0.470004    (0.03257735474891127, 1.0103672529713676e-09)
0.530628    (0.033042770023911816, 5.456976614561023e-10)
0.587787     (0.03396250448855004, 1.574404486410344e-10)
0.641854    (0.03380000822864859, 1.9659431226291915e-10)
0.693147    (0.034306982158272215, 9.797409099464328e-11)
dtype: object
</pre>

<p>
Look at the fitted distributions.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(4, 2)
<span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 1000)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> res:
  <span class="org-variable-name">F</span> = ashr.cdf_ash(res[k], grid)
  plt.plot(grid, np.array(F.rx2(<span class="org-string">'y'</span>)).ravel(), c=colorcet.cm[<span class="org-string">'fire'</span>](k), lw=1)
plt.xlim(.005, .015)
plt.xlabel(<span class="org-string">'Latent gene expression'</span>)
plt.ylabel(<span class="org-string">'CDF'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954-cdf.png" alt="ENSG00000177954-cdf.png">
</p>
</div>

<p>
Overlay the estimated density with the empirical density of the naive MLE.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">grid</span> = np.linspace(lam.<span class="org-builtin">min</span>(), lam.<span class="org-builtin">max</span>(), 5000)
<span class="org-variable-name">F</span> = ashr.cdf_ash(res[k], grid)
<span class="org-variable-name">f</span> = np.diff(np.array(F.rx2(<span class="org-string">'y'</span>)).ravel()) / np.diff(grid)

plt.clf()
plt.gcf().set_size_inches(4, 3)
plt.plot(grid[:-1], np.log(f + 1e-8), c=<span class="org-string">'r'</span>, lw=1, label=<span class="org-string">'Shrunk'</span>)
<span class="org-variable-name">h</span> = np.histogram(dat.X[:,0] / dat.obs[<span class="org-string">'size'</span>], density=<span class="org-constant">True</span>, bins=50)
plt.plot(h[1][:-1], np.log(h[0] + 1e-8), c=<span class="org-string">'k'</span>, lw=1, label=<span class="org-string">'Empirical'</span>)
plt.legend(frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
plt.xlabel(<span class="org-string">'Relative expression'</span>)
plt.ylabel(<span class="org-string">'Log density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000177954-fitted-g.png" alt="ENSG00000177954-fitted-g.png">
</p>
</div>

<p>
<i>CXCR4</i> departs from Gamma, but not from point-Gamma.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(6, 4)

ax[0].hist(dat.X[:,1], bins=np.arange(dat.X[:,1].<span class="org-builtin">max</span>() + 1), color=<span class="org-string">'k'</span>)
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Number of cells'</span>)
ax[0].set_title(gene_info.loc[dat.var.index[1], <span class="org-string">'name'</span>])

<span class="org-keyword">for</span> i, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>((<span class="org-string">'gamma'</span>, <span class="org-string">'unimodal'</span>)):
  ax[1].plot(*<span class="org-builtin">getattr</span>(scmodes.deconvolve, f<span class="org-string">'fit_{m}'</span>)(dat.X[:,1], dat.obs[<span class="org-string">'size'</span>]), c=cm(i), label=m.capitalize())
ax[1].legend(frameon=<span class="org-constant">False</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/gof.org/ENSG00000121966.png" alt="ENSG00000121966.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-07-25 Sat 12:06</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
