<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-08-31 Mon 17:14 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Clarifying confusion in scRNA-seq analysis</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Clarifying confusion in scRNA-seq analysis</h1>
<p>
The development of single cell RNA sequencing (scRNA-seq) technologies has
greatly facilitated investigation of important biological questions that were
previously difficult or impossible to study. However, how to model and analyze
scRNA-seq data has been the subject of considerable confusion and debate. We
develop a simple way of thinking and talking about scRNA-seq data that can
clarify much of this confusion. We outline a general framework which explicitly
models, and distinguishes between, the measurement process and variation in
actual expression levels between cells. Although our simple framework cannot on
its own address all issues that arise in scRNA-seq data, we show how it can
help clarify thinking, guide interpretation of existing analysis methods,
address questions of biological interest, and suggest directions for
development of new methods and analysis strategies.
</p>

<p>
Sarkar and Stephens,
<a href="https://dx.doi.org/10.1101/2020.04.07.030007">&ldquo;Separating measurement and
expression models clarifies confusion in single cell RNA-seq analysis.&rdquo;</a>
<i>BioRxiv</i> 2020.
</p>

<div id="outline-container-org7b1d935" class="outline-2">
<h2 id="org7b1d935">Empirical results</h2>
<div class="outline-text-2" id="text-org7b1d935">
<ul class="org-ul">
<li><a href="llr.html">Marginal log likelihood comparison of expression models</a></li>
<li><a href="ebnbm.html">Negative binomial measurement model</a></li>
<li><a href="lra.html">Low rank approximation</a></li>
<li><a href="fitted-values.html">Comparison of fitted values</a></li>
<li><a href="mlcb.html">MLCB figures</a></li>
<li><a href="figures.html">Manuscript figures</a></li>
</ul>
</div>
</div>

<div id="outline-container-org819fb0c" class="outline-2">
<h2 id="org819fb0c">Additional analyses</h2>
<div class="outline-text-2" id="text-org819fb0c">
<ul class="org-ul">
<li><a href="deconvolution.html">Comparison of distribution deconvolution methods</a></li>
<li><a href="deconv-examples.html">Distribution deconvolution examples</a></li>
<li><a href="gof.html">Goodness of fit of deconvolved distributions</a> (<a href="http://shiny.stephenslab.uchicago.edu:5009/gof">interactive browser</a>)</li>
<li><a href="marker-genes.html">Marker genes in <i>in silico</i> mixtures</a>
(<a href="http://shiny.stephenslab.uchicago.edu:5009/markers">interactive browser</a>)</li>
<li><a href="immune-census.html">Expression variation in Census of Immune Cells</a></li>
<li><a href="chromium3.html">Expression variation in 10X v3 PBMC data</a></li>
<li><a href="brain-dronc-seq.html">Expression variation in human brain cells</a></li>
<li><a href="retina.html">Expression variation in human retina cells</a></li>
<li><a href="liver.html">Expression variation in human liver</a></li>
<li><a href="imputation.html">Imputation of count matrices</a></li>
</ul>
</div>
</div>

<div id="outline-container-org012444b" class="outline-2">
<h2 id="org012444b">Technical details</h2>
<div class="outline-text-2" id="text-org012444b">
<ul class="org-ul">
<li><a href="pitfalls.html">Technical zero-generating mechanism in scRNA-seq data</a></li>
<li><a href="modes.html">Estimating the number of modes</a></li>
<li><a href="browser.html">Interactive simulations</a></li>
<li><a href="capture.html">Effect of capture rate on sampling variation</a></li>
<li><a href="ash-estmode.html">Speed up ash mode estimation</a></li>
<li><a href="mix-nb.html">Mixture of negative binomials</a></li>
<li><a href="gamma.html">Convergence of Gamma deconvolution</a></li>
<li><a href="poisson.html">Deconvolution of near-Poisson data</a></li>
<li><a href="validation.html">Validation set log likelihood comparison</a></li>
<li><a href="npmle.html">NPMLE on C1 spike-in data</a></li>
<li><a href="deconv-de.html">Differential expression from distribution deconvolution</a></li>
<li><a href="ep.html">Posterior approximation for scRNA-seq data</a></li>
<li><a href="link.html">Link functions in Poisson LRA</a></li>
<li><a href="random-effect.html">Distributional assumptions in Poisson LRA</a></li>
<li><a href="pois-truncnorm.html">Poisson-truncated normal data</a></li>
<li><a href="gamma-mix.html">Poisson-unimodal Gamma mixture model</a></li>
<li><a href="pois-gam-lik.html">Poisson-Gamma log likelihood surface</a></li>
<li><a href="vae.html">Variational autoencoders for scRNA-seq data</a></li>
<li><a href="wnbmf.html">Weighted Negative Binomial Matrix Factorization</a></li>
<li><a href="ash-pois-log-link.html">Log link in Poisson ash</a></li>
<li><a href="transforms.html">Transformations of deconvolved gene expression
distributions</a></li>
<li><a href="gaussian.html">Gaussian methods</a></li>
<li><a href="quantiles.html">Randomized quantiles</a></li>
<li><a href="independence.html">Relaxing the independence assumption on expression models</a></li>
<li><a href="zinb-em.html">EM algorithm for point-Gamma expression model</a></li>
<li><a href="npmle-grid.html">Iterative refinement of NPMLE grid in ashr</a></li>
<li><a href="halfuniform.html">Uniform vs half-uniform mixture prior</a></li>
</ul>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-08-31 Mon 17:14</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
