<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-09-07 Mon 14:23 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Marginal log likelihood comparison of expression models</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Marginal log likelihood comparison of expression models</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#results">Results</a>
<ul>
<li><a href="#data">Data</a></li>
<li><a href="#estimate-marginal-llik">Estimate marginal log likelihood</a></li>
<li><a href="#control-data">Application to control data sets</a></li>
<li><a href="#biological-data">Application to biological data sets</a></li>
<li><a href="#drop-seq">Drop-Seq spike-in gene</a></li>
<li><a href="#ash">Issue in ash</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org0f03d8f" class="outline-2">
<h2 id="intro"><a id="org0f03d8f"></a>Introduction</h2>
<div class="outline-text-2" id="text-intro">
<p>
We previously used a <a href="gof.html">goodness of fit test</a> and
<a href="deconvolution.html">out-of-sample (marginal) log likelihood</a> to assess
whether the data were adequately described by a given expression model. Here,
we compute the in-sample marginal log likelihood, in order to make direct
comparisons between expression models.
</p>
</div>
</div>

<div id="outline-container-orgb8f7cf6" class="outline-2">
<h2 id="setup"><a id="orgb8f7cf6"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> os
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scanpy <span class="org-keyword">as</span> sc
<span class="org-keyword">import</span> scipy.optimize <span class="org-keyword">as</span> so
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
rpy2.robjects.pandas2ri.activate()
<span class="org-variable-name">ashr</span> = rpy2.robjects.packages.importr(<span class="org-string">'ashr'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orge99efa2" class="outline-2">
<h2 id="results"><a id="orge99efa2"></a>Results</h2>
<div class="outline-text-2" id="text-results">
</div>
<div id="outline-container-org4e37009" class="outline-3">
<h3 id="data"><a id="org4e37009"></a>Data</h3>
<div class="outline-text-3" id="text-data">
<p>
Prepare the data in <code>h5ad</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orga462cce"><span class="org-keyword">import</span> functools <span class="org-keyword">as</span> ft

<span class="org-keyword">def</span> <span class="org-function-name">read_chromium</span>(sample):
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/svensson_chromium_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[x.obs[<span class="org-string">'sample'</span>] == sample,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x.var</span> = x.var.reset_index()
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">read_dropseq</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/macosko_dropseq_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x.var</span> = x.var.reset_index()
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">read_indrops</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/klein_indrops_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x.var</span> = x.var.reset_index()
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">read_gemcode</span>():
  <span class="org-variable-name">x</span> = sc.read(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/negative-controls/zheng_gemcode_control.h5ad'</span>)
  <span class="org-variable-name">x</span> = x[:,x.var.<span class="org-builtin">filter</span>(like=<span class="org-string">'ERCC'</span>, axis=<span class="org-string">'index'</span>).index]
  sc.pp.filter_genes(x, min_cells=1)
  <span class="org-variable-name">x.var</span> = x.var.reset_index()
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">_read_10x</span>(k, return_df=<span class="org-constant">False</span>, min_detect=0.01, chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k}/filtered_matrices_mex/hg19/'</span>,
                                  return_adata=<span class="org-keyword">not</span> return_df, return_df=return_df, min_detect=min_detect)
  <span class="org-variable-name">x.var.columns</span> = [<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>]
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">_mix_10x</span>(k1, k2, min_detect=0.01, return_y=<span class="org-constant">False</span>, chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x1</span> = _read_10x(k1, min_detect=0)
  <span class="org-variable-name">x2</span> = _read_10x(k2, min_detect=0)
  <span class="org-variable-name">x</span> = x1.concatenate(x2)
  sc.pp.filter_genes(x, min_cells=min_detect * x.shape[0])
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>(*args, **kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, *args, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_cyto_naive_mix</span>(*args, **kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>, *args, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">read_ipsc</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/ipsc/ipsc.h5ad'</span>)
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">read_liver</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/human-cell-atlas/liver-caudate-lobe/liver-caudate-lobe.h5ad'</span>)
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">read_kidney</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/human-cell-atlas/kidney/kidney.h5ad'</span>)
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">read_brain</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/gtex-droncseq/gtex-droncseq.h5ad'</span>)
  sc.pp.filter_genes(x, min_counts=.01 * x.shape[0])
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">read_retina</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/human-cell-atlas/adult-retina/adult-retina.h5ad'</span>)
  <span class="org-variable-name">query</span> = x.obs[<span class="org-string">'donor_organism.provenance.document_id'</span>] == <span class="org-string">'427c0a62-9baf-42ab-a3a3-f48d10544280'</span>
  <span class="org-variable-name">y</span> = x[query]
  sc.pp.filter_genes(y, min_cells=.01 * y.shape[0])
  <span class="org-variable-name">y.obs</span>[<span class="org-string">'size'</span>] = y.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> y
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> y[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-keyword">def</span> <span class="org-function-name">read_pbmc_10k_v3</span>(chunk=<span class="org-constant">None</span>, chunksize=500):
  <span class="org-variable-name">x</span> = anndata.read_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/modes/10k_pbmc_v3.h5ad'</span>)
  sc.pp.filter_genes(x, min_cells=0.01 * x.shape[0])
  <span class="org-variable-name">x.obs</span>[<span class="org-string">'size'</span>] = x.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
  <span class="org-keyword">if</span> chunk <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-keyword">return</span> x
  <span class="org-keyword">else</span>:
    <span class="org-keyword">return</span> x[:,chunk * chunksize:(chunk + 1) * chunksize]

<span class="org-variable-name">data</span> = {
  <span class="org-string">'dropseq'</span>: read_dropseq,
  <span class="org-string">'indrops'</span>: read_indrops,
  <span class="org-string">'chromium1'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20311'</span>),
  <span class="org-string">'chromium2'</span>: <span class="org-keyword">lambda</span>: read_chromium(<span class="org-string">'20312'</span>),
  <span class="org-string">'gemcode'</span>: read_gemcode,
  <span class="org-string">'cytotoxic_t'</span>: ft.partial(_read_10x, k=<span class="org-string">'cytotoxic_t'</span>),
  <span class="org-string">'b_cells'</span>: ft.partial(_read_10x, k=<span class="org-string">'b_cells'</span>),
  <span class="org-string">'ipsc'</span>: read_ipsc,
  <span class="org-string">'cytotoxic_t-b_cells'</span>: _cd8_cd19_mix,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: _cyto_naive_mix,
  <span class="org-string">'pbmc_10k_v3'</span>: read_pbmc_10k_v3,
  <span class="org-string">'liver-caudate-lobe'</span>: read_liver,
  <span class="org-string">'kidney'</span>: read_kidney,
  <span class="org-string">'brain'</span>: read_brain,
  <span class="org-string">'retina'</span>: read_retina,
}

<span class="org-variable-name">chunksize</span> = 500
<span class="org-variable-name">chunks</span> = {
  <span class="org-string">'b_cells'</span>: 6417 // chunksize,
  <span class="org-string">'brain'</span>: 11744 // chunksize,
  <span class="org-string">'cytotoxic_t'</span>: 6530 // chunksize,
  <span class="org-string">'cytotoxic_t-b_cells'</span>: 6647 // chunksize,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: 6246 // chunksize,
  <span class="org-string">'ipsc'</span>: 9957 // chunksize,
  <span class="org-string">'kidney'</span>: 15496 // chunksize,
  <span class="org-string">'liver-caudate-lobe'</span>: 3181 // chunksize,
  <span class="org-string">'pbmc_10k_v3'</span>: 12144 // chunksize,
  <span class="org-string">'retina'</span>: 10047 // chunksize,
}

<span class="org-variable-name">control</span> = <span class="org-builtin">list</span>(data.keys())[:5]
<span class="org-variable-name">non_control</span> = <span class="org-builtin">list</span>(data.keys())[5:]
</pre>
</div>

<p>
Report the dimensions of each data set.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame([data[k]().shape <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data],
             columns=[<span class="org-string">'num_samples'</span>, <span class="org-string">'num_genes'</span>],
             index=data.keys())
</pre>
</div>

<pre class="example">
num_samples  num_genes
dropseq                       84         81
indrops                      953        103
chromium1                   2000         88
chromium2                   2000         88
gemcode                     1015         91
cytotoxic_t                10209       6530
b_cells                    10085       6417
ipsc                        5597       9957
cytotoxic_t-b_cells        20294       6647
cytotoxic_t-naive_t        20688       6246
pbmc_10k_v3                11769      12144
liver-caudate-lobe          8856      16200
kidney                     11233      15496
brain                      14963      11744
retina                     21285      10047
</pre>
</div>
</div>

<div id="outline-container-orge0340e9" class="outline-3">
<h3 id="estimate-marginal-llik"><a id="orge0340e9"></a>Estimate marginal log likelihood</h3>
<div class="outline-text-3" id="text-estimate-marginal-llik">
<p>
Estimate the marginal likelihood for each data set, for each gene, for each
family of expression models.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org7e39abd">sbatch --partition=broadwl -n1 -c28 --exclusive --job-name=llik --time=30:00 -a 0-48
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import anndata</span>
<span class="org-sh-heredoc">import multiprocessing as mp</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">tasks = [(m, d) for m in ('point', 'gamma', 'point_gamma', 'unimodal', 'npmle') </span>
<span class="org-sh-heredoc">         for d in data</span>
<span class="org-sh-heredoc">         if not (m in ('unimodal', 'npmle') and d in chunks) and not (m == 'npmle' and d in control)]</span>
<span class="org-sh-heredoc">m, d = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">with mp.Pool() as pool:</span>
<span class="org-sh-heredoc">  x = data[d]()</span>
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_llik(x, pool=pool, methods=[m])</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/llik/{d}-{m}.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Fit non-parametric expression models serially for control data, to avoid
memory issues.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">for</span> d <span class="org-keyword">in</span> control:
  (scmodes.benchmark.evaluate_llik(data[d](), methods=[<span class="org-string">'npmle'</span>], max_grid_updates=20, tol=1e-5)
   .to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/llik/{d}-npmle.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>))
</pre>
</div>

<p>
Shard data sets to fit unimodal/non-parametric expression models within the
<code>midway2</code> time/memory limits.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org5dd3f6d">sbatch --partition=broadwl -n1 -c28 --exclusive --job-name=llik --time=4:00:00 -a 324-343
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import multiprocessing as mp</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">tasks = [(m, d, c)</span>
<span class="org-sh-heredoc">         for m in ('unimodal', 'npmle')</span>
<span class="org-sh-heredoc">         for d in chunks</span>
<span class="org-sh-heredoc">         for c in range(chunks[d])]</span>
<span class="org-sh-heredoc">m, d, c = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">kwargs = dict()</span>
<span class="org-sh-heredoc">if m == 'npmle':</span>
<span class="org-sh-heredoc">  kwargs = {'max_grid_updates': 20, 'tol': 1e-4}</span>
<span class="org-sh-heredoc">with mp.Pool() as pool:</span>
<span class="org-sh-heredoc">  x = data[d](chunk=c)</span>
<span class="org-sh-heredoc">  res = scmodes.benchmark.evaluate_llik(x, s=x.obs['size'], pool=pool, methods=[m], **kwargs)</span>
<span class="org-sh-heredoc">  res.to_csv(f'/scratch/midway2/aksarkar/modes/llik/{d}-{m}-{c}.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="orge2bfc78">rsync -au /scratch/midway2/aksarkar/modes/llik/ /project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/
</pre>
</div>

<p>
Combine the sharded results.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org0f8304f"><span class="org-keyword">for</span> m <span class="org-keyword">in</span> (<span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>):
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> chunks:
    <span class="org-keyword">if</span> os.path.exists(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/{k}-{m}-0.txt.gz'</span>):
      (pd.concat([pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/{k}-{m}-{i}.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
                  <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(chunks[k])])
       .to_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/{k}-{m}.txt.gz'</span>, sep=<span class="org-string">'\t'</span>))
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">llik</span> = (pd.concat(
  {
    k: pd.concat([
      pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/{k}-{m}.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>)
      <span class="org-keyword">for</span> m <span class="org-keyword">in</span> (<span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>)
      <span class="org-keyword">if</span> os.path.exists(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/llik/{k}-{m}.txt.gz'</span>)])
    <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data
  })
        .reset_index(level=0)
        .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1))
</pre>
</div>

<p>
Report the number (fraction) of genes that have likelihood ratio greater
than some threshold over some baseline. Ensure monotonicity post-hoc, to get
around an issue where the unimodal model sometimes
<a href="halfuniform.html#orgf28a0b8">does worse
than expected</a> due to not optimizing over the granularity of the grid
jointly with the mode.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">fix_llik</span>(llik):
  <span class="org-variable-name">llik</span>[<span class="org-string">'gamma'</span>] = np.fmax(llik[<span class="org-string">'gamma'</span>], llik[<span class="org-string">'point'</span>])
  <span class="org-variable-name">llik</span>[<span class="org-string">'point_gamma'</span>] = np.fmax(llik[<span class="org-string">'point_gamma'</span>], llik[<span class="org-string">'gamma'</span>])
  <span class="org-variable-name">llik</span>[<span class="org-string">'unimodal'</span>] = np.fmax(llik[<span class="org-string">'unimodal'</span>], llik[<span class="org-string">'gamma'</span>])
  <span class="org-variable-name">llik</span>[<span class="org-string">'npmle'</span>] = np.fmax.<span class="org-builtin">reduce</span>([llik[<span class="org-string">'npmle'</span>], llik[<span class="org-string">'unimodal'</span>], llik[<span class="org-string">'point_gamma'</span>]])
  <span class="org-keyword">return</span> llik

<span class="org-keyword">def</span> <span class="org-function-name">compare</span>(llik, baseline, thresh=10, agg=<span class="org-builtin">sum</span>):
  <span class="org-variable-name">labels</span> = [<span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>]
  <span class="org-keyword">assert</span> baseline <span class="org-keyword">in</span> labels
  <span class="org-variable-name">res</span> = <span class="org-builtin">dict</span>()
  <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik.groupby(<span class="org-string">'dataset'</span>):
    <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
    <span class="org-variable-name">query</span> = fix_llik(query)
    <span class="org-variable-name">res</span>[k] = (query.sub(query[baseline], axis=0) &gt; np.log(thresh)).agg(agg, axis=0)
  <span class="org-variable-name">res</span> = (pd.concat(res)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
         .pivot_table(index=<span class="org-string">'dataset'</span>, columns=<span class="org-string">'method'</span>)
         [0])
  <span class="org-keyword">del</span> res[baseline]
  <span class="org-keyword">return</span> res[[x <span class="org-keyword">for</span> x <span class="org-keyword">in</span> labels <span class="org-keyword">if</span> x != baseline]]
</pre>
</div>
</div>
</div>

<div id="outline-container-org2483eb2" class="outline-3">
<h3 id="control-data"><a id="org2483eb2"></a>Application to control data sets</h3>
<div class="outline-text-3" id="text-control-data">
<p>
For each control data set, for each expression model, report the fraction of
genes better fit by a Gamma expression model than a point mass expression
model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">compare(llik[llik[<span class="org-string">'dataset'</span>].isin(control)], baseline=<span class="org-string">'point'</span>)
</pre>
</div>

<pre class="example">
method     gamma  point_gamma  unimodal  npmle
dataset
chromium1     15           15        17     17
chromium2     48           48        51     52
dropseq       77           77        78     78
gemcode       52           52        56     56
indrops        4            4         4      4
</pre>

<p>
For each control data set, for each expression model, report the number of
genes which have improvement in marginal log likelihood above some threshold
over the Gamma expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">compare(llik[llik[<span class="org-string">'dataset'</span>].isin(control)], baseline=<span class="org-string">'gamma'</span>)
</pre>
</div>

<pre class="example">
method     point  point_gamma  unimodal  npmle
dataset
chromium1      0            0         8      8
chromium2      0            0         4      5
dropseq        0            2         6     20
gemcode        1            1        53     53
indrops        1            0         2      2
</pre>

<p>
Find the genes where a point-Gamma model improved over a Gamma model.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik[llik[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
  <span class="org-variable-name">temp</span>[k] = query.loc[query[<span class="org-string">'point_gamma'</span>] &gt; query[<span class="org-string">'gamma'</span>] + thresh]
<span class="org-variable-name">query</span> = (pd.concat(temp)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
         [[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>, <span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>]])
query
</pre>
</div>

<pre class="example">
method  dataset  dataset        gene        point        gamma  point_gamma  \
0       dropseq  dropseq  ERCC-00078  -765.076024  -279.353443  -264.522869
1       dropseq  dropseq  ERCC-00131  -723.520929  -356.755781  -341.742929
2       gemcode  gemcode  ERCC-00002 -6108.645264 -6148.879840 -6132.663007

method     unimodal        npmle
0       -263.565367  -261.356008
1       -347.518213  -337.359459
2      -5944.693604 -5944.434618
</pre>

<p>
For each control data set, for each expression model, report the number of
genes which have improvement in marginal log likelihood above some threshold
over the unimodal expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">compare(llik[llik[<span class="org-string">'dataset'</span>].isin(control)], baseline=<span class="org-string">'unimodal'</span>)
</pre>
</div>

<pre class="example">
method     point  gamma  point_gamma  npmle
dataset
chromium1      4      4            4      4
chromium2      7      7            8      8
dropseq        0      0            0      1
gemcode        0      0            0      0
indrops        0      0            0      0
</pre>

<p>
Look at the cases where NPMLE did better than unimodal.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik[llik[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
  <span class="org-variable-name">temp</span>[k] = query.loc[query[<span class="org-string">'npmle'</span>] &gt; query[<span class="org-string">'unimodal'</span>] + thresh]
<span class="org-variable-name">query</span> = (pd.concat(temp)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
         [[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>, <span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>]])
query
</pre>
</div>

<pre class="example">
method    dataset        gene       point       gamma  point_gamma  \
0       chromium1  ERCC-00024  -45.460132  -45.477957   -45.478005
1       chromium1  ERCC-00069 -140.401215 -139.854752  -139.944119
2       chromium1  ERCC-00104  -71.913572  -71.246536   -69.867679
3       chromium1  ERCC-00154 -272.657972 -273.072955  -273.094156
4       chromium2  ERCC-00016  -44.980123  -44.980123   -44.995680
5       chromium2  ERCC-00041  -45.187759  -45.204457   -45.204466
6       chromium2  ERCC-00073 -144.039409 -137.072157  -141.523956
7       chromium2  ERCC-00075  -15.306939  -15.306939   -15.308376
8       chromium2  ERCC-00086  -43.580107  -43.610757   -43.611190
9       chromium2  ERCC-00116 -179.364459 -177.229879  -174.857339
10      chromium2  ERCC-00142  -28.006585  -28.006585   -28.011618
11      chromium2  ERCC-00164  -55.455464  -53.541417   -48.993924
12        dropseq  ERCC-00131 -723.520929 -356.755781  -341.742929

method    unimodal       npmle
0      -121.644443  -45.841260
1      -238.635997 -139.705769
2      -123.375871  -72.716830
3      -331.171738 -272.658181
4      -173.512727  -44.997467
5      -116.825209  -45.188332
6      -198.975064 -136.364247
7      -186.398614  -15.307474
8      -107.533467  -43.580519
9      -186.885089 -174.791905
10     -148.213405  -43.356179
11      -80.198402  -56.292263
12     -347.518213 -337.359459
</pre>

<p>
Filter the cases where NPMLE did better than unimodal, retaining only genes
for which NPMLE also did better than Gamma.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">thresh</span> = 10
query[query[<span class="org-string">'npmle'</span>] &gt; query[<span class="org-string">'gamma'</span>] + thresh]
</pre>
</div>

<pre class="example">
method  dataset        gene       point       gamma  point_gamma    unimodal  \
12      dropseq  ERCC-00131 -723.520929 -356.755781  -341.742929 -347.518213

method       npmle
12     -337.359459
</pre>

<p>
For each control data set, for each expression model, report the number of
genes which have improvement in marginal log likelihood above some threshold
over the non-parametric expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">compare(llik[llik[<span class="org-string">'dataset'</span>].isin(control)], baseline=<span class="org-string">'npmle'</span>)
</pre>
</div>

<pre class="example">
method     point  gamma  point_gamma  unimodal
dataset
chromium1      0      1            0         2
chromium2      1      1            1         0
dropseq        0      0            0         0
gemcode        0      0            0         0
indrops        0      0            0         0
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">thresh</span> = 10
<span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik[llik[<span class="org-string">'dataset'</span>].isin(control)].groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
  <span class="org-variable-name">temp</span>[k] = query.loc[(query.sub(query[<span class="org-string">'npmle'</span>], axis=0) &gt; thresh).<span class="org-builtin">apply</span>(<span class="org-builtin">any</span>, axis=1)]
<span class="org-variable-name">query</span> = (pd.concat(temp)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
         [[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>, <span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>]])
query
</pre>
</div>

<pre class="example">
method    dataset        gene       point       gamma  point_gamma  \
0       chromium1  ERCC-00054 -140.374338 -139.785657  -139.403474
1       chromium1  ERCC-00116 -110.987988 -103.145967  -108.102592
2       chromium2  ERCC-00142  -28.006585  -28.006585   -28.011618

method    unimodal       npmle
0      -139.360976 -149.364211
1      -102.207858 -115.402711
2      -148.213405  -43.356179
</pre>

<p>
These cases can be explained by not starting the iterative refinement
procedure with a fine enough grid.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> query.groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">dat</span> = data[k]()
  <span class="org-keyword">for</span> _, row <span class="org-keyword">in</span> g.iterrows():
    temp[(k, row[<span class="org-string">'gene'</span>])] = scmodes.ebpm.ebpm_npmle(
      dat[:,dat.var.iloc[:,0] == row[<span class="org-string">'gene'</span>]].X.A.ravel(),
      dat.X.<span class="org-builtin">sum</span>(axis=1).A.ravel(),
      K=1024,
      max_grid_updates=10,
      tol=1e-5,
      verbose=<span class="org-constant">True</span>).rx2(<span class="org-string">'loglik'</span>)[0]
pd.Series(temp).reset_index().rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>, <span class="org-string">'level_1'</span>: <span class="org-string">'gene'</span>, 0: <span class="org-string">'npmle'</span>}, axis=1)
</pre>
</div>

<pre class="example">
dataset        gene       npmle
0  chromium1  ERCC-00054 -139.356572
1  chromium1  ERCC-00116 -102.119712
2  chromium2  ERCC-00142  -28.282124
</pre>
</div>
</div>

<div id="outline-container-orgfb7d234" class="outline-3">
<h3 id="biological-data"><a id="orgfb7d234"></a>Application to biological data sets</h3>
<div class="outline-text-3" id="text-biological-data">
<p>
Implement a helper function to plot.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">_plot</span>(llik, thresh):
  <span class="org-variable-name">query1</span> = compare(llik[llik[<span class="org-string">'dataset'</span>].isin(non_control)], baseline=<span class="org-string">'gamma'</span>, thresh=thresh, agg=<span class="org-string">'mean'</span>)
  <span class="org-variable-name">query2</span> = compare(llik[llik[<span class="org-string">'dataset'</span>].isin(non_control)], baseline=<span class="org-string">'unimodal'</span>, thresh=thresh, agg=<span class="org-string">'mean'</span>)
  <span class="org-variable-name">keys</span> = [<span class="org-string">'b_cells'</span>, <span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'ipsc'</span>, <span class="org-string">'cytotoxic_t-b_cells'</span>, <span class="org-string">'cytotoxic_t-naive_t'</span>, <span class="org-string">'kidney'</span>, <span class="org-string">'liver-caudate-lobe'</span>, <span class="org-string">'pbmc_10k_v3'</span>, <span class="org-string">'brain'</span>, <span class="org-string">'retina'</span>]
  <span class="org-variable-name">labels</span> = [<span class="org-string">'B cell'</span>, <span class="org-string">'T cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cyto/naive T'</span>, <span class="org-string">'Kidney'</span>, <span class="org-string">'Liver'</span>, <span class="org-string">'PBMC'</span>, <span class="org-string">'Brain'</span>, <span class="org-string">'Retina'</span>]
  plt.clf()
  <span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3)
  fig.set_size_inches(7.5, 3)
  ax[0].bar(<span class="org-builtin">range</span>(query1.shape[0]), query1.loc[keys, <span class="org-string">'point_gamma'</span>], color=<span class="org-string">'k'</span>)
  ax[0].set_xticks(<span class="org-builtin">range</span>(query1.shape[0]))
  ax[0].set_xticklabels(labels, rotation=90)
  ax[0].set_title(<span class="org-string">'Point-Gamma vs. Gamma'</span>)
  ax[0].set_ylabel(f<span class="org-string">'Fraction of genes with\nlikelihood ratio &gt; {thresh}'</span>)
  ax[1].bar(<span class="org-builtin">range</span>(query1.shape[0]), query1.loc[keys, <span class="org-string">'unimodal'</span>], color=<span class="org-string">'k'</span>)
  ax[1].set_xticks(<span class="org-builtin">range</span>(query1.shape[0]))
  ax[1].set_xticklabels(labels, rotation=90)
  ax[1].set_title(<span class="org-string">'Unimodal vs. Gamma'</span>)
  ax[1].set_xlabel(<span class="org-string">'Dataset'</span>)
  ax[2].bar(<span class="org-builtin">range</span>(query2.shape[0]), query2.loc[keys, <span class="org-string">'npmle'</span>], color=<span class="org-string">'k'</span>)
  ax[2].set_xticks(<span class="org-builtin">range</span>(query2.shape[0]))
  ax[2].set_xticklabels(labels, rotation=90)
  ax[2].set_title(<span class="org-string">'Non-parametric vs. Unimodal'</span>)
  fig.tight_layout()
</pre>
</div>

<p>
Report the fraction of genes with likelihood ratio &gt; 10 in favor of the more
complex expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">_plot(llik, thresh=10)
</pre>
</div>


<div class="figure">
<p><img src="figure/llr.org/biological-point_gamma-gamma-10.png" alt="biological-point_gamma-gamma-10.png">
</p>
</div>

<p>
Report the fraction of genes with likelihood ratio &gt; 50 in favor of the more
complex expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">_plot(llik, thresh=50)
</pre>
</div>


<div class="figure">
<p><img src="figure/llr.org/biological-point_gamma-gamma-50.png" alt="biological-point_gamma-gamma-50.png">
</p>
</div>

<p>
Report the fraction of genes with likelihood ratio &gt; 100 in favor of the
more complex expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">_plot(llik, thresh=100)
</pre>
</div>


<div class="figure">
<p><img src="figure/llr.org/biological-point_gamma-gamma-100.png" alt="biological-point_gamma-gamma-100.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org5fc1e78" class="outline-3">
<h3 id="drop-seq"><a id="org5fc1e78"></a>Drop-Seq spike-in gene</h3>
<div class="outline-text-3" id="text-drop-seq">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = data[<span class="org-string">'dropseq'</span>]()
<span class="org-variable-name">x</span> = dat[:,dat.var.iloc[:,0] == <span class="org-string">'ERCC-00131'</span>].X.A.ravel()
<span class="org-variable-name">s</span> = dat.X.<span class="org-builtin">sum</span>(axis=1).A.ravel()
<span class="org-variable-name">y</span> = np.arange(x.<span class="org-builtin">max</span>() + 1)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">grid</span> = np.linspace(0, (x / s).<span class="org-builtin">max</span>(), 1000)

<span class="org-variable-name">pmf</span> = <span class="org-builtin">dict</span>()
<span class="org-variable-name">cdf</span> = <span class="org-builtin">dict</span>()

<span class="org-variable-name">gamma_res</span> = scmodes.ebpm.ebpm_gamma(x, s)
<span class="org-variable-name">pmf</span>[<span class="org-string">'Gamma'</span>] = np.array([scmodes.benchmark.gof._zig_pmf(k, size=s, log_mu=gamma_res[0], log_phi=-gamma_res[1]).mean()
                         <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
<span class="org-variable-name">cdf</span>[<span class="org-string">'Gamma'</span>] = st.gamma(a=np.exp(gamma_res[1]), scale=np.exp(gamma_res[0] - gamma_res[1])).cdf(grid)

<span class="org-variable-name">point_gamma_res</span> = scmodes.ebpm.ebpm_point_gamma(x, s)
<span class="org-variable-name">cdf</span>[<span class="org-string">'ZIG'</span>] = sp.expit(point_gamma_res[2]) + sp.expit(-point_gamma_res[2]) * st.gamma(a=np.exp(point_gamma_res[1]), scale=np.exp(point_gamma_res[0] - point_gamma_res[1])).cdf(grid)
<span class="org-variable-name">pmf</span>[<span class="org-string">'ZIG'</span>] = np.array([scmodes.benchmark.gof._zig_pmf(k, size=s, log_mu=point_gamma_res[0], log_phi=-point_gamma_res[1], logodds=point_gamma_res[2]).mean()
                       <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])

<span class="org-comment-delimiter"># </span><span class="org-comment">Important: tune gridmult</span>
<span class="org-variable-name">unimodal_res</span> = ebpm_unimodal_tune(x, s)
<span class="org-variable-name">g</span> = np.array(unimodal_res.rx2(<span class="org-string">'fitted_g'</span>))
<span class="org-variable-name">g</span> = g[:,g[0] &gt; 1e-8]
<span class="org-variable-name">a</span> = np.fmin(g[1], g[2])
<span class="org-variable-name">b</span> = np.fmax(g[1], g[2])
<span class="org-variable-name">comp_dens_conv</span> = np.array([((st.gamma(a=k + 1, scale=1 / s.reshape(-1, 1)).cdf(b.reshape(1, -1)) - st.gamma(a=k + 1, scale=1 / s.reshape(-1, 1)).cdf(a.reshape(1, -1))) / np.outer(s, b - a)).mean(axis=0) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
<span class="org-variable-name">comp_dens_conv</span>[:,0] = st.poisson(mu=s.reshape(-1, 1) * b[0]).pmf(y).mean(axis=0)
<span class="org-variable-name">pmf</span>[<span class="org-string">'Unimodal'</span>] = comp_dens_conv @ g[0]
<span class="org-variable-name">cdf</span>[<span class="org-string">'Unimodal'</span>] = ashr.cdf_ash(unimodal_res, grid).rx2(<span class="org-string">'y'</span>).ravel()

<span class="org-variable-name">npmle_res</span> = scmodes.ebpm.ebpm_npmle(x, s, max_grid_updates=5, tol=1e-5)
<span class="org-variable-name">g</span> = np.array(npmle_res.rx2(<span class="org-string">'fitted_g'</span>))
<span class="org-variable-name">g</span> = g[:,g[0] &gt; 1e-8]
<span class="org-variable-name">a</span> = np.fmin(g[1], g[2])
<span class="org-variable-name">b</span> = np.fmax(g[1], g[2])
<span class="org-variable-name">comp_dens_conv</span> = np.array([((st.gamma(a=k + 1, scale=1 / s.reshape(-1, 1)).cdf(b.reshape(1, -1)) - st.gamma(a=k + 1, scale=1 / s.reshape(-1, 1)).cdf(a.reshape(1, -1))) / np.outer(s, b - a)).mean(axis=0) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> y])
<span class="org-variable-name">pmf</span>[<span class="org-string">'NPMLE'</span>] = comp_dens_conv @ g[0]
<span class="org-variable-name">cdf</span>[<span class="org-string">'NPMLE'</span>] = ashr.cdf_ash(npmle_res, grid).rx2(<span class="org-string">'y'</span>).ravel()
</pre>
</div>

<p>
1 - f4fd818b-e60b-4cc1-9c8b-69cfa67ba463
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Paired'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1)
fig.set_size_inches(4.5, 4)

ax[0].hist(x, bins=y, color=<span class="org-string">'0.7'</span>, density=<span class="org-constant">True</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[0].plot(y + .5, pmf[k], lw=1, c=cm(i))
ax[0].set_xlabel(<span class="org-string">'Number of molecules'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
ax[0].set_title(<span class="org-string">'ERCC-00131'</span>)

<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(pmf):
  ax[1].plot(grid, cdf[k], lw=1, c=cm(i), label=k)
ax[1].legend(frameon=<span class="org-constant">False</span>)
ax[1].set_xlabel(<span class="org-string">'Latent gene expression'</span>)
ax[1].set_ylabel(<span class="org-string">'CDF'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/llr.org/dropseq-ERCC-00131.png" alt="dropseq-ERCC-00131.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org50064dc" class="outline-3">
<h3 id="ash"><a id="org50064dc"></a>Issue in ash</h3>
<div class="outline-text-3" id="text-ash">
<p>
Find all genes where a simpler expression model (point, Gamma, point-Gamma)
did better than a unimodal expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik[llik[<span class="org-string">'dataset'</span>].isin(non_control)].groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
  <span class="org-variable-name">temp</span>[k] = query.loc[(query.sub(query[<span class="org-string">'unimodal'</span>], axis=0)[[<span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>]] &gt; thresh).<span class="org-builtin">apply</span>(<span class="org-builtin">any</span>, axis=1)]
<span class="org-variable-name">query</span> = (pd.concat(temp)
         .reset_index()
         .rename({<span class="org-string">'level_0'</span>: <span class="org-string">'dataset'</span>}, axis=1)
         [[<span class="org-string">'dataset'</span>, <span class="org-string">'gene'</span>, <span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'point_gamma'</span>, <span class="org-string">'unimodal'</span>, <span class="org-string">'npmle'</span>]])
query.shape
</pre>
</div>

<pre class="example">
(3253, 7)

</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org4bbf7ed"><span class="org-keyword">def</span> <span class="org-function-name">ebpm_unimodal_tune</span>(x, s):
  <span class="org-variable-name">x</span> = pd.Series(x)
  <span class="org-variable-name">s</span> = pd.Series(s)

  <span class="org-keyword">def</span> <span class="org-function-name">_loss_gridmult</span>(gridmult, mode, x, s):
    <span class="org-variable-name">llik</span> = ashr.ash_pois(x, s, mixcompdist=<span class="org-string">'halfuniform'</span>, mode=mode,
                         gridmult=gridmult, outputlevel=<span class="org-string">'loglik'</span>,
                         control=rpy2.robjects.ListVector({<span class="org-string">'tol.svd'</span>: 0})).rx2(<span class="org-string">'loglik'</span>)[0]
    <span class="org-keyword">print</span>(f<span class="org-string">'{mode:&gt;9.6g} {gridmult:&gt;9.6g} {llik:&gt;9.6g}'</span>)
    <span class="org-keyword">return</span> -llik

  <span class="org-keyword">def</span> <span class="org-function-name">_loss_mode</span>(mode, x, s):
    <span class="org-keyword">print</span>(f<span class="org-string">'{"mode":&gt;9s} {"gridmult":&gt;9s} {"llik":&gt;9s}'</span>)
    <span class="org-variable-name">opt</span> = so.minimize_scalar(_loss_gridmult, bracket=[1.1, 2], bounds=[1, 2],
                             method=<span class="org-string">'bounded'</span>, args=(mode, x, s), tol=1e-2)
    <span class="org-keyword">assert</span> opt.success
    <span class="org-keyword">return</span> opt.fun

  <span class="org-variable-name">opt</span> = so.minimize_scalar(_loss_mode, bracket=[0, (x / s).<span class="org-builtin">max</span>()],
                           bounds=[0, (x / s).<span class="org-builtin">max</span>()], method=<span class="org-string">'bounded'</span>, args=(x, s),
                           tol=1e-5)
  <span class="org-variable-name">mode</span> = opt.x
  <span class="org-keyword">assert</span> opt.success
  <span class="org-keyword">print</span>(f<span class="org-string">'Converged to mode={mode:&gt;9.6g}'</span>)

  <span class="org-variable-name">opt</span> = so.minimize_scalar(_loss_gridmult, bracket=[1.1, 2], bounds=[1, 2],
                           method=<span class="org-string">'bounded'</span>, args=(mode, x, s), tol=1e-2)
  <span class="org-keyword">assert</span> opt.success
  <span class="org-variable-name">gridmult</span> = opt.x
  <span class="org-keyword">print</span>(f<span class="org-string">'Converged to {gridmult:&gt;9.6g}'</span>)

  <span class="org-variable-name">res</span> = ashr.ash_pois(x, s, mixcompdist=<span class="org-string">'halfuniform'</span>, mode=mode, gridmult=gridmult, control=rpy2.robjects.ListVector({<span class="org-string">'tol.svd'</span>: 0}))
  <span class="org-keyword">return</span> res
</pre>
</div>

<p>
Report the fraction of genes for which NPMLE did better than the best of
point mass, Gamma, and unimodal.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">thresh</span> = 10
<span class="org-variable-name">temp</span> = <span class="org-builtin">dict</span>()
<span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> llik[llik[<span class="org-string">'dataset'</span>].isin(non_control)].groupby(<span class="org-string">'dataset'</span>):
  <span class="org-variable-name">query</span> = g.pivot_table(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'method'</span>, values=<span class="org-string">'llik'</span>)
  <span class="org-variable-name">baseline</span> = query[[<span class="org-string">'point'</span>, <span class="org-string">'gamma'</span>, <span class="org-string">'unimodal'</span>]].<span class="org-builtin">max</span>(axis=1)
  <span class="org-variable-name">temp</span>[k] = (query[<span class="org-string">'npmle'</span>] &gt; baseline + thresh).mean(axis=0)
<span class="org-variable-name">res</span> = (pd.Series(temp)
         .reset_index()
         .rename({<span class="org-string">'index'</span>: <span class="org-string">'dataset'</span>, 0: <span class="org-string">'npmle'</span>}, axis=1))
res
</pre>
</div>

<pre class="example">
dataset     npmle
0              b_cells  0.000156
1                brain  0.001448
2          cytotoxic_t  0.000766
3  cytotoxic_t-b_cells  0.003460
4  cytotoxic_t-naive_t  0.001441
5                 ipsc  0.001908
6               kidney  0.003293
7   liver-caudate-lobe  0.000074
8          pbmc_10k_v3  0.010293
9               retina  0.013935
</pre>

<p>
For each data set, for each expression model, report the number of genes
which have improvement in marginal log likelihood above some threshold over
the non-parametric expression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">compare(llik[llik[<span class="org-string">'dataset'</span>].isin(non_control)], baseline=<span class="org-string">'npmle'</span>, agg=<span class="org-string">'mean'</span>)
</pre>
</div>

<pre class="example">
method                  point     gamma  point_gamma  unimodal
dataset
b_cells              0.106436  0.116254     0.116098  0.093657
brain                0.067439  0.103713     0.108396  0.112568
cytotoxic_t          0.096325  0.101378     0.101378  0.058499
cytotoxic_t-b_cells  0.113886  0.129081     0.130886  0.115390
cytotoxic_t-naive_t  0.124560  0.133045     0.134806  0.100865
ipsc                 0.000000  0.000000     0.000000  0.000000
kidney               0.000000  0.017705     0.018140  0.018761
liver-caudate-lobe   0.000000  0.021605     0.021716  0.017582
pbmc_10k_v3          0.018281  0.034091     0.036397  0.037961
retina               0.000299  0.035334     0.036031  0.033841
</pre>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-09-07 Mon 14:23</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
