<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-07-25 Sat 12:42 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Low rank approximation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Low rank approximation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orga9ab311">Introduction</a></li>
<li><a href="#org5889bba">Setup</a></li>
<li><a href="#org77d27ef">Results</a>
<ul>
<li><a href="#org42bd0a9">Datasets</a></li>
<li><a href="#orgfa9be3b">Poisson thinning benchmark</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orga9ab311" class="outline-2">
<h2 id="orga9ab311">Introduction</h2>
<div class="outline-text-2" id="text-orga9ab311">
<p>
The key idea of our approach to modeling scRNA-seq is to separate sampling
variation and expression variation. This approach leads to the following
multi-gene model for scRNA-seq data: \(
  \newcommand\ml{\mathbf{L}}
  \newcommand\mf{\mathbf{F}}
  \newcommand\my{\mathbf{Y}}
  \)
</p>

\begin{align*}
  x_{ij} &\sim \operatorname{Poisson}(\lambda_{ij})\\
  \lambda_{ij} &= h^{-1}((\ml\mf')_{ij})
\end{align*}

<p>
where \(i = 1, \ldots, n\), \(j = 1, \ldots, p\), \(\ml\) is an \(n \times
  K\) matrix, and \(\mf\) is a \(p \times K\) matrix. (Here, we absorb the size
factor into \(\ml\).) A number of methods have been proposed to fit this
model (or extensions to it), several specialized for scRNA-seq data.
</p>

<ul class="org-ul">
<li>Non-negative matrix factorization
(<a href="https://papers.nips.cc/paper/1861-algorithms-for-non-negative-matrix-factorization.pdf">Lee
and Seung 2001</a>, <a href="https://arxiv.org/abs/1010.1763">Févotte and Idier
2011</a>)</li>
<li>Negative Binomial Matrix Factorization
(<a href="https://arxiv.org/abs/1801.01708">Gouvert et al. 2018</a>)</li>
<li>Latent dirichlet allocation
(<a href="http://www.jmlr.org/papers/volume3/blei03a/blei03a.pdf">Blei et al
2003</a>,
<a href="https://papers.nips.cc/paper/3902-online-learning-for-latent-dirichlet-allocation">Hoffman
et al. 2010</a>, <a href="http://proceedings.mlr.press/v22/taddy12.html">Taddy
2012</a>)</li>
<li>Hierarchical Bayesian Poisson Factorization
(<a href="https://www.hindawi.com/journals/cin/2009/785152/">Cemgil 2009</a>,
<a href="http://www.cs.columbia.edu/~blei/papers/GopalanHofmanBlei2015.pdf">Gopalan
et al. 2015</a>,
<a href="https://onlinelibrary.wiley.com/doi/full/10.15252/msb.20188557">Levitin
et al. 2019</a>)</li>
<li>ZINB-WAVE (<a href="https://www.nature.com/articles/s41467-017-02554-5">Risso et
al. 2018</a>)</li>
<li>scVI (<a href="https://www.nature.com/articles/s41592-018-0229-2">Lopez et al.
2018</a>)</li>
<li>DCA (<a href="https://www.nature.com/articles/s41467-018-07931-2">Eraslan et
al. 2019</a>)</li>
<li>GLM-PCA (<a href="https://www.biorxiv.org/content/10.1101/574574v1">Townes et
al. 2019</a>)</li>
</ul>

<p>
Here, we evaluate methods on their ability to estimate \(\lambda\). We use
Poisson thinning (<a href="https://arxiv.org/abs/1705.08393">Gerard and Stephens
2017</a>, <a href="https://www.biorxiv.org/content/10.1101/758524v1">Gerard 2019</a>) of
real data
</p>

\begin{align*}
  x_{ij} &\sim \mathrm{Poisson}(\lambda_{ij})\\
  y_{ij} &\sim \mathrm{Binomial}(x_{ij}, 0.5)\\
  \tilde{y}_{ij} &= x_{ij} - y_{ij}
\end{align*}

<p>
resulting in two matrices \(\my\) and \(\tilde{\my}\) with identical
\(\lambda\). Our benchmark estimates \(\ml, \mf\) from \(\my\) and evaluates
the estimate using the log likelihood of \(\tilde{\my}\).
</p>
</div>
</div>

<div id="outline-container-org5889bba" class="outline-2">
<h2 id="org5889bba">Setup</h2>
<div class="outline-text-2" id="text-org5889bba">
<div class="org-src-container">
<pre class="src src-ipython" id="org5d4dd9b"><span class="org-keyword">import</span> collections
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> os.path
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org77d27ef" class="outline-2">
<h2 id="org77d27ef">Results</h2>
<div class="outline-text-2" id="text-org77d27ef">
</div>
<div id="outline-container-org42bd0a9" class="outline-3">
<h3 id="org42bd0a9">Datasets</h3>
<div class="outline-text-3" id="text-org42bd0a9">
<p>
Take genes with non-zero counts present in at least 25% of cells. In the
iPSC data, this is all 9,957 genes, so take a random sample of 500 genes.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgbf99fa6"><span class="org-keyword">def</span> <span class="org-function-name">_read_10x</span>(k, min_detect=0.25, n_cells=1000, seed=1):
  <span class="org-keyword">return</span> scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k}/filtered_matrices_mex/hg19/'</span>, min_detect=0.25, return_df=<span class="org-constant">True</span>).sample(n=n_cells, axis=0, random_state=seed)

<span class="org-keyword">def</span> <span class="org-function-name">_mix_10x</span>(k1, k2, min_detect=0.25, n_cells=1000, seed=1):
  <span class="org-variable-name">x1</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k1}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-variable-name">x2</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{k2}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
  <span class="org-keyword">return</span> scmodes.dataset.synthetic_mix(x1, x2, min_detect=min_detect)[0].sample(n=n_cells, axis=0, random_state=seed)

<span class="org-keyword">def</span> <span class="org-function-name">_cd8_cd19_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>, **kwargs)

<span class="org-keyword">def</span> <span class="org-function-name">_cyto_naive_mix</span>(**kwargs):
  <span class="org-keyword">return</span> _mix_10x(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>, **kwargs)

<span class="org-variable-name">data</span> = {
  <span class="org-string">'cytotoxic_t'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'cytotoxic_t'</span>),
  <span class="org-string">'b_cells'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'b_cells'</span>),
  <span class="org-string">'ipsc'</span>: <span class="org-keyword">lambda</span>: scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>, return_df=<span class="org-constant">True</span>).sample(n=1000, axis=0, random_state=1).sample(n=500, axis=1, random_state=2),
  <span class="org-string">'cytotoxic_t-b_cells'</span>: _cd8_cd19_mix,
  <span class="org-string">'cytotoxic_t-naive_t'</span>: _cyto_naive_mix,
  <span class="org-string">'pbmcs_68k'</span>: <span class="org-keyword">lambda</span>: _read_10x(<span class="org-string">'fresh_68k_pbmc_donor_a'</span>),
}
</pre>
</div>

<p>
Report the data dimensions.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame([data[k]().shape <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data],
             columns=[<span class="org-string">'num_cells'</span>, <span class="org-string">'num_genes'</span>],
             index=data.keys())
</pre>
</div>

<pre class="example">
num_cells  num_genes
cytotoxic_t               1000        461
b_cells                   1000        375
ipsc                      1000        500
cytotoxic_t-b_cells       1000        404
cytotoxic_t-naive_t       1000        419
pbmcs_68k                 1000        404
</pre>

<p>
Report the proportion of zeros.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.Series({k: (data[k]() == 0).values.mean() <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data})
</pre>
</div>

<pre class="example">
cytotoxic_t            0.437247
b_cells                0.416360
ipsc                   0.301654
cytotoxic_t-b_cells    0.430270
cytotoxic_t-naive_t    0.437422
pbmcs_68k              0.444396
dtype: float64
</pre>
</div>
</div>

<div id="outline-container-orgfa9be3b" class="outline-3">
<h3 id="orgfa9be3b">Poisson thinning benchmark</h3>
<div class="outline-text-3" id="text-orgfa9be3b">
<p>
Run the CPU methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -n1 --mem=10G --time=24:00:00 --job-name=lra-generalization -a 31,39
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">ranks = [1, 2, 4, 8]</span>
<span class="org-sh-heredoc">methods = ['nmf', 'glmpca']</span>
<span class="org-sh-heredoc">tasks = [(d, m, r) for d in data for m in methods for r in ranks]</span>
<span class="org-sh-heredoc">d, m, r = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[d]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_lra_generalization(x, methods=[m], n_components=r, tol=1e-2, n_trials=5)</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/lra-generalization/{d}-{m}-{r}.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Run the GPU methods.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --time=60:00 --job-name=lra-generalization -a 0-47%4
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;imports&gt;&gt;</span>
<span class="org-sh-heredoc">import os</span>
<span class="org-sh-heredoc">&lt;&lt;data&gt;&gt;</span>
<span class="org-sh-heredoc">methods = ['pvae', 'nbvae']</span>
<span class="org-sh-heredoc">ranks = [1, 2, 4, 8]</span>
<span class="org-sh-heredoc">tasks = [(d, m, r) for d in data for m in methods for r in ranks]</span>
<span class="org-sh-heredoc">d, m, r = tasks[int(os.environ['SLURM_ARRAY_TASK_ID'])]</span>
<span class="org-sh-heredoc">x = data[d]()</span>
<span class="org-sh-heredoc">res = scmodes.benchmark.evaluate_lra_generalization(x, methods=[m], n_components=r, n_trials=5)</span>
<span class="org-sh-heredoc">res.to_csv(f'/scratch/midway2/aksarkar/modes/lra-generalization/{d}-{m}-{r}.txt.gz', compression='gzip', sep='\t')</span>
<span class="org-sh-heredoc">EOF</span>
</pre>
</div>

<p>
Collect the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">pois_thin_res</span> = collections.defaultdict(<span class="org-builtin">list</span>)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> data:
  <span class="org-keyword">for</span> m <span class="org-keyword">in</span> (<span class="org-string">'nmf'</span>, <span class="org-string">'glmpca'</span>, <span class="org-string">'pvae'</span>, <span class="org-string">'nbvae'</span>):
    <span class="org-keyword">for</span> r <span class="org-keyword">in</span> (1, 2, 4, 8):
      <span class="org-variable-name">f</span> = f<span class="org-string">'/scratch/midway2/aksarkar/modes/lra-generalization/{k}-{m}-{r}.txt.gz'</span>
      <span class="org-keyword">if</span> os.path.exists(f):
        pois_thin_res[(k, r)].append(pd.read_csv(f, header=[0, 1], index_col=0, sep=<span class="org-string">'\t'</span>))
<span class="org-variable-name">pois_thin_res</span> = pd.concat({k: pd.concat(pois_thin_res[k], axis=1) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> pois_thin_res}, axis=0)
<span class="org-variable-name">pois_thin_res.index.names</span> = [<span class="org-string">'data'</span>, <span class="org-string">'rank'</span>, <span class="org-string">'trial'</span>]
pois_thin_res.to_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/lra-generalization/lra-generalization.txt.gz'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">pois_thin_res</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/lra-generalization/lra-generalization.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=[0, 1], index_col=[0, 1, 2])
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">titles</span> = [<span class="org-string">'T cell'</span>, <span class="org-string">'B cell'</span>, <span class="org-string">'iPSC'</span>, <span class="org-string">'T cell/B cell'</span>, <span class="org-string">'Cytotoxic/naive T'</span>, <span class="org-string">'PBMC'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, (k, g), t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), pois_thin_res.xs(1, level=1)[<span class="org-string">'validation'</span>].groupby(level=0), titles):
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(g.columns):
    a.scatter(x + np.random.normal(scale=0.1, size=g.shape[0]), g[m].values, s=4, c=<span class="org-string">'k'</span>, zorder=3)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
  a.set_xlim(-0.5, g.shape[1] - 0.5)
  a.set_xticks(np.arange(g.shape[1]))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> g.columns], rotation=90)
  a.set_xlabel(<span class="org-string">'Method'</span>)
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Log likelihood of\nheld-out molecules'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/lra.org/lra-bench-1.png" alt="lra-bench-1.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, (k, g), t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), pois_thin_res.xs(2, level=1)[<span class="org-string">'validation'</span>].groupby(level=0), titles):
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(g.columns):
    a.scatter(x + np.random.normal(scale=0.1, size=g.shape[0]), g[m].values, s=4, c=<span class="org-string">'k'</span>, zorder=3)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
  a.set_xlim(-0.5, g.shape[1] - 0.5)
  a.set_xticks(np.arange(g.shape[1]))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> g.columns], rotation=90)
  a.set_xlabel(<span class="org-string">'Method'</span>)
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Log likelihood of\nheld-out molecules'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/lra.org/lra-bench-2.png" alt="lra-bench-2.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, (k, g), t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), pois_thin_res.xs(4, level=1)[<span class="org-string">'validation'</span>].groupby(level=0), titles):
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(g.columns):
    a.scatter(x + np.random.normal(scale=0.1, size=g.shape[0]), g[m].values, s=4, c=<span class="org-string">'k'</span>, zorder=3)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
  a.set_xlim(-0.5, g.shape[1] - 0.5)
  a.set_xticks(np.arange(g.shape[1]))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> g.columns], rotation=90)
  a.set_xlabel(<span class="org-string">'Method'</span>)
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Log likelihood of\nheld-out molecules'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/lra.org/lra-bench-4.png" alt="lra-bench-4.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 6, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2.5)
<span class="org-keyword">for</span> a, (k, g), t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax.ravel(), pois_thin_res.xs(8, level=1)[<span class="org-string">'validation'</span>].groupby(level=0), titles):
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(g.columns):
    a.scatter(x + np.random.normal(scale=0.1, size=g.shape[0]), g[m].values, s=4, c=<span class="org-string">'k'</span>, zorder=3)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
  a.set_xlim(-0.5, g.shape[1] - 0.5)
  a.set_xticks(np.arange(g.shape[1]))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> g.columns], rotation=90)
  a.set_xlabel(<span class="org-string">'Method'</span>)
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Log likelihood of\nheld-out molecules'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/lra.org/lra-bench-8.png" alt="lra-bench-8.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-07-25 Sat 12:42</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
