<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-09-08 Sun 21:44 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Low rank approximation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Low rank approximation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orga9ab311">Introduction</a></li>
<li><a href="#org5889bba">Setup</a></li>
<li><a href="#org9a4fac9">Methods</a>
<ul>
<li><a href="#org42bd0a9">Datasets</a></li>
<li><a href="#org1dfce43">Implementation</a></li>
</ul>
</li>
<li><a href="#org77d27ef">Results</a>
<ul>
<li><a href="#ipsc">PCA on (log) counts</a></li>
<li><a href="#orge68d487">Sorted cells benchmark</a></li>
<li><a href="#org924d356">Synthetic mixtures benchmark</a></li>
<li><a href="#pbmc">PBMCs benchmark</a></li>
<li><a href="#org443406e">Plot the results</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orga9ab311" class="outline-2">
<h2 id="orga9ab311">Introduction</h2>
<div class="outline-text-2" id="text-orga9ab311">
<p>
Suppose we have observations \(x_{ij} \sim
  \operatorname{Pois}(\lambda_{ij})\). Rather than making a distributional
assumption \(\lambda_j \sim g_j(\cdot)\) (as in
<a href="deconvolution.html">expression deconvolution</a>), we might make a <i>low
rank assumption</i> \(h(\lambda_{ij}) = [\mathbf{L F}]_{ij}\).  This assumption
can be interpreted as regularizing cells/genes towards each other, or as
discovering clusters in the data.
</p>

<p>
A number of methods have been proposed to estimate low rank structure from
count data, several specialized for scRNA-seq data. 
</p>

<ul class="org-ul">
<li>Non-negative matrix factorization
(<a href="https://papers.nips.cc/paper/1861-algorithms-for-non-negative-matrix-factorization.pdf">Lee
and Seung 2001</a>, <a href="https://arxiv.org/abs/1010.1763">FÃ©votte and Idier
2011</a>)</li>
<li>Latent dirichlet allocation
(<a href="http://www.jmlr.org/papers/volume3/blei03a/blei03a.pdf">Blei et al
2003</a>,
<a href="https://papers.nips.cc/paper/3902-online-learning-for-latent-dirichlet-allocation">Hoffman
et al. 2010</a>, <a href="http://proceedings.mlr.press/v22/taddy12.html">Taddy
2012</a>)</li>
<li>Hierarchical Bayesian Poisson Factorization
(<a href="http://www.cs.columbia.edu/~blei/papers/GopalanHofmanBlei2015.pdf">Gopalan
et al. 2015</a>,
<a href="https://onlinelibrary.wiley.com/doi/full/10.15252/msb.20188557">Levitin
et al. 2019</a>)</li>
<li>ZINB-WAVE (<a href="https://www.nature.com/articles/s41467-017-02554-5">Risso et
al. 2018</a>)</li>
<li>scVI (<a href="https://www.nature.com/articles/s41592-018-0229-2">Lopez et al.
2018</a>)</li>
<li>DCA (<a href="https://www.nature.com/articles/s41467-018-07931-2">Eraslan et
al. 2019</a>)</li>
<li>GLM-PCA (<a href="https://www.biorxiv.org/content/10.1101/574574v1">Townes et
al. 2019</a>)</li>
</ul>

<p>
There is one fundamental modelling decision which is not obvious from first
principles: what precisely is low rank? In other words, what is the
functional form of \(h(\cdot)\)? The methods above use either: \(h(x) = x\),
\(h(x) = \ln(x)\), or learn \(h\) from the data.
</p>

<p>
Here, we evaluate methods on their ability to generalize to new data. We use
real data, assuming:
</p>

\begin{align*}
  x_{ij} &\sim \mathrm{Poisson}(s_i \lambda_{ij})\\
  h(\lambda_{ij}) &= [\mathbf{L F}]_{ij}
\end{align*}

<p>
and hold out molecules by randomly thinning the observed counts:
</p>

\begin{align*}
  y_{ij} &\sim \mathrm{Binomial}(x_{ij}, 0.5)\\
  \tilde{y}_{ij} &= x_{ij} - y_{ij}\\
\end{align*}

<p>
This approach leaves the relative abundance of the transcripts unchanged in
expectation, implying that the low rank structure learned in \(\mathbf{Y}\)
should explain the data in \(\tilde{\mathbf{Y}}\).
</p>

<p>
Our metric is then the likelihood of the held-out data. We simply need to
re-scale to account for different size factors \(s_i\).
</p>
</div>
</div>

<div id="outline-container-org5889bba" class="outline-2">
<h2 id="org5889bba">Setup</h2>
<div class="outline-text-2" id="text-org5889bba">
<div class="org-src-container">
<pre class="src src-ipython" id="org5d4dd9b"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
<span class="org-keyword">import</span> sklearn.decomposition <span class="org-keyword">as</span> skd
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> wlra

rpy2.robjects.pandas2ri.activate()
<span class="org-variable-name">glmpca</span> = rpy2.robjects.packages.importr(<span class="org-string">'glmpca'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org9a4fac9" class="outline-2">
<h2 id="org9a4fac9">Methods</h2>
<div class="outline-text-2" id="text-org9a4fac9">
</div>
<div id="outline-container-org42bd0a9" class="outline-3">
<h3 id="org42bd0a9">Datasets</h3>
<div class="outline-text-3" id="text-org42bd0a9">
<p>
We use real scRNA-seq datasets in heterogeneous tissues to benchmark
methods:
</p>

<ol class="org-ol">
<li>Mouse neuron cells sequenced on the Fluidigm C1 platform
(<a href="https://science.sciencemag.org/content/347/6226/1138.full">Zeisel et
al. 2015</a>)</li>

<li>Fresh PBMCs sequenced on the 10X platform
(<a href="https://www.nature.com/articles/ncomms14049">Zheng et al. 2017</a>)</li>
</ol>
</div>
</div>

<div id="outline-container-org1dfce43" class="outline-3">
<h3 id="org1dfce43">Implementation</h3>
<div class="outline-text-3" id="text-org1dfce43">
<p>
Wrap around GLM-PCA.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">_glmpca</span>(x, n_components, max_iters=<span class="org-constant">None</span>, eps=<span class="org-constant">None</span>):
  <span class="org-variable-name">ctl</span> = {<span class="org-string">'maxIter'</span>: 100, <span class="org-string">'eps'</span>: 1e-4}
  <span class="org-keyword">if</span> max_iters <span class="org-keyword">is</span> <span class="org-keyword">not</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">ctl</span>[<span class="org-string">'maxIter'</span>] = max_iters
  <span class="org-keyword">if</span> eps <span class="org-keyword">is</span> <span class="org-keyword">not</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">ctl</span>[<span class="org-string">'eps'</span>] = eps
  <span class="org-variable-name">ctl</span> = rpy2.robjects.ListVector(ctl)
  <span class="org-comment-delimiter"># </span><span class="org-comment">GLM-PCA expects [genes, samples]</span>
  <span class="org-variable-name">res</span> = glmpca.glmpca(x.T, L=n_components, ctl=ctl)
  <span class="org-comment-delimiter"># </span><span class="org-comment">Follow GLM-PCA code for size factors, not the paper</span>
  <span class="org-variable-name">s</span> = np.log(x.mean(axis=1))
  <span class="org-variable-name">L</span> = np.array(res.rx2(<span class="org-string">'loadings'</span>))
  <span class="org-variable-name">F</span> = np.array(res.rx2(<span class="org-string">'factors'</span>))
  <span class="org-variable-name">etahat</span> = s.reshape(-1, 1) + F.T.dot(L)
  <span class="org-variable-name">trace</span> = np.array(res.rx2(<span class="org-string">'dev'</span>))
  <span class="org-keyword">return</span> etahat, trace
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org77d27ef" class="outline-2">
<h2 id="org77d27ef">Results</h2>
<div class="outline-text-2" id="text-org77d27ef">
</div>
<div id="outline-container-org8391426" class="outline-3">
<h3 id="ipsc"><a id="org8391426"></a>PCA on (log) counts</h3>
<div class="outline-text-3" id="text-ipsc">
<p>
Investigate whether PCA on counts or log counts gives better log likelihood.
We can write principal components analysis as a generative model (Tipping
1999):
</p>

<p>
\[ \mathbf{y}_i \sim \mathcal{N}(\mathbf{y}_i; \mathbf{W z}_i, \sigma^2 \mathbf{I}) \]
</p>

<p>
Now, suppose \(y_{ij} = \ln(x_{ij} + \epsilon)\). Then, via
change of variables we have:
</p>

<p>
\[ p(\mathbf{x}_i) = \frac{1}{\mathbf{x}_i + \epsilon}
   \mathcal{N}(\ln(\mathbf{x}_i + \epsilon); \mathbf{W z}_i, \sigma^2
   \mathbf{I}) \]
</p>

<p>
where we abuse notation for element-wise operations. In general, this means
we can compare the likelihood of the observed data under models of different
transformations of the observations.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org73942cb">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> sklearn.decomposition <span class="org-keyword">as</span> skd

<span class="org-variable-name">ipsc</span> = scmodes.dataset.ipsc(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/'</span>)
<span class="org-variable-name">res0</span> = skd.PCA(n_components=10).fit(ipsc)
<span class="org-variable-name">z_0</span> = res0.transform(ipsc)
<span class="org-variable-name">llik0</span> = st.norm(loc=z_0.dot(res0.components_), scale=np.sqrt(res0.noise_variance_)).logpdf(ipsc).<span class="org-builtin">sum</span>()

<span class="org-variable-name">log_ipsc</span> = np.log(ipsc + 1)
<span class="org-variable-name">res1</span> = skd.PCA(n_components=10).fit(log_ipsc)
<span class="org-variable-name">z_1</span> = res1.transform(ipsc)
<span class="org-variable-name">llik1</span> = (-log_ipsc + st.norm(loc=z_1.dot(res1.components_), scale=np.sqrt(res1.noise_variance_)).logpdf(log_ipsc)).<span class="org-builtin">sum</span>()
<span class="org-keyword">print</span>(f<span class="org-string">'PCA of untransformed counts: {llik0:.4g}'</span>)
<span class="org-keyword">print</span>(f<span class="org-string">'PCA of log1p counts: {llik1:.4g}'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=mstephens --mem=8G
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python &lt;&lt;EOF
<span class="org-sh-heredoc">&lt;&lt;pca&gt;&gt;</span>
EOF
</pre>
</div>

<pre class="example">
PCA of untransformed counts: -4.712e+08
PCA of log1p counts: -3.57e+09
</pre>
</div>
</div>

<div id="outline-container-orge68d487" class="outline-3">
<h3 id="orge68d487">Sorted cells benchmark</h3>
<div class="outline-text-3" id="text-orge68d487">
<div class="org-src-container">
<pre class="src src-ipython" id="orgc308207">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
<span class="org-variable-name">methods</span> = [<span class="org-string">'nmf'</span>, <span class="org-string">'glmpca'</span>, <span class="org-string">'pvae'</span>]
<span class="org-variable-name">datasets</span> = [<span class="org-string">'b_cells'</span>, <span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>]
<span class="org-variable-name">m</span> = methods[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-keyword">for</span> d <span class="org-keyword">in</span> datasets:
  <span class="org-variable-name">data</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{d}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>)
  <span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_lra_generalization(data, methods=[m], n_trials=10)
  res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/lra-generalization/{d}-{m}.txt.gz'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 -a 2 --time=2:00:00 --job-name=lra-benchmark
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python /project2/mstephens/aksarkar/projects/singlecell-modes/code/zheng-bench-lra.py
</pre>
</div>
</div>
</div>

<div id="outline-container-org924d356" class="outline-3">
<h3 id="org924d356">Synthetic mixtures benchmark</h3>
<div class="outline-text-3" id="text-org924d356">
<div class="org-src-container">
<pre class="src src-ipython" id="orgec82fc0">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
<span class="org-variable-name">methods</span> = [<span class="org-string">'pvae'</span>]
<span class="org-variable-name">datasets</span> = [(<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'b_cells'</span>), (<span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>)]
<span class="org-variable-name">d1</span>, <span class="org-variable-name">d2</span> = datasets[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-variable-name">x1</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{d1}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">x2</span> = scmodes.dataset.read_10x(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/{d2}/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>, min_detect=0)
<span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = scmodes.dataset.synthetic_mix(x1, x2, min_detect=.25)
<span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_lra_generalization(x, methods=methods, n_trials=10)
res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/lra-generalization/{d1}-{d2}-pvae.txt.gz'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 -a 0-1 --mem=16G --time=6:00:00 --job-name=lra-benchmark
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python /project2/mstephens/aksarkar/projects/singlecell-modes/code/synthetic-mix-lra.py
</pre>
</div>
</div>
</div>

<div id="outline-container-org00024ce" class="outline-3">
<h3 id="pbmc"><a id="org00024ce"></a>PBMCs benchmark</h3>
<div class="outline-text-3" id="text-pbmc">
<div class="org-src-container">
<pre class="src src-ipython">&lt;&lt;imports&gt;&gt;
<span class="org-keyword">import</span> os
<span class="org-variable-name">tasks</span> = [<span class="org-string">'nmf'</span>, <span class="org-string">'glmpca'</span>, <span class="org-string">'pvae'</span>]
<span class="org-variable-name">task</span> = tasks[<span class="org-builtin">int</span>(os.environ[<span class="org-string">'SLURM_ARRAY_TASK_ID'</span>])]
<span class="org-variable-name">data</span> = scmodes.dataset.read_10x(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/10xgenomics/fresh_68k_pbmc_donor_a/filtered_matrices_mex/hg19/'</span>, return_df=<span class="org-constant">True</span>)
<span class="org-variable-name">res</span> = scmodes.benchmark.evaluate_lra_generalization(data, methods=[task], n_trials=10)
res.to_csv(f<span class="org-string">'/scratch/midway2/aksarkar/modes/lra-generalization/pbmcs-{task}.txt.gz'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -a 1 -n1 -c28 --exclusive --time=6:00:00 --job-name=lra-benchmark
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python /project2/mstephens/aksarkar/projects/singlecell-modes/code/pbmc-bench-lra.py
</pre>
</div>

<p>
Run PVAE on GPU.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 -a 2 --mem=16G --time=6:00:00 --job-name=lra-benchmark
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scmodes
python /project2/mstephens/aksarkar/projects/singlecell-modes/code/pbmc-bench-lra.py
</pre>
</div>
</div>
</div>

<div id="outline-container-org443406e" class="outline-3">
<h3 id="org443406e">Plot the results</h3>
<div class="outline-text-3" id="text-org443406e">
<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh">rsync -au /scratch/midway2/aksarkar/modes/lra-generalization/ /project2/mstephens/aksarkar/projects/singlecell-modes/data/lra-generalization/
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">methods</span> = [<span class="org-string">'nmf'</span>, <span class="org-string">'glmpca'</span>, <span class="org-string">'pvae'</span>]
<span class="org-variable-name">data</span> = [<span class="org-string">'b_cells'</span>, <span class="org-string">'cytotoxic_t'</span>, <span class="org-string">'naive_t'</span>, <span class="org-string">'cytotoxic_t-b_cells'</span>, <span class="org-string">'cytotoxic_t-naive_t'</span>, <span class="org-string">'pbmcs'</span>]
<span class="org-variable-name">titles</span> = [<span class="org-string">'B cell'</span>, <span class="org-string">'Cytotoxic T'</span>, <span class="org-string">'Naive T'</span>, <span class="org-string">'B cell/T cell'</span>, <span class="org-string">'Naive/cytotoxic T'</span>, <span class="org-string">'PBMC'</span>]
<span class="org-variable-name">results</span> = {k: pd.concat([pd.read_csv(f<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-modes/data/lra-generalization/{k}-{m}.txt.gz'</span>,
                                     header=[0, 1],
                                     index_col=0)
                         <span class="org-keyword">for</span> m <span class="org-keyword">in</span> methods], axis=<span class="org-string">'columns'</span>)
           <span class="org-keyword">for</span> k <span class="org-keyword">in</span> data}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, <span class="org-builtin">len</span>(results), sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 2)
<span class="org-keyword">for</span> a, k, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax, results, titles):
  <span class="org-variable-name">llik</span> = results[k][<span class="org-string">'validation'</span>]
  <span class="org-keyword">for</span> x, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(llik):
    a.scatter(x + np.random.normal(scale=0.1, size=llik.shape[0]), llik[m].values, s=1, c=<span class="org-string">'k'</span>, zorder=3)
  a.grid(c=<span class="org-string">'0.8'</span>, lw=1, axis=<span class="org-string">'x'</span>)
  a.set_xlim(-0.5, 2.5)
  a.set_xticks(np.arange(3))
  a.set_xticklabels([m.upper() <span class="org-keyword">for</span> m <span class="org-keyword">in</span> methods], rotation=90)
  a.set_title(t)
ax[0].set_ylabel(<span class="org-string">'Validation log lik'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/lra.org/mlcb-bench-lra.png" alt="mlcb-bench-lra.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-09-08 Sun 21:44</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
