<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-03-14 Thu 21:42 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Technical zero-generating mechanism in scRNA-seq data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Technical zero-generating mechanism in scRNA-seq data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org669fb46">Introduction</a></li>
<li><a href="#org1eab7a8">Setup</a></li>
<li><a href="#orgde1f3e7">Joint distribution of expression across genes</a></li>
<li><a href="#org63fce9e">Single cell/bulk concordance</a></li>
<li><a href="#orgc9e381a">Mode on zero</a></li>
</ul>
</div>
</div>

<div id="outline-container-org669fb46" class="outline-2">
<h2 id="org669fb46">Introduction</h2>
<div class="outline-text-2" id="text-org669fb46">
<p>
Important recent work has demonstrated that Multinomial (equivalently,
Poisson) sampling is sufficient to model the noise introduced by sequencing
(<a href="https://dx.doi.org/10.1073/pnas.1721085115">Wang et al. 2018</a>,
<a href="https://www.biorxiv.org/content/early/2019/03/11/574574">Townes et
al. 2019</a>). Here, we show that previously cited evidence of a technical
zero-generating mechanism can be replicated in simulated data without such a
mechanism.
</p>
</div>
</div>

<div id="outline-container-org1eab7a8" class="outline-2">
<h2 id="org1eab7a8">Setup</h2>
<div class="outline-text-2" id="text-org1eab7a8">
<p>
Submitted batch job 51133345
</p>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])

<span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> gzip
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> sklearn.neighbors <span class="org-keyword">as</span> skn

<span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
<span class="org-keyword">import</span> rpy2.robjects.numpy2ri

rpy2.robjects.pandas2ri.activate()
rpy2.robjects.numpy2ri.activate()
<span class="org-variable-name">mclust</span> = rpy2.robjects.packages.importr(<span class="org-string">'mclust'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgde1f3e7" class="outline-2">
<h2 id="orgde1f3e7">Joint distribution of expression across genes</h2>
<div class="outline-text-2" id="text-orgde1f3e7">
<p>
Reproduce <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4112276/bin/nihms-590625-f0001.jpg">Fig 1C</a> of <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4112276/">Kharchenko et al 2014</a>. They analyzed scRNA-Seq of mouse
embryonic fibroblast cells generated by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3129258/">Islam et al 2011</a>.
</p>

<div class="org-src-container">
<pre class="src src-sh">curl -s -o GSE29087_L139_expression_tab.txt.gz <span class="org-string">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE29087&amp;format=file&amp;file=GSE29087_L139_expression_tab.txt.gz"</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-R">islam_counts <span class="org-ess-assignment">&lt;-</span> read.delim(gzfile(<span class="org-string">'/scratch/midway2/aksarkar/modes/GSE29087_L139_expression_tab.txt.gz'</span>), skip=6, header=F)
x <span class="org-ess-assignment">&lt;-</span> log10(islam_counts[,92] + 1)
y <span class="org-ess-assignment">&lt;-</span> log10(islam_counts[,93] + 1)
keep <span class="org-ess-assignment">&lt;-</span> x != 0 | y != 0
png(<span class="org-string">'temp.png'</span>, width=4, height=4, res=300, units=<span class="org-string">"in"</span>)
smoothScatter(x[keep],y[keep] )
dev.off()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/islam-smoothscatter.png" alt="islam-smoothscatter.png" width="300">
</p>
</div>

<p>
Parse the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgbac7ac6"><span class="org-keyword">with</span> gzip.<span class="org-builtin">open</span>(<span class="org-string">'/scratch/midway2/aksarkar/modes/GSE29087_L139_expression_tab.txt.gz'</span>, <span class="org-string">'rt'</span>) <span class="org-keyword">as</span> f:
  <span class="org-builtin">next</span>(f)  <span class="org-comment-delimiter"># </span><span class="org-comment">first three lines are description</span>
  <span class="org-builtin">next</span>(f)
  <span class="org-builtin">next</span>(f)
  <span class="org-variable-name">barcodes</span> = <span class="org-builtin">next</span>(f).split()
  <span class="org-variable-name">samples</span> = <span class="org-builtin">next</span>(f).split()
  <span class="org-builtin">next</span>(f)  <span class="org-comment-delimiter"># </span><span class="org-comment">skip header</span>
  <span class="org-variable-name">features</span> = []
  <span class="org-variable-name">counts</span> = []
  <span class="org-keyword">for</span> line <span class="org-keyword">in</span> f:
    <span class="org-variable-name">row</span> = line.split(<span class="org-string">'\t'</span>)
    features.append(row[:6])
    counts.append([<span class="org-builtin">int</span>(x) <span class="org-keyword">if</span> x <span class="org-keyword">else</span> np.nan <span class="org-keyword">for</span> x <span class="org-keyword">in</span> row[6:]])
  <span class="org-variable-name">counts</span> = np.ma.masked_invalid(np.array(counts).T)
</pre>
</div>

<p>
Plot the joint distribution of \(\log_2\) expression values across a pair of
cells.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_joint_dist_two_cells</span>(x, y):
  <span class="org-variable-name">x</span> = np.log(x + 1) / np.log(2)
  <span class="org-variable-name">y</span> = np.log(y + 1) / np.log(2)
  <span class="org-variable-name">keep</span> = (x + y) &gt; 0

  plt.clf()
  plt.gcf().set_size_inches(3, 3)
  plt.hexbin(x[keep], y[keep], gridsize=25, bins=<span class="org-string">'log'</span>, cmap=colorcet.cm[<span class="org-string">'bgyw_r'</span>])
  <span class="org-variable-name">H</span>, <span class="org-variable-name">xx</span>, <span class="org-variable-name">yy</span> = np.histogram2d(x[keep], y[keep], bins=30)
  <span class="org-variable-name">gridx</span>, <span class="org-variable-name">gridy</span> = np.meshgrid(xx[1:] + (xx[1] - xx[0]) / 2, yy[1:] + (yy[1] - yy[0]) / 2)
  <span class="org-variable-name">low_pass</span> = np.logical_and(H &gt; 0, H &lt; 3)
  plt.scatter(gridx[low_pass].ravel(), gridy[low_pass].ravel(), c=<span class="org-string">'k'</span>, s=H[low_pass], alpha=0.5)

  plt.xlim([-0.5, 16])
  plt.ylim([-0.5, 16])
  plt.xticks(np.arange(0, 16, 5))
  plt.yticks(np.arange(0, 16, 5))
  plt.xlabel(<span class="org-string">'$\log_2(r + 1)$ of cell 1'</span>)
  plt.ylabel(<span class="org-string">'$\log_2(r + 1)$ of cell 2'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_joint_dist_two_cells(counts[91], counts[92])
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/islam-mef.png" alt="islam-mef.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span> = np.log(counts[91] + 1) / np.log(2)
<span class="org-variable-name">y</span> = np.log(counts[92] + 1) / np.log(2)
plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(x, y, c=<span class="org-string">'k'</span>, s=1, alpha=0.1)
plt.axis(<span class="org-string">'equal'</span>)
plt.xlabel(<span class="org-string">'$\log_2(r + 1)$ of cell 1'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'$\log_2(r + 1)$ of cell 2'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/islam-mef-scatter.png" alt="islam-mef-scatter.png">
</p>
</div>

<p>
Now, simulate count data, assuming 25,000 molecules per
cell and a range of parameters <a href="https://users.rcc.uchicago.edu/~aksarkar/singlecell-qtl/figure/zinb.org/joint-distribution.png">matching real data</a>:
</p>

<p>
\[ r_{ij} \sim \mathrm{Poisson}(R_i \lambda_{ij}) \]
</p>

<p>
\[ \lambda_{ij} \sim \mathrm{Gamma}(\mu, \phi) \]
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">simulate_counts</span>(num_cells, num_genes, size, nb=<span class="org-constant">False</span>, zi=<span class="org-constant">False</span>, seed=<span class="org-constant">None</span>):
  <span class="org-keyword">if</span> seed <span class="org-keyword">is</span> <span class="org-keyword">not</span> <span class="org-constant">None</span>:
    np.random.seed(seed)
  <span class="org-variable-name">mu</span> = np.exp(np.random.uniform(-15, -4, size=(1, num_genes)))
  <span class="org-keyword">if</span> nb:
    <span class="org-variable-name">phi</span> = np.exp(np.random.uniform(-4, 1, size=(1, num_genes)))
    <span class="org-variable-name">u</span> = np.random.gamma(phi, phi, size=(num_cells, num_genes))
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">u</span> = 1
  <span class="org-variable-name">x</span> = np.random.poisson(lam=size * mu * u, size=(num_cells, num_genes))
  <span class="org-keyword">return</span> x
</pre>
</div>

<p>
First, simulate Poisson data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = simulate_counts(num_cells=2, num_genes=10000, size=2.5e4, seed=0)
plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(np.log(x + 1) / np.log(2), np.log(y + 1) / np.log(2), c=<span class="org-string">'k'</span>, s=1)
plt.xlabel(<span class="org-string">'$\log_2(r + 1)$ of cell 1'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'$\log_2(r + 1)$ of cell 2'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/pois.png" alt="pois.png">
</p>
</div>

<p>
Now, simulate negative binomial data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = simulate_counts(num_cells=2, num_genes=10000, size=2.5e4, nb=<span class="org-constant">True</span>, seed=0)
plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(np.log(x + 1) / np.log(2), np.log(y + 1) / np.log(2), c=<span class="org-string">'k'</span>, s=1)
plt.xlabel(<span class="org-string">'$\log_2(r + 1)$ of cell 1'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'$\log_2(r + 1)$ of cell 2'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/nb.png" alt="nb.png">
</p>
</div>

<p>
The results suggest that the observed pattern of zeros in the joint
distribution of genes across cells can arise simply from Poisson
sampling. The observed outliers can be explained by overdispersion as well.
</p>
</div>
</div>

<div id="outline-container-org63fce9e" class="outline-2">
<h2 id="org63fce9e">Single cell/bulk concordance</h2>
<div class="outline-text-2" id="text-org63fce9e">
<p>
Reproduce Fig. 2 of <a href="https://academic.oup.com/biostatistics/article-lookup/doi/10.1093/biostatistics/kxx053">Hicks et al. 2017</a> (<a href="https://github.com/stephaniehicks/scBatchPaper/blob/master/scripts/createFigures.Rmd">associated repository</a>), which is Fig.
1C of <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3683364/">Shalek et al. 2013</a>.
</p>

<p>
Download the data.
</p>

<div class="org-src-container">
<pre class="src src-sh">curl <span class="org-string">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE41265&amp;format=file&amp;file=GSE41265_allGenesTPM.txt.gz"</span> -o GSE41265_allGenesTPM.txt.gz
</pre>
</div>

<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">shalek_tpm</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/ideas/GSE41265_allGenesTPM.txt.gz'</span>)
</pre>
</div>

<p>
Shalek et al filtered genes that don't have TPM &gt; 1 in at least three cells.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">shalek_tpm_pass</span> = shalek_tpm.loc[((shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>) &gt; 1).<span class="org-builtin">sum</span>(axis=1) &gt;= 3).values]
shalek_tpm_pass.shape
</pre>
</div>

<pre class="example">
(6312, 22)

</pre>

<p>
Plot 10K populations against each other.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 2)
fig.set_size_inches(6, 6)

ax[0][0].scatter(np.log(shalek_tpm_pass[<span class="org-string">'P1'</span>] + 1), np.log(shalek_tpm_pass[<span class="org-string">'P2'</span>] + 1), c=<span class="org-string">'k'</span>, s=2, alpha=0.1)
ax[0][0].set_xlabel(<span class="org-string">'Population 1 ln (TPM + 1)'</span>)
ax[0][0].set_ylabel(<span class="org-string">'Population 2 ln (TPM + 1)'</span>)

ax[0][1].scatter(np.log(shalek_tpm_pass[<span class="org-string">'P1'</span>] + 1), np.log(shalek_tpm_pass[<span class="org-string">'P3'</span>] + 1), c=<span class="org-string">'k'</span>, s=2, alpha=0.1)
ax[0][1].set_xlabel(<span class="org-string">'Population 1 ln (TPM + 1)'</span>)
ax[0][1].set_ylabel(<span class="org-string">'Population 3 ln (TPM + 1)'</span>)

ax[1][0].scatter(np.log(shalek_tpm_pass[<span class="org-string">'P2'</span>] + 1), np.log(shalek_tpm_pass[<span class="org-string">'P3'</span>] + 1), c=<span class="org-string">'k'</span>, s=2, alpha=0.1)
ax[1][0].set_xlabel(<span class="org-string">'Population 2 ln (TPM + 1)'</span>)
ax[1][0].set_ylabel(<span class="org-string">'Population 3 ln (TPM + 1)'</span>)

ax[1][1].set_axis_off()
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/shalek-fig-1a.png" alt="shalek-fig-1a.png">
</p>
</div>

<p>
Plot the average single cell log (TPM + 1) against each 10K population log
(TPM + 1)
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3)
fig.set_size_inches(9, 3)
<span class="org-keyword">for</span> i, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax):
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: "When calculating the &#8220;average&#8221; single-cell expression level, we</span>
  <span class="org-comment-delimiter"># </span><span class="org-comment">first averaged TPM levels from each of the 18 single cells, and then</span>
  <span class="org-comment-delimiter"># </span><span class="org-comment">transformed this average estimate into log space." (Shalek et al 2013)</span>
  a.scatter(np.log(shalek_tpm_pass.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1) + 1), np.log(shalek_tpm_pass[f<span class="org-string">'P{i + 1}'</span>] + 1), s=2, alpha=0.1, c=<span class="org-string">'k'</span>)
  a.plot([0, 12], [0, 12], c=<span class="org-string">'r'</span>, lw=1)
  a.set_title(f<span class="org-string">'Population {i + 1}'</span>)
  a.set_xlabel(<span class="org-string">'10K cells log (TPM + 1)'</span>)
ax[0].set_ylabel(<span class="org-string">'Single cell sample mean log (TPM + 1)'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/shalek-fig-1c.png" alt="shalek-fig-1c.png">
</p>
</div>

<p>
Reproduce Fig. 2A of Hicks et al.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(np.log(shalek_tpm[<span class="org-string">'P1'</span>] + 1) / np.log(2), np.log(shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1) + 1) / np.log(2), c=<span class="org-string">'k'</span>, s=1, alpha=0.1)
plt.plot([0, 10], [0, 10], c=<span class="org-string">'r'</span>, lw=1)
plt.xlabel(<span class="org-string">'Population 1 $\log_2(\mathrm{TPM} + 1)$'</span>)
plt.ylabel(<span class="org-string">'Single cell mean $\log_2(\mathrm{TPM} + 1)$'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Single cell mean $\\log_2(\\mathrm{TPM} + 1)$')

</pre>

<div class="figure">
<p><img src="figure/pitfalls.org/hicks-fig-2a.png" alt="hicks-fig-2a.png">
</p>
</div>

<p>
Reproduce Fig. 2B of Hicks et al. In their plotting code, they truncate the
\(y\) limits to \([-5, 5]\), which actually cuts off some data.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3)
fig.set_size_inches(8.5, 3)

<span class="org-keyword">for</span> i, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax):
  <span class="org-variable-name">M</span> = (np.log(shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1) + 1) / np.log(2) + np.log(shalek_tpm[f<span class="org-string">'P{i + 1}'</span>] + 1) / np.log(2)) / 2
  <span class="org-variable-name">A</span> = np.log(shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1) + 1) / np.log(2) - np.log(shalek_tpm[f<span class="org-string">'P{i + 1}'</span>] + 1) / np.log(2)
  a.scatter(M, A, s=1, c=<span class="org-string">'k'</span>, alpha=0.1)

  <span class="org-variable-name">grid</span> = np.linspace(0, M.<span class="org-builtin">max</span>(), 25)
  a.plot(grid, A.groupby(np.digitize(M, bins=grid)).agg(np.mean), c=<span class="org-string">'r'</span>, ls=<span class="org-string">'--'</span>)

  a.set_title(f<span class="org-string">'Population {i + 1}'</span>)
  a.set_xlabel(<span class="org-string">'M (mean average)'</span>)
  a.set_ylabel(<span class="org-string">'A (log ratio)'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/hicks-fig-2b.png" alt="hicks-fig-2b.png">
</p>
</div>

<p>
From these plots, Hicks et al. conclude that the single cell expression is
systematically smaller than the bulk expression for lowly expressed genes.
</p>

<p>
Now, TPM is proportional to relative abundance, where the constant of
proportionality is constant across samples (<a href="https://dx.doi.org/10.1007/s12064-012-0162-3">Wagner et al. 2012</a>). 
</p>

<p>
However, it is still possible that TPM can be systematically different across
many genes simply because some gene is highly expressed in one sample,
altering the relative abundances of all other genes. Therefore, naive
comparison of TPM could give misleading results.
</p>

<p>
To investigate whether this is the case, instead plot relative abundances
against each other.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(9, 3)

<span class="org-variable-name">sc_mean_tpm</span> = shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1)
<span class="org-variable-name">sc_ln_abundance</span> = np.ma.masked_invalid(np.log(sc_mean_tpm / sc_mean_tpm.<span class="org-builtin">sum</span>()))
<span class="org-variable-name">lim</span> = [-50,
       <span class="org-builtin">max</span>(bulk_ln_abundance.<span class="org-builtin">max</span>(), sc_ln_abundance.<span class="org-builtin">max</span>())]

<span class="org-keyword">for</span> i, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax):
  <span class="org-variable-name">bulk_ln_abundance</span> = np.ma.masked_invalid(np.log(shalek_tpm[f<span class="org-string">'P{i + 1}'</span>] / shalek_tpm[f<span class="org-string">'P{i + 1}'</span>].<span class="org-builtin">sum</span>()))

  a.scatter(bulk_ln_abundance, sc_ln_abundance, s=1, c=<span class="org-string">'0.75'</span>, alpha=0.1)

  <span class="org-variable-name">outlier</span> = np.where(np.logical_or(sc_ln_abundance &lt; -30, bulk_ln_abundance &lt; -30).filled(0))[0]
  a.scatter(bulk_ln_abundance[outlier], sc_ln_abundance[outlier], s=4, c=<span class="org-string">'k'</span>)
  <span class="org-keyword">for</span> o <span class="org-keyword">in</span> outlier:
    a.text(bulk_ln_abundance[o], sc_ln_abundance[o], shalek_tpm.iloc[o][<span class="org-string">'GENE'</span>])

  a.plot(lim, lim, c=<span class="org-string">'r'</span>, lw=1)
  a.set_xlabel(f<span class="org-string">'Population {i + 1} ln relative abundance'</span>)
  ax[0].set_ylabel(<span class="org-string">'Single cell ln relative abundance'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/shalek-abundances.png" alt="shalek-abundances.png">
</p>
</div>

<p>
From these plots, we would conclude that the relative abundances in bulk are
systematically smaller than in single cell, which is exactly the opposite of
the conclusion in Hicks et al. 
</p>

<p>
Our results are explained by the fact that exactly two genes, <i>ACAT3</i> and
<i>HSPA1B</i>, have systematically higher estimated abundance in bulk versus
single cell.
</p>

<p>
Reproduce Fig. 3 of Hicks et al. 2017. The data were generated in Shalek et
al 2013, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4122333/">Trapnell et al. 2014</a>, and <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4022966/">Wu et al. 2014</a>.
</p>

<p>
Download the data generated by Trapnell et al.
</p>

<div class="org-src-container">
<pre class="src src-sh">curl <span class="org-string">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE52529&amp;format=file&amp;file=GSE52529_fpkm_matrix.txt.gz"</span> -o GSE52529_fpkm_matrix.txt.gz
curl <span class="org-string">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE52529&amp;format=file&amp;file=GSE52529_truseq_fpkm_matrix.txt.gz"</span> -o GSE52529_truseq_fpkm_matrix.txt.gz
</pre>
</div>

<p>
Download the data generated by Wu et al.
</p>

<div class="org-src-container">
<pre class="src src-sh">curl -O --ftp-pasv <span class="org-string">"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE51nnn/GSE51254/suppl/GSE51254_RAW.tar"</span>
tar xf GSE51254_RAW.tar
</pre>
</div>

<p>
Wu et al generated 2 bulk RNA-Seq replicates for each of two different
technologies, and 96 scRNA-Seq libraries using the C1 platform.
</p>

<p>
The RNA-Seq data were quantified using cufflinks. Concatenate the output to
simplify parsing.
</p>

<div class="org-src-container">
<pre class="src src-sh">zcat GSM*bulk*.sorted.genes.fpkm_tracking.txt.gz | cut -f10 | gzip &gt;wu-bulk-fpkm.txt.gz
zcat GSM*_C*_IL*.sorted.genes.fpkm_tracking.txt.gz | cut -f10 | gzip &gt;wu-c1-fpkm.txt.gz
</pre>
</div>

<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">trapnell_sc_fpkm</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/ideas/GSE52529_fpkm_matrix.txt.gz'</span>)
<span class="org-variable-name">trapnell_bulk_fpkm</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/ideas/GSE52529_truseq_fpkm_matrix.txt.gz'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">wu_bulk_fpkm</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/ideas/wu-bulk-fpkm.txt.gz'</span>, header=<span class="org-constant">None</span>).values.reshape(4, -1).T
<span class="org-variable-name">wu_sc_fpkm</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/ideas/wu-c1-fpkm.txt.gz'</span>, header=<span class="org-constant">None</span>).values.reshape(-1, wu_bulk_fpkm.shape[0]).T
</pre>
</div>

<p>
Plot single cell mean against bulk expression. <b>Important: Hicks et al
truncate the visualization.</b>
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3)
fig.set_size_inches(9, 3)

ax[0].scatter(shalek_tpm[<span class="org-string">'P1'</span>], shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[0].set_xlim(0, 30)
ax[0].set_ylim(0, 30)
<span class="org-variable-name">grid</span> = np.linspace(0, 30, 15)
<span class="org-variable-name">sm</span> = shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>).mean(axis=1).groupby(np.digitize(shalek_tpm[<span class="org-string">'P1'</span>], bins=grid)).agg(np.mean)
ax[0].plot(grid[sm.index - 1], sm, c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">'--'</span>)


ax[1].scatter(wu_bulk_fpkm[:,0], wu_sc_fpkm.mean(axis=1), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[1].set_xlim(0, 50)
ax[1].set_ylim(0, 50)
<span class="org-variable-name">grid</span> = np.linspace(0, 30, 15)
<span class="org-variable-name">sm</span> = pd.Series(wu_sc_fpkm.mean(axis=1)).groupby(np.digitize(wu_bulk_fpkm[:,0], bins=grid)).agg(np.mean)
ax[1].plot(grid[sm.index - 1][1:], sm[1:], c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">'--'</span>)

ax[2].scatter(trapnell_bulk_fpkm[<span class="org-string">'T72_0'</span>], trapnell_sc_fpkm.mean(axis=1), s=1, c=<span class="org-string">'k'</span>, alpha=0.1)
ax[2].set_xlim(0, 30)
ax[2].set_ylim(0, 30)
<span class="org-variable-name">grid</span> = np.linspace(0, 30, 15)
<span class="org-variable-name">sm</span> = trapnell_sc_fpkm.mean(axis=1).groupby(np.digitize(trapnell_bulk_fpkm[<span class="org-string">'T72_0'</span>], bins=grid)).agg(np.mean)
ax[2].plot(grid[sm.index - 1], sm, c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">'--'</span>)

<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'Bulk TPM'</span>)
  a.set_ylabel(<span class="org-string">'Single cell mean TPM'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/hicks-fig-3.png" alt="hicks-fig-3.png">
</p>
</div>

<p>
As we argued above, this comparison is not sensible. Instead, plot relative
abundances against each other.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">get_abundance</span>(bulk, sc):
  <span class="org-variable-name">bulk</span> = np.ma.masked_values(bulk / bulk.<span class="org-builtin">sum</span>(), 0)
  <span class="org-variable-name">mean</span> = sc.mean(axis=1)
  <span class="org-variable-name">sc</span> = np.ma.masked_values(mean / mean.<span class="org-builtin">sum</span>(), 0)
  <span class="org-keyword">return</span> bulk, sc

<span class="org-variable-name">shalek_bulk_abundance</span>, <span class="org-variable-name">shalek_sc_abundance</span> = get_abundance(shalek_tpm[<span class="org-string">'P1'</span>], shalek_tpm.<span class="org-builtin">filter</span>(like=<span class="org-string">'S'</span>))
<span class="org-variable-name">wu_bulk_abundance</span>, <span class="org-variable-name">wu_sc_abundance</span> = get_abundance(wu_bulk_fpkm[:,0], wu_sc_fpkm)
<span class="org-variable-name">trapnell_bulk_abundance</span>, <span class="org-variable-name">trapnell_sc_abundance</span> = get_abundance(trapnell_bulk_fpkm[<span class="org-string">'T72_0'</span>], trapnell_sc_fpkm)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(8, 3)

<span class="org-variable-name">lim</span> = [5e-9, .1]
ax[0].set_ylabel(<span class="org-string">'Single cell mean relative abundance'</span>)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xlabel(<span class="org-string">'Bulk relative abundance'</span>)
  a.set_xscale(<span class="org-string">'log'</span>)
  a.set_yscale(<span class="org-string">'log'</span>)
  a.plot(lim, lim, c=<span class="org-string">'r'</span>, lw=1)

ax[0].scatter(shalek_bulk_abundance, shalek_sc_abundance, c=<span class="org-string">'k'</span>, s=2, alpha=0.1)
ax[0].set_title(<span class="org-string">'Shalek et al.'</span>)

ax[1].scatter(wu_bulk_abundance, wu_sc_abundance, c=<span class="org-string">'k'</span>, s=1, alpha=0.1)
ax[1].set_title(<span class="org-string">'Wu et al.'</span>)

ax[2].scatter(trapnell_bulk_abundance, trapnell_sc_abundance, c=<span class="org-string">'k'</span>, s=1, alpha=0.1)
ax[2].set_title(<span class="org-string">'Trapnell et al.'</span>)

fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/sc-bulk-abundance.png" alt="sc-bulk-abundance.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgc9e381a" class="outline-2">
<h2 id="orgc9e381a">Mode on zero</h2>
<div class="outline-text-2" id="text-orgc9e381a">
<p>
Reproduce <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4112276/bin/nihms-590625-f0001.jpg">Fig. 1E</a> of <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4112276/">Kharchenko et al 2014</a>. We observed a similar
relationship between \(\mathrm{logit}(\pi)\) and \(\ln\mu\) in iPSCs.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_mu</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">log_phi</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">logodds</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-logodds.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = log_mu[<span class="org-string">'NA18507'</span>].align(logodds[<span class="org-string">'NA18507'</span>], join=<span class="org-string">'inner'</span>)
plt.scatter(x, y, c=<span class="org-string">'k'</span>, s=1, alpha=0.25)
plt.xlabel(<span class="org-string">'$\ln(\mu)$'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'$\mathrm{logit}(\pi)$'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/ipsc-zinb.png" alt="ipsc-zinb.png">
</p>
</div>

<p>
Now, simulate Poisson data and compute the empirical fraction of zeros.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span> = simulate_counts(num_cells=1000, num_genes=10000, size=2.5e4, seed=2)
plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(np.log(x.mean(axis=0) / 2.5e4), (x == 0).mean(axis=0), c=<span class="org-string">'k'</span>, s=1, alpha=0.25)
plt.xlabel(<span class="org-string">'$\ln(\hat\mu)$'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'Empirical fraction of zeros'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/pois-pi0.png" alt="pois-pi0.png">
</p>
</div>

<p>
Do the same for negative binomial data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span> = simulate_counts(num_cells=1000, num_genes=10000, size=2.5e4, nb=<span class="org-constant">True</span>, seed=2)
plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.scatter(np.log(x.mean(axis=0) / 2.5e4), (x == 0).mean(axis=0), c=<span class="org-string">'k'</span>, s=1, alpha=0.25)
plt.xlabel(<span class="org-string">'$\ln(\hat\mu)$'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'Empirical fraction of zeros'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/pitfalls.org/nb-pi0.png" alt="nb-pi0.png">
</p>
</div>

<p>
The results suggest that even non-zero inflated data will have a sigmoidal
relationship between the fraction of zeros and mean expression. 
</p>

<p>
However, the results also suggest that zero-inflation is needed to explain
the particular shape of the dependence observed in real data.
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-03-14 Thu 21:42</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
