<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2019-04-29 Mon 18:14 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Transformations of deconvolved gene expression distributions</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="css/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/supp.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Transformations of deconvolved gene expression distributions</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5e4bead">Introduction</a></li>
<li><a href="#org7acce31">Setup</a></li>
<li><a href="#orgfd058f2">Methods</a></li>
<li><a href="#org48a149b">Results</a>
<ul>
<li><a href="#orgb3b5997">Untransformed distribution</a></li>
<li><a href="#orgfa4ab82">Log transform</a></li>
<li><a href="#org2249f09">Square root transform</a></li>
<li><a href="#org111a2b2">Variance effects versus mean effects</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org5e4bead" class="outline-2">
<h2 id="org5e4bead">Introduction</h2>
<div class="outline-text-2" id="text-org5e4bead">
<p>
From the perspective of <a href="deconvolution.html">distribution
deconvolution</a>, scRNA-seq count data has a mean-variance relationship simply
by convolution with a Poisson technical noise model.
</p>

<p>
Here, we investigate whether the latent gene expression distribution also has
a mean-variance relationship, and what happens to it when we transform the
distribution.
</p>
</div>
</div>

<div id="outline-container-org7acce31" class="outline-2">
<h2 id="org7acce31">Setup</h2>
<div class="outline-text-2" id="text-org7acce31">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scmodes
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgfd058f2" class="outline-2">
<h2 id="orgfd058f2">Methods</h2>
<div class="outline-text-2" id="text-orgfd058f2">
<p>
The key idea of our approach is to deconvolve the gene expression of iPSCs
assuming:
</p>

<p>
\[ x_{ijk} \sim \mathrm{Poisson}(s_{ij} \exp(\mathbf{z_i}' \beta_j)
  \lambda_{ijk}) \]
</p>

<p>
\[ \lambda_{ijk} \sim g_{ik}(\cdot) = \pi_{ik} \delta_0(\cdot) + (1 - \pi_ik)
  \mathrm{Gamma}(\mu_{ik}, \phi_{ik}) \]
</p>

<p>
From the fitted \(\hat{g}_1, \ldots, \hat{g}_p\), we can investigate whether
there is a mean-variance relationship in latent gene expression.
</p>
</div>
</div>

<div id="outline-container-org48a149b" class="outline-2">
<h2 id="org48a149b">Results</h2>
<div class="outline-text-2" id="text-org48a149b">
</div>
<div id="outline-container-orgb3b5997" class="outline-3">
<h3 id="orgb3b5997">Untransformed distribution</h3>
<div class="outline-text-3" id="text-orgb3b5997">
<p>
Read the deconvolved distributions.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_mu</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">logodds</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-logodds.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
</pre>
</div>

<p>
Throw out the individual with evidence of contamination.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">for</span> x <span class="org-keyword">in</span> (log_mu, log_phi, logodds):
  <span class="org-keyword">del</span> x[<span class="org-string">'NA18498'</span>]
</pre>
</div>

<p>
Recover the mean and variance of latent gene expression per gene, per individual.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-comment-delimiter"># </span><span class="org-comment">Important: log(sigmoid(x)) = -softplus(-x)</span>
<span class="org-variable-name">mean</span> = np.exp(log_mu - np.log1p(np.exp(logodds)))
<span class="org-variable-name">variance</span> = np.exp(2 * log_mu + log_phi - np.log1p(np.exp(logodds))) + np.exp(-np.log1p(np.exp(logodds)) - np.log1p(np.exp(-logodds)) + 2 * log_mu)
</pre>
</div>

<p>
Plot the mean/variance relationship.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.semilogx()
plt.semilogy()
plt.scatter(mean.values.ravel(), variance.values.ravel(), c=<span class="org-string">'k'</span>, s=4, alpha=0.1)
plt.xlabel(<span class="org-string">'Latent gene expression mean'</span>)
plt.ylabel(<span class="org-string">'Latent gene expression variance'</span>)
</pre>
</div>

<pre class="example">
Text(0, 0.5, 'Latent gene expression variance')

</pre>

<div class="figure">
<p><img src="figure/transforms.org/zig-mean-var.png" alt="zig-mean-var.png">
</p>
</div>

<p>
Restrict to the individual with the most cells (\(n=281\)).
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.semilogx()
plt.semilogy()
plt.scatter(mean[<span class="org-string">'NA18507'</span>], variance[<span class="org-string">'NA18507'</span>].values.ravel(), c=<span class="org-string">'k'</span>, s=4, alpha=0.1)
plt.title(<span class="org-string">'NA18507'</span>)
plt.xlabel(<span class="org-string">'Latent gene expression mean'</span>)
plt.ylabel(<span class="org-string">'Latent gene expression variance'</span>)
</pre>
</div>

<pre class="example">
Text(0, 0.5, 'Latent gene expression variance')

</pre>

<div class="figure">
<p><img src="figure/transforms.org/zig-mean-var-NA18507.png" alt="zig-mean-var-NA18507.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgfa4ab82" class="outline-3">
<h3 id="orgfa4ab82">Log transform</h3>
<div class="outline-text-3" id="text-orgfa4ab82">
<p>
Assume \(\lambda_{ijk} \sim g_{ik}(\cdot)\), as above. Plot the approximate
relationship between \(E[\ln(\lambda + \epsilon)]\) and
\(V[\ln(\lambda + \epsilon)]\) by first-order Taylor expansion.
</p>

<p>
\[ E[\ln(x + \epsilon)] \approx \ln(E[x] + \epsilon) - \frac{V[x]}{(E[x] + \epsilon)^2} \]
</p>

<p>
\[ V[\ln(x + \epsilon)] \approx \frac{2 V[x]}{(E[x] + \epsilon)^2} - \frac{V[x]^2}{(E[x] + \epsilon)^4} \]
</p>

<p>
Importantly, \(\epsilon\) needs to be on the same scale as \(\lambda\). But
\(x \sim \mathrm{Poisson}(s_i \lambda)\), so \(\lambda\) is on the scale of
\(1 / s_i\). Instead of assuming \(x_{ijk} \sim F_{ijk}(\cdot)\), assume:
</p>

<p>
\[ x_{ijk} \sim \frac{1}{n} \sum_j F_{ijk}(\cdot) \]
</p>

<p>
Then, \(E[x_{ijk}] = \frac{1}{n}\sum_j s_{ij} E[\lambda_{ijk}]\). This gives
us a single scaling factor for \(g_{ik}\).
</p>

<p>
Read the metadata for the iPSCs.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">annotations</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">keep_samples</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
<span class="org-variable-name">annotations</span> = annotations.loc[keep_samples.values.ravel()]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">s</span> = annotations.loc[(annotations[<span class="org-string">'chip_id'</span>] == <span class="org-string">'NA18507'</span>).values, <span class="org-string">'mol_hs'</span>].mean()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 3)
<span class="org-keyword">for</span> a, eps, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax, np.array([0, 0.5, 1]) / s, [<span class="org-string">'0'</span>, <span class="org-string">'1/(2s)'</span>, <span class="org-string">'1/s'</span>]):
  a.set_yscale(<span class="org-string">'log'</span>)
  a.scatter(np.log(mean[<span class="org-string">'NA18507'</span>] + eps), (2 * variance[<span class="org-string">'NA18507'</span>]) / np.square(mean[<span class="org-string">'NA18507'</span>] + eps), c=<span class="org-string">'k'</span>, s=4, alpha=0.1)
  a.set_title(f<span class="org-string">'$\epsilon = {t}$'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent log gene expression mean'</span>)
ax[0].set_ylabel(<span class="org-string">'Latent log gene expression variance'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/transforms.org/zig-log-mean-var-NA18507.png" alt="zig-log-mean-var-NA18507.png">
</p>
</div>

<p>
Zoom in on highly expressed genes.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 3)
<span class="org-keyword">for</span> a, eps, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax, np.array([0, 0.5, 1]) / s, [<span class="org-string">'0'</span>, <span class="org-string">'1/(2s)'</span>, <span class="org-string">'1/s'</span>]):
  a.set_yscale(<span class="org-string">'log'</span>)
  a.set_xlim(-8, -6)
  a.set_ylim(1e-2, 10)
  a.scatter(np.log(mean[<span class="org-string">'NA18507'</span>] + eps), (2 * variance[<span class="org-string">'NA18507'</span>]) / np.square(mean[<span class="org-string">'NA18507'</span>] + eps), c=<span class="org-string">'k'</span>, s=4, alpha=0.25)
  a.set_title(f<span class="org-string">'$\epsilon = {t}$'</span>)
ax[1].set_xlabel(<span class="org-string">'Latent log gene expression mean'</span>)
ax[0].set_ylabel(<span class="org-string">'Latent log gene expression variance'</span>)
fig.tight_layout()

</pre>
</div>


<div class="figure">
<p><img src="figure/transforms.org/zig-log-mean-var-NA18507-inset.png" alt="zig-log-mean-var-NA18507-inset.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org2249f09" class="outline-3">
<h3 id="org2249f09">Square root transform</h3>
<div class="outline-text-3" id="text-org2249f09">
<p>
Assume \(\lambda_{ijk} \sim g_{ik}(\cdot)\), as above. Plot the approximate
relationship between \(E[\lambda^{1/2}]\) and
\(V[\lambda^{1/2}]\) by first-order Taylor expansion.
</p>

<p>
\[ E[x^{1/2}] \approx (E[x])^{1/2} - \frac{V[x]}{4 E[x]^{3/2}} \]
</p>

<p>
\[ V[x^{1/2}] \approx \frac{V[x]}{2 E[x]} - \frac{V[x]^2}{16 E[x]^3} \]
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.semilogx()
plt.semilogy()
plt.scatter(np.sqrt(mean[<span class="org-string">'NA18507'</span>]), variance[<span class="org-string">'NA18507'</span>] / (2 * mean[<span class="org-string">'NA18507'</span>]), c=<span class="org-string">'k'</span>, s=4, alpha=0.1)
plt.xlabel(<span class="org-string">'Latent sqrt gene expression mean'</span>)
plt.ylabel(<span class="org-string">'Latent sqrt gene expression variance'</span>)
plt.gcf().tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/transforms.org/zig-sqrt-mean-var-NA18507.png" alt="zig-sqrt-mean-var-NA18507.png">
</p>
</div>
</div>
</div>
<div id="outline-container-org111a2b2" class="outline-3">
<h3 id="org111a2b2">Variance effects versus mean effects</h3>
<div class="outline-text-3" id="text-org111a2b2">
<p>
Look at 
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2019-04-29 Mon 18:14</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
